// ngrx-forms/esm5/src/array/reducer/move-control.js
import * as tslib_1 from "tslib";
import { MoveArrayControlAction } from '../../actions';
import { computeArrayState } from '../../state';
import { childReducer, updateIdRecursive } from './util';
export function move(array, fromIndex, toIndex) {
    var item = array[fromIndex];
    var length = array.length;
    if (fromIndex > toIndex) {
        return tslib_1.__spread(array.slice(0, toIndex), [
            item
        ], array.slice(toIndex, fromIndex), array.slice(fromIndex + 1, length));
    }
    else {
        var targetIndex = toIndex + 1;
        return tslib_1.__spread(array.slice(0, fromIndex), array.slice(fromIndex + 1, targetIndex), [
            item
        ], array.slice(targetIndex, length));
    }
}
export function moveControlReducer(state, action) {
    if (action.type !== MoveArrayControlAction.TYPE) {
        return state;
    }
    if (action.controlId !== state.id) {
        return childReducer(state, action);
    }
    var fromIndex = action.fromIndex;
    var toIndex = action.toIndex;
    if (fromIndex === toIndex) {
        return state;
    }
    if (fromIndex < 0 || toIndex < 0) {
        throw new Error("fromIndex " + fromIndex + " or toIndex " + fromIndex + " was negative");
    }
    if (fromIndex >= state.controls.length || toIndex >= state.controls.length) {
        throw new Error("fromIndex " + fromIndex + " or toIndex " + toIndex + " is out of bounds with the length of the controls " + state.controls.length);
    }
    var controls = move(state.controls, fromIndex, toIndex);
    controls = controls.map(function (c, i) { return updateIdRecursive(c, state.id + "." + i); });
    return computeArrayState(state.id, controls, state.value, state.errors, state.pendingValidations, state.userDefinedProperties, {
        wasOrShouldBeDirty: true,
        wasOrShouldBeEnabled: state.isEnabled,
        wasOrShouldBeTouched: state.isTouched,
        wasOrShouldBeSubmitted: state.isSubmitted,
    });
}
