// ngrx-forms/esm5/src/array/reducer/util.js
import * as tslib_1 from "tslib";
import { formStateReducer } from '../../reducer';
import { computeArrayState, isArrayState, isGroupState } from '../../state';
export function dispatchActionPerChild(controls, actionCreator) {
    var hasChanged = false;
    var newControls = controls
        .map(function (state) {
        var newState = formStateReducer(state, actionCreator(state.id));
        hasChanged = hasChanged || state !== newState;
        return newState;
    });
    return hasChanged ? newControls : controls;
}
function callChildReducers(controls, action) {
    var hasChanged = false;
    var newControls = controls
        .map(function (state) {
        var newState = formStateReducer(state, action);
        hasChanged = hasChanged || state !== newState;
        return newState;
    });
    return hasChanged ? newControls : controls;
}
export function childReducer(state, action) {
    var controls = callChildReducers(state.controls, action);
    if (state.controls === controls) {
        return state;
    }
    return computeArrayState(state.id, controls, state.value, state.errors, state.pendingValidations, state.userDefinedProperties, {
        wasOrShouldBeDirty: state.isDirty,
        wasOrShouldBeEnabled: state.isEnabled,
        wasOrShouldBeTouched: state.isTouched,
        wasOrShouldBeSubmitted: state.isSubmitted,
    });
}
export function updateIdRecursiveForGroup(state, newId) {
    var controls = Object.keys(state.controls)
        .reduce(function (agg, key) {
        var _a;
        return Object.assign(agg, (_a = {},
            _a[key] = updateIdRecursive(state.controls[key], newId + "." + key),
            _a));
    }, {});
    return tslib_1.__assign({}, state, { id: newId, controls: controls });
}
export function updateIdRecursiveForArray(state, newId) {
    var controls = state.controls.map(function (c, i) { return updateIdRecursive(c, newId + "." + i); });
    return tslib_1.__assign({}, state, { id: newId, controls: controls });
}
export function updateIdRecursive(state, newId) {
    if (state.id === newId) {
        return state;
    }
    if (isGroupState(state)) {
        return updateIdRecursiveForGroup(state, newId);
    }
    if (isArrayState(state)) {
        return updateIdRecursiveForArray(state, newId);
    }
    return tslib_1.__assign({}, state, { id: newId });
}
