// survey-lib/esm2015/lib/survey-lib.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/survey-lib.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CUSTOM_ELEMENTS_SCHEMA, NgModule, NO_ERRORS_SCHEMA } from '@angular/core';
import { SurveyLibComponent } from './survey-lib.component';
import { PagesService } from './shared/pages-services/pages.service';
import { PersistenceService } from './shared/persistence.service';
import { PagesComponent } from './pages/pages.component';
import { WelcomePageComponent } from './pages/welcome-page/welcome-page.component';
import { QuestionnaireComponent } from './pages/questionnaire/questionnaire.component';
import { ThankYouPageComponent } from './pages/thank-you-page/thank-you-page.component';
import { ChoiceComponent } from './pages/questionnaire/choice/choice.component';
import { YesNoComponent } from './pages/questionnaire/yes-no/yes-no.component';
import { ShortTextComponent } from './pages/questionnaire/short-text/short-text.component';
import { LongTextComponent } from './pages/questionnaire/long-text/long-text.component';
import { StatementComponent } from './pages/questionnaire/statement/statement.component';
import { EmailComponent } from './pages/questionnaire/email/email.component';
import { NpsComponent } from './pages/questionnaire/nps/nps.component';
import { DateComponent } from './pages/questionnaire/date/date.component';
import { RatingComponent } from './pages/questionnaire/rating/rating.component';
import { DropDownComponent } from './pages/questionnaire/drop-down/drop-down.component';
import { ContactFormComponent } from './pages/questionnaire/contact-form/contact-form.component';
import { AddressComponent } from './pages/questionnaire/address/address.component';
import { PhoneComponent } from './pages/questionnaire/phone/phone.component';
import { NameComponent } from './pages/questionnaire/name/name.component';
import { AgreeDisagreeComponent } from './pages/questionnaire/agree-disagree/agree-disagree.component';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { StoreModule } from '@ngrx/store';
import { surveyReducers } from './pages/store/survey.reducers';
import { WebsiteComponent } from './pages/questionnaire/website/website.component';
import { NumberComponent } from './pages/questionnaire/number/number.component';
import { ShufflePipe } from './shared/pipe/shuffle.pipe';
import { AutofocusDirective } from './shared/directives/auto-focus.directive';
import { DefaultThankYouPageComponent } from './pages/default-thank-you-page/default-thank-you-page.component';
export class SurveyLibModule {
}
SurveyLibModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    SurveyLibComponent,
                    PagesComponent,
                    WelcomePageComponent,
                    QuestionnaireComponent,
                    ThankYouPageComponent,
                    ChoiceComponent,
                    YesNoComponent,
                    ShortTextComponent,
                    LongTextComponent,
                    StatementComponent,
                    EmailComponent,
                    NpsComponent,
                    DateComponent,
                    RatingComponent,
                    DropDownComponent,
                    ContactFormComponent,
                    AddressComponent,
                    PhoneComponent,
                    NameComponent,
                    AgreeDisagreeComponent,
                    WebsiteComponent,
                    NumberComponent,
                    DefaultThankYouPageComponent,
                    ShufflePipe,
                    AutofocusDirective
                ],
                imports: [
                    CommonModule,
                    ReactiveFormsModule,
                    FormsModule,
                    StoreModule.forRoot({ survey: surveyReducers })
                ],
                exports: [SurveyLibComponent],
                providers: [PagesService, PersistenceService],
                schemas: [CUSTOM_ELEMENTS_SCHEMA, NO_ERRORS_SCHEMA]
            },] }
];
