// survey-lib/esm5/lib/pages/questionnaire/contact-form/contact-form.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/questionnaire/contact-form/contact-form.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { Question } from '../../pages.model';
import { FormControl, FormGroup, Validators } from '@angular/forms';
import { PagesService } from '../../../shared/pages-services/pages.service';
var ContactFormComponent = /** @class */ (function () {
    function ContactFormComponent(pagesService) {
        this.pagesService = pagesService;
    }
    /**
     * @return {?}
     */
    ContactFormComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.createForm();
    };
    /**
     * @return {?}
     */
    ContactFormComponent.prototype.createForm = /**
     * @return {?}
     */
    function () {
        var e_1, _a;
        /** @type {?} */
        var formGroup = {};
        try {
            for (var _b = tslib_1.__values(Object.keys(this.contactForm.contactForm)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var prop = _c.value;
                formGroup[prop] = new FormControl(this.contactForm.contactForm[prop].value || '', ContactFormComponent.mapValidators(this.contactForm.contactForm[prop].validation));
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        this.form = new FormGroup(formGroup);
    };
    /**
     * @private
     * @param {?} validators
     * @return {?}
     */
    ContactFormComponent.mapValidators = /**
     * @private
     * @param {?} validators
     * @return {?}
     */
    function (validators) {
        var e_2, _a;
        /** @type {?} */
        var formValidators = [];
        if (validators) {
            try {
                for (var _b = tslib_1.__values(Object.keys(validators)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var validation = _c.value;
                    if (validation === 'required') {
                        formValidators.push(Validators.required);
                    }
                    else if (validation === 'min') {
                        formValidators.push(Validators.min(validators[validation]));
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
        return formValidators;
    };
    /**
     * @return {?}
     */
    ContactFormComponent.prototype.onSubmit = /**
     * @return {?}
     */
    function () {
        this.response.contactInfo = this.objectProps;
        this.response.questionId = this.contactForm.questionId;
        this.pagesService.persistSurvey(this.response);
        //console.log(this.dataObject);
        /*  for (let controlsKey in this.form.controls) {
            this.dataObject[controlsKey].value = form[controlsKey];
          }
          console.log(this.dataObject);
          this.response.questionId = this.contactForm.questionId;*/
        //this.response.contactInfo
    };
    ContactFormComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-contact-form',
                    template: "\n<div>\n\n  <form novalidate (ngSubmit)=\"onSubmit()\" [formGroup]=\"form\">\n    <div *ngFor=\"let prop of objectProps\">\n      <label [attr.for]=\"prop\">{{prop.label}}</label>\n\n      <div [ngSwitch]=\"prop.type\">\n        <input *ngSwitchCase=\"'text'\"\n               [formControlName]=\"prop.key\"\n               [id]=\"prop.key\" [type]=\"prop.type\"\n               [(ngModel)] = \"prop.value\">\n\n      </div>\n<!--\n      <div class=\"error\" *ngIf=\"form.get(prop.key).invalid && (form.get(prop.key).dirty || form.get(prop.key).touched)\">\n        <div *ngIf=\"form.get(prop.key).errors.required\">\n          <h6>{{ prop.label }} is required.</h6>\n        </div>\n      </div>-->\n\n\n    </div>\n    <p>\n      <button type=\"submit\">Done</button>\n    </p>\n  </form>\n\n</div>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ContactFormComponent.ctorParameters = function () { return [
        { type: PagesService }
    ]; };
    ContactFormComponent.propDecorators = {
        contactForm: [{ type: Input }]
    };
    return ContactFormComponent;
}());
export { ContactFormComponent };
if (false) {
    /** @type {?} */
    ContactFormComponent.prototype.contactForm;
    /** @type {?} */
    ContactFormComponent.prototype.form;
    /** @type {?} */
    ContactFormComponent.prototype.objectProps;
    /**
     * @type {?}
     * @private
     */
    ContactFormComponent.prototype.response;
    /**
     * @type {?}
     * @private
     */
    ContactFormComponent.prototype.surveyState;
    /**
     * @type {?}
     * @private
     */
    ContactFormComponent.prototype.pagesService;
}
