// survey-lib/esm5/lib/pages/questionnaire/rating/rating.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/questionnaire/rating/rating.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Question } from '../../pages.model';
import { SurveyResponses } from '../../survey-response.model';
import { PagesService } from '../../../shared/pages-services/pages.service';
var RatingComponent = /** @class */ (function () {
    function RatingComponent(pageService) {
        this.pageService = pageService;
        this.submitEvent = new EventEmitter();
        this.isActive = false;
    }
    /**
     * @return {?}
     */
    RatingComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.pageService.stage == 'entry-next') {
            document.getElementById('rating-question').className = 'entry-next';
            document.getElementById('rating-question-delay').className = 'entry-next-delay';
        }
        else if (this.pageService.stage == 'entry-previous') {
            document.getElementById('rating-question').className = 'entry-previous-delay';
            document.getElementById('rating-question-delay').className = 'entry-previous';
        }
    };
    /**
     * @return {?}
     */
    RatingComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        console.log(this.ratingQuestion.attributes.steps.length);
        if (this.ratingQuestion.attributes.steps.length > 6) {
            /** @type {?} */
            var els = document.getElementsByTagName("label");
            console.log(els[0]);
            for (var i = 0, all = els.length; i < all; i++) {
                els[i].classList.add('ratingIconAbove6');
            }
        }
    };
    /**
     * @param {?} rating
     * @return {?}
     */
    RatingComponent.prototype.onRatingSelect = /**
     * @param {?} rating
     * @return {?}
     */
    function (rating) {
        var _this = this;
        console.warn(rating);
        this.isActive = true;
        this.surveyResponse.question = this.ratingQuestion.question;
        this.surveyResponse.questionType = this.ratingQuestion.questionType;
        this.surveyResponse.questionId = this.ratingQuestion.questionId;
        this.surveyResponse.rating = rating;
        this.surveyResponse.identifier = this.ratingQuestion.identifier;
        this.pageService.persistSurvey(this.surveyResponse);
        console.log(this.ratingQuestion.attributes.ratingIcon);
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.submitEvent.emit();
        }), 350);
    };
    /**
     * @return {?}
     */
    RatingComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
    };
    RatingComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-rating',
                    template: "<div id=\"rating-question\">\n<div class=\"survey-statement\">\n  <span *ngIf=\"ratingQuestion.attributes.required\" class=\"isRequired\">*</span>{{ratingQuestion.question}}</div>\n  <p class=\"question-description\" [ngStyle]=\"{'display':'ratingQuestion.attributes.description == undefined' ? 'none' :'block'}\">\n  {{ratingQuestion.attributes.description}}\n</p>\n\n</div>\n\n\n\n<div id=\"rating-question-delay\">\n<div class=\"rating\">\n    <ng-container *ngFor=\"let step of ratingQuestion.attributes.steps; let stepIndex = index\">\n      <input type=\"radio\" [id]=\"ratingQuestion.attributes.steps.length - stepIndex+'rating'\"\n             [value]=\"step.value\"\n             [(ngModel)]=\"surveyResponse.rating\"\n             (click)=\"onRatingSelect(step.value); $event.stopPropagation()\"\n             name=\"star-rating\" >\n      <label [for] = \"ratingQuestion.attributes.steps.length - stepIndex+'rating'\">\n        <div *ngIf=\"ratingQuestion.attributes.ratingIcon == 'star'\">\n          <svg class=\"ratingOutlineIcon\" xmlns=\"http://www.w3.org/2000/svg\" width=\"30\" height=\"30\" viewBox=\"0 0 24 24\"><path d=\"M15.668 8.626l8.332 1.159-6.065 5.874 1.48 8.341-7.416-3.997-7.416 3.997 1.481-8.341-6.064-5.874 8.331-1.159 3.668-7.626 3.669 7.626zm-6.67.925l-6.818.948 4.963 4.807-1.212 6.825 6.068-3.271 6.069 3.271-1.212-6.826 4.964-4.806-6.819-.948-3.002-6.241-3.001 6.241z\"/></svg>\n          <svg class=\"ratingFillIcon\" xmlns=\"http://www.w3.org/2000/svg\" width=\"30\" height=\"30\" viewBox=\"0 0 24 24\"><path d=\"M12 .587l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z\"/></svg>\n        </div>\n        <div *ngIf=\"ratingQuestion.attributes.ratingIcon == 'heart'\">\n          <svg class=\"ratingOutlineIcon\" width=\"30\" height=\"30\" xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 24 24\"><path d=\"M12 21.593c-5.63-5.539-11-10.297-11-14.402 0-3.791 3.068-5.191 5.281-5.191 1.312 0 4.151.501 5.719 4.457 1.59-3.968 4.464-4.447 5.726-4.447 2.54 0 5.274 1.621 5.274 5.181 0 4.069-5.136 8.625-11 14.402m5.726-20.583c-2.203 0-4.446 1.042-5.726 3.238-1.285-2.206-3.522-3.248-5.719-3.248-3.183 0-6.281 2.187-6.281 6.191 0 4.661 5.571 9.429 12 15.809 6.43-6.38 12-11.148 12-15.809 0-4.011-3.095-6.181-6.274-6.181\"/></svg>\n          <svg class=\"ratingFillIcon\" width=\"30\" height=\"30\" xmlns=\"http://www.w3.org/2000/svg\"  viewBox=\"0 0 24 24\"><path d=\"M12 4.248c-3.148-5.402-12-3.825-12 2.944 0 4.661 5.571 9.427 12 15.808 6.43-6.381 12-11.147 12-15.808 0-6.792-8.875-8.306-12-2.944z\"/></svg>\n        </div>\n        <div *ngIf=\"ratingQuestion.attributes.ratingIcon == 'thumbsUp'\">\n          <svg class=\"ratingOutlineIcon\" xmlns=\"http://www.w3.org/2000/svg\" width=\"30\" height=\"30\" viewBox=\"0 0 24 24\"><path d=\"M21.406 9.558c-1.21-.051-2.87-.278-3.977-.744.809-3.283 1.253-8.814-2.196-8.814-1.861 0-2.351 1.668-2.833 3.329-1.548 5.336-3.946 6.816-6.4 7.401v-.73h-6v12h6v-.904c2.378.228 4.119.864 6.169 1.746 1.257.541 3.053 1.158 5.336 1.158 2.538 0 4.295-.997 5.009-3.686.5-1.877 1.486-7.25 1.486-8.25 0-1.648-1.168-2.446-2.594-2.506zm-17.406 10.442h-2v-8h2v8zm15.896-5.583s.201.01 1.069-.027c1.082-.046 1.051 1.469.004 1.563l-1.761.099c-.734.094-.656 1.203.141 1.172 0 0 .686-.017 1.143-.041 1.068-.056 1.016 1.429.04 1.551-.424.053-1.745.115-1.745.115-.811.072-.706 1.235.109 1.141l.771-.031c.822-.074 1.003.825-.292 1.661-1.567.881-4.685.131-6.416-.614-2.239-.965-4.438-1.934-6.959-2.006v-6c3.264-.749 6.328-2.254 8.321-9.113.898-3.092 1.679-1.931 1.679.574 0 2.071-.49 3.786-.921 5.533 1.061.543 3.371 1.402 6.12 1.556 1.055.059 1.024 1.455-.051 1.584l-1.394.167s-.608 1.111.142 1.116z\"/></svg>\n          <svg class=\"ratingFillIcon\" xmlns=\"http://www.w3.org/2000/svg\" width=\"30\" height=\"30\" viewBox=\"0 0 24 24\"><path d=\"M5 22h-5v-12h5v12zm17.615-8.412c-.857-.115-.578-.734.031-.922.521-.16 1.354-.5 1.354-1.51 0-.672-.5-1.562-2.271-1.49-1.228.05-3.666-.198-4.979-.885.906-3.656.688-8.781-1.688-8.781-1.594 0-1.896 1.807-2.375 3.469-1.221 4.242-3.312 6.017-5.687 6.885v10.878c4.382.701 6.345 2.768 10.505 2.768 3.198 0 4.852-1.735 4.852-2.666 0-.335-.272-.573-.96-.626-.811-.062-.734-.812.031-.953 1.268-.234 1.826-.914 1.826-1.543 0-.529-.396-1.022-1.098-1.181-.837-.189-.664-.757.031-.812 1.133-.09 1.688-.764 1.688-1.41 0-.565-.424-1.109-1.26-1.221z\"/></svg>\n        </div>\n      </label>\n    </ng-container>\n</div>\n\n  <div style=\"margin-top:2%\" class=\"skip-this-question\" *ngIf=\"!ratingQuestion.attributes.required\">\n  <span (click)=\"onRatingSelect(undefined)\">Skip</span>\n  </div>\n\n</div>\n\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    RatingComponent.ctorParameters = function () { return [
        { type: PagesService }
    ]; };
    RatingComponent.propDecorators = {
        ratingQuestion: [{ type: Input }],
        surveyResponse: [{ type: Input }],
        submitEvent: [{ type: Output }]
    };
    return RatingComponent;
}());
export { RatingComponent };
if (false) {
    /** @type {?} */
    RatingComponent.prototype.ratingQuestion;
    /** @type {?} */
    RatingComponent.prototype.surveyResponse;
    /** @type {?} */
    RatingComponent.prototype.submitEvent;
    /** @type {?} */
    RatingComponent.prototype.isActive;
    /**
     * @type {?}
     * @private
     */
    RatingComponent.prototype.pageService;
}
