// survey-lib/esm2015/lib/pages/store/survey-actions.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/store/survey-actions.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const CURRENT_STATE = 'current_state';
/** @type {?} */
export const SAVE_CHOICE = 'save_choice';
/** @type {?} */
export const SAVE_EMAIL = 'save_email';
/** @type {?} */
export const SAVE_AGREE_DISAGREE = 'save_agree_disagree';
/** @type {?} */
export const SAVE_PHONE = 'save_phone';
/** @type {?} */
export const SAVE_ADDRESS = 'save_address';
/** @type {?} */
export const SAVE_DATE = 'save_date';
/** @type {?} */
export const SAVE_NPS = 'save_nps';
/** @type {?} */
export const SAVE_YES_NO = 'save_yes_no';
/** @type {?} */
export const SAVE_RATING = 'save_rating';
/** @type {?} */
export const SAVE_DROP_DOWN = 'save_rating';
/** @type {?} */
export const SAVE_LONG_TEXT = 'save_long_text';
/** @type {?} */
export const SAVE_SHORT_TEXT = 'save_short_text';
/** @type {?} */
export const RESET_RESPONSE = 'reset_response';
/** @type {?} */
export const REMOVE_RESPONSES = 'remove_responses';
/** @type {?} */
export const REMOVE_RESPONSES_OF_EXCLUDED_QUESTION_IDS = 'remove_responses_of_excluded_question_ids';
export class Email {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = SAVE_EMAIL;
    }
}
if (false) {
    /** @type {?} */
    Email.prototype.type;
    /** @type {?} */
    Email.prototype.payload;
}
export class Phone {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = SAVE_PHONE;
    }
}
if (false) {
    /** @type {?} */
    Phone.prototype.type;
    /** @type {?} */
    Phone.prototype.payload;
}
export class AgreeDisagree {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = SAVE_AGREE_DISAGREE;
    }
}
if (false) {
    /** @type {?} */
    AgreeDisagree.prototype.type;
    /** @type {?} */
    AgreeDisagree.prototype.payload;
}
export class Choice {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = SAVE_CHOICE;
    }
}
if (false) {
    /** @type {?} */
    Choice.prototype.type;
    /** @type {?} */
    Choice.prototype.payload;
}
export class Address {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = SAVE_ADDRESS;
    }
}
if (false) {
    /** @type {?} */
    Address.prototype.type;
    /** @type {?} */
    Address.prototype.payload;
}
export class Date {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = SAVE_DATE;
    }
}
if (false) {
    /** @type {?} */
    Date.prototype.type;
    /** @type {?} */
    Date.prototype.payload;
}
export class Nps {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = SAVE_NPS;
    }
}
if (false) {
    /** @type {?} */
    Nps.prototype.type;
    /** @type {?} */
    Nps.prototype.payload;
}
export class YesNo {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = SAVE_YES_NO;
    }
}
if (false) {
    /** @type {?} */
    YesNo.prototype.type;
    /** @type {?} */
    YesNo.prototype.payload;
}
export class Rating {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = SAVE_RATING;
    }
}
if (false) {
    /** @type {?} */
    Rating.prototype.type;
    /** @type {?} */
    Rating.prototype.payload;
}
export class DropDown {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = SAVE_DROP_DOWN;
    }
}
if (false) {
    /** @type {?} */
    DropDown.prototype.type;
    /** @type {?} */
    DropDown.prototype.payload;
}
export class LongText {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = SAVE_LONG_TEXT;
    }
}
if (false) {
    /** @type {?} */
    LongText.prototype.type;
    /** @type {?} */
    LongText.prototype.payload;
}
export class ShortText {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = SAVE_SHORT_TEXT;
    }
}
if (false) {
    /** @type {?} */
    ShortText.prototype.type;
    /** @type {?} */
    ShortText.prototype.payload;
}
export class RemoveResponses {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = REMOVE_RESPONSES;
    }
}
if (false) {
    /** @type {?} */
    RemoveResponses.prototype.type;
    /** @type {?} */
    RemoveResponses.prototype.payload;
}
export class RemoveResponsesOfExcludedQuestionIds {
    /**
     * @param {?} payload
     */
    constructor(payload) {
        this.payload = payload;
        this.type = REMOVE_RESPONSES_OF_EXCLUDED_QUESTION_IDS;
    }
}
if (false) {
    /** @type {?} */
    RemoveResponsesOfExcludedQuestionIds.prototype.type;
    /** @type {?} */
    RemoveResponsesOfExcludedQuestionIds.prototype.payload;
}
export class Reset {
    constructor() {
        this.type = RESET_RESPONSE;
    }
}
if (false) {
    /** @type {?} */
    Reset.prototype.type;
}
export class CurrentState {
    constructor() {
        this.type = CURRENT_STATE;
    }
}
if (false) {
    /** @type {?} */
    CurrentState.prototype.type;
}
