// survey-lib/esm2015/lib/pages/questionnaire/nps/nps.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/questionnaire/nps/nps.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Question } from '../../pages.model';
import { SurveyResponses } from '../../survey-response.model';
import { PagesService } from '../../../shared/pages-services/pages.service';
export class NpsComponent {
    /**
     * @param {?} pageService
     */
    constructor(pageService) {
        this.pageService = pageService;
        this.submitEvent = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.pageService.stage == 'entry-next') {
            document.getElementById('nps-question').className = 'entry-next';
            document.getElementById('nps-question-delay').className = 'entry-next-delay';
        }
        else if (this.pageService.stage == 'entry-previous') {
            document.getElementById('nps-question').className = 'entry-previous-delay';
            document.getElementById('nps-question-delay').className = 'entry-previous';
        }
    }
    /**
     * @param {?} rating
     * @return {?}
     */
    onSelect(rating) {
        if (rating != undefined) {
            this.surveyResponse.question = this.npsQuestion.question;
            this.surveyResponse.questionType = this.npsQuestion.questionType;
            this.surveyResponse.identifier = this.npsQuestion.identifier;
            this.surveyResponse.questionId = this.npsQuestion.questionId;
            this.surveyResponse.rating = rating;
            this.pageService.persistSurvey(this.surveyResponse);
            document.getElementById('rating-' + rating).classList.add('blink-btn-animation');
            setTimeout((/**
             * @return {?}
             */
            () => {
                this.submitEvent.emit();
            }), 350);
        }
        else {
            this.submitEvent.emit();
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        /*this.surveyState*/
    }
}
NpsComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-nps',
                template: "<div id=\"nps-question\">\n\n<div class=\"survey-statement\"><u></u>\n  <span *ngIf=\"npsQuestion.attributes.required\" class=\"isRequired\">*</span>{{ npsQuestion.question }} </div>\n  <p class=\"question-description\" *ngIf=\"npsQuestion.attributes.isDescriptionEnabled\">\n  {{npsQuestion.attributes.description}}\n</p>\n</div>\n\n<div id=\"nps-question-delay\">\n<div class=\"nps-btn-group\">\n\n    <button id=\"rating-0\" class=\"btn-secondary nps-btn\" [ngClass]=\"{'nps-btn-focus':surveyResponse.rating==0 }\" (click)=\"onSelect(0)\">0</button>\n\n    <button id=\"rating-1\" class=\"btn-secondary nps-btn\" [ngClass]=\"{'nps-btn-focus':surveyResponse.rating==1 }\" (click)=\"onSelect(1)\">1</button>\n\n    <button id=\"rating-2\" class=\"btn-secondary nps-btn\" [ngClass]=\"{'nps-btn-focus':surveyResponse.rating==2 }\" (click)=\"onSelect(2)\">2</button>\n\n    <button id=\"rating-3\" class=\"btn-secondary nps-btn\" [ngClass]=\"{'nps-btn-focus':surveyResponse.rating==3 }\" (click)=\"onSelect(3)\">3</button>\n\n    <button id=\"rating-4\" class=\"btn-secondary nps-btn\" [ngClass]=\"{'nps-btn-focus':surveyResponse.rating==4 }\" (click)=\"onSelect(4)\">4</button>\n\n    <button id=\"rating-5\" class=\"btn-secondary nps-btn\" [ngClass]=\"{'nps-btn-focus':surveyResponse.rating==5 }\" (click)=\"onSelect(5)\">5</button>\n\n    <button id=\"rating-6\" class=\"btn-secondary nps-btn\" [ngClass]=\"{'nps-btn-focus':surveyResponse.rating==6 }\" (click)=\"onSelect(6)\">6</button>\n\n    <button id=\"rating-7\" class=\"btn-secondary nps-btn\" [ngClass]=\"{'nps-btn-focus':surveyResponse.rating==7 }\" (click)=\"onSelect(7)\">7</button>\n\n    <button id=\"rating-8\" class=\"btn-secondary nps-btn\" [ngClass]=\"{'nps-btn-focus':surveyResponse.rating==8 }\" (click)=\"onSelect(8)\">8</button>\n\n    <button id=\"rating-9\" class=\"btn-secondary nps-btn\" [ngClass]=\"{'nps-btn-focus':surveyResponse.rating==9 }\" (click)=\"onSelect(9)\">9</button>\n\n    <button id=\"rating-10\" class=\"btn-secondary nps-btn\" [ngClass]=\"{'nps-btn-focus':surveyResponse.rating==10 }\" (click)=\"onSelect(10)\">10</button>\n\n</div>\n<div class=\"nps-legend\">\n  <span id=\"left-nps-label\">{{npsQuestion.attributes.leftLabel}}</span>\n  <span id=\"center-nps-label\">{{npsQuestion.attributes.centerLabel}}</span>\n  <span id=\"right-nps-label\">{{npsQuestion.attributes.rightLabel}}</span>\n</div>\n\n<div  class=\"skip-this-question\" style=\"margin-top:8%;\" *ngIf=\"!npsQuestion.attributes.required\">\n  <span (click)=\"onSelect(undefined)\" >Skip</span>\n</div>\n\n</div>\n",
                styles: [""]
            }] }
];
/** @nocollapse */
NpsComponent.ctorParameters = () => [
    { type: PagesService }
];
NpsComponent.propDecorators = {
    npsQuestion: [{ type: Input }],
    surveyResponse: [{ type: Input }],
    submitEvent: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    NpsComponent.prototype.npsQuestion;
    /** @type {?} */
    NpsComponent.prototype.surveyResponse;
    /** @type {?} */
    NpsComponent.prototype.submitEvent;
    /**
     * @type {?}
     * @private
     */
    NpsComponent.prototype.pageService;
}
