// ngrx-forms/esm5/src/status-css-classes.directive.js
import * as tslib_1 from "tslib";
import { Directive, HostBinding, Input } from '@angular/core';
/**
 * Lists the available status class names based on the property
 * they are depending on.
 */
export var NGRX_STATUS_CLASS_NAMES = {
    isValid: 'ngrx-forms-valid',
    isInvalid: 'ngrx-forms-invalid',
    isDirty: 'ngrx-forms-dirty',
    isPristine: 'ngrx-forms-pristine',
    isTouched: 'ngrx-forms-touched',
    isUntouched: 'ngrx-forms-untouched',
    isSubmitted: 'ngrx-forms-submitted',
    isUnsubmitted: 'ngrx-forms-unsubmitted',
    isValidationPending: 'ngrx-forms-validation-pending',
};
var NgrxStatusCssClassesDirective = /** @class */ (function () {
    function NgrxStatusCssClassesDirective() {
    }
    Object.defineProperty(NgrxStatusCssClassesDirective.prototype, "ngrxFormControlState", {
        set: function (state) {
            this.state = state;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgrxStatusCssClassesDirective.prototype, "ngrxFormState", {
        set: function (state) {
            this.state = state;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgrxStatusCssClassesDirective.prototype, "isValid", {
        get: function () {
            return this.state.isValid;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgrxStatusCssClassesDirective.prototype, "isInvalid", {
        get: function () {
            return this.state.isInvalid;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgrxStatusCssClassesDirective.prototype, "isDirty", {
        get: function () {
            return this.state.isDirty;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgrxStatusCssClassesDirective.prototype, "isPristine", {
        get: function () {
            return this.state.isPristine;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgrxStatusCssClassesDirective.prototype, "isTouched", {
        get: function () {
            return this.state.isTouched;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgrxStatusCssClassesDirective.prototype, "isUntouched", {
        get: function () {
            return this.state.isUntouched;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgrxStatusCssClassesDirective.prototype, "isSubmitted", {
        get: function () {
            return this.state.isSubmitted;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgrxStatusCssClassesDirective.prototype, "isUnsubmitted", {
        get: function () {
            return this.state.isUnsubmitted;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgrxStatusCssClassesDirective.prototype, "isValidationPending", {
        get: function () {
            return this.state.isValidationPending;
        },
        enumerable: true,
        configurable: true
    });
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], NgrxStatusCssClassesDirective.prototype, "ngrxFormControlState", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], NgrxStatusCssClassesDirective.prototype, "ngrxFormState", null);
    tslib_1.__decorate([
        HostBinding("class." + NGRX_STATUS_CLASS_NAMES.isValid),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [])
    ], NgrxStatusCssClassesDirective.prototype, "isValid", null);
    tslib_1.__decorate([
        HostBinding("class." + NGRX_STATUS_CLASS_NAMES.isInvalid),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [])
    ], NgrxStatusCssClassesDirective.prototype, "isInvalid", null);
    tslib_1.__decorate([
        HostBinding("class." + NGRX_STATUS_CLASS_NAMES.isDirty),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [])
    ], NgrxStatusCssClassesDirective.prototype, "isDirty", null);
    tslib_1.__decorate([
        HostBinding("class." + NGRX_STATUS_CLASS_NAMES.isPristine),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [])
    ], NgrxStatusCssClassesDirective.prototype, "isPristine", null);
    tslib_1.__decorate([
        HostBinding("class." + NGRX_STATUS_CLASS_NAMES.isTouched),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [])
    ], NgrxStatusCssClassesDirective.prototype, "isTouched", null);
    tslib_1.__decorate([
        HostBinding("class." + NGRX_STATUS_CLASS_NAMES.isUntouched),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [])
    ], NgrxStatusCssClassesDirective.prototype, "isUntouched", null);
    tslib_1.__decorate([
        HostBinding("class." + NGRX_STATUS_CLASS_NAMES.isSubmitted),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [])
    ], NgrxStatusCssClassesDirective.prototype, "isSubmitted", null);
    tslib_1.__decorate([
        HostBinding("class." + NGRX_STATUS_CLASS_NAMES.isUnsubmitted),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [])
    ], NgrxStatusCssClassesDirective.prototype, "isUnsubmitted", null);
    tslib_1.__decorate([
        HostBinding("class." + NGRX_STATUS_CLASS_NAMES.isValidationPending),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [])
    ], NgrxStatusCssClassesDirective.prototype, "isValidationPending", null);
    NgrxStatusCssClassesDirective = tslib_1.__decorate([
        Directive({
            // tslint:disable-next-line:directive-selector
            selector: 'form[ngrxFormState],[ngrxFormControlState]',
        })
    ], NgrxStatusCssClassesDirective);
    return NgrxStatusCssClassesDirective;
}());
export { NgrxStatusCssClassesDirective };
