// survey-lib/esm2015/lib/pages/questionnaire/contact-form/contact-form.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/questionnaire/contact-form/contact-form.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { Question } from '../../pages.model';
import { FormControl, FormGroup, Validators } from '@angular/forms';
import { PagesService } from '../../../shared/pages-services/pages.service';
export class ContactFormComponent {
    /**
     * @param {?} pagesService
     */
    constructor(pagesService) {
        this.pagesService = pagesService;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.createForm();
    }
    /**
     * @return {?}
     */
    createForm() {
        /** @type {?} */
        const formGroup = {};
        for (let prop of Object.keys(this.contactForm.contactForm)) {
            formGroup[prop] = new FormControl(this.contactForm.contactForm[prop].value || '', ContactFormComponent.mapValidators(this.contactForm.contactForm[prop].validation));
        }
        this.form = new FormGroup(formGroup);
    }
    /**
     * @private
     * @param {?} validators
     * @return {?}
     */
    static mapValidators(validators) {
        /** @type {?} */
        const formValidators = [];
        if (validators) {
            for (const validation of Object.keys(validators)) {
                if (validation === 'required') {
                    formValidators.push(Validators.required);
                }
                else if (validation === 'min') {
                    formValidators.push(Validators.min(validators[validation]));
                }
            }
        }
        return formValidators;
    }
    /**
     * @return {?}
     */
    onSubmit() {
        this.response.contactInfo = this.objectProps;
        this.response.questionId = this.contactForm.questionId;
        this.pagesService.persistSurvey(this.response);
        //console.log(this.dataObject);
        /*  for (let controlsKey in this.form.controls) {
            this.dataObject[controlsKey].value = form[controlsKey];
          }
          console.log(this.dataObject);
          this.response.questionId = this.contactForm.questionId;*/
        //this.response.contactInfo
    }
}
ContactFormComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-contact-form',
                template: "\n<div>\n\n  <form novalidate (ngSubmit)=\"onSubmit()\" [formGroup]=\"form\">\n    <div *ngFor=\"let prop of objectProps\">\n      <label [attr.for]=\"prop\">{{prop.label}}</label>\n\n      <div [ngSwitch]=\"prop.type\">\n        <input *ngSwitchCase=\"'text'\"\n               [formControlName]=\"prop.key\"\n               [id]=\"prop.key\" [type]=\"prop.type\"\n               [(ngModel)] = \"prop.value\">\n\n      </div>\n<!--\n      <div class=\"error\" *ngIf=\"form.get(prop.key).invalid && (form.get(prop.key).dirty || form.get(prop.key).touched)\">\n        <div *ngIf=\"form.get(prop.key).errors.required\">\n          <h6>{{ prop.label }} is required.</h6>\n        </div>\n      </div>-->\n\n\n    </div>\n    <p>\n      <button type=\"submit\">Done</button>\n    </p>\n  </form>\n\n</div>\n",
                styles: [""]
            }] }
];
/** @nocollapse */
ContactFormComponent.ctorParameters = () => [
    { type: PagesService }
];
ContactFormComponent.propDecorators = {
    contactForm: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    ContactFormComponent.prototype.contactForm;
    /** @type {?} */
    ContactFormComponent.prototype.form;
    /** @type {?} */
    ContactFormComponent.prototype.objectProps;
    /**
     * @type {?}
     * @private
     */
    ContactFormComponent.prototype.response;
    /**
     * @type {?}
     * @private
     */
    ContactFormComponent.prototype.surveyState;
    /**
     * @type {?}
     * @private
     */
    ContactFormComponent.prototype.pagesService;
}
