// ngrx-forms/esm5/src/control/directive.js
import * as tslib_1 from "tslib";
import { DOCUMENT } from '@angular/common';
import { Directive, ElementRef, HostBinding, HostListener, Inject, Input, Optional, Self, } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { ActionsSubject } from '@ngrx/store';
import { FocusAction, MarkAsDirtyAction, MarkAsTouchedAction, SetValueAction, UnfocusAction } from '../actions';
import { selectViewAdapter } from '../view-adapter/util';
import { NGRX_FORM_VIEW_ADAPTER } from '../view-adapter/view-adapter';
import { NgrxValueConverters } from './value-converter';
export var NGRX_UPDATE_ON_TYPE;
(function (NGRX_UPDATE_ON_TYPE) {
    NGRX_UPDATE_ON_TYPE["CHANGE"] = "change";
    NGRX_UPDATE_ON_TYPE["BLUR"] = "blur";
    NGRX_UPDATE_ON_TYPE["NEVER"] = "never";
})(NGRX_UPDATE_ON_TYPE || (NGRX_UPDATE_ON_TYPE = {}));
var ControlValueAccessorAdapter = /** @class */ (function () {
    function ControlValueAccessorAdapter(valueAccessor) {
        this.valueAccessor = valueAccessor;
    }
    ControlValueAccessorAdapter.prototype.setViewValue = function (value) {
        this.valueAccessor.writeValue(value);
    };
    ControlValueAccessorAdapter.prototype.setOnChangeCallback = function (fn) {
        this.valueAccessor.registerOnChange(fn);
    };
    ControlValueAccessorAdapter.prototype.setOnTouchedCallback = function (fn) {
        this.valueAccessor.registerOnTouched(fn);
    };
    ControlValueAccessorAdapter.prototype.setIsDisabled = function (isDisabled) {
        if (this.valueAccessor.setDisabledState) {
            this.valueAccessor.setDisabledState(isDisabled);
        }
    };
    return ControlValueAccessorAdapter;
}());
var NgrxFormControlDirective = /** @class */ (function () {
    function NgrxFormControlDirective(el, 
    // for the dom parameter the `null` type must be last to ensure that in the compiled output
    // there is no reference to the Document type to support non-browser platforms
    dom, actionsSubject, viewAdapters, valueAccessors) {
        this.el = el;
        this.dom = dom;
        this.actionsSubject = actionsSubject;
        this.isInitialized = false;
        this.focusTrackingIsEnabled = false;
        this.ngrxUpdateOn = NGRX_UPDATE_ON_TYPE.CHANGE;
        this.ngrxValueConverter = NgrxValueConverters.default();
        viewAdapters = viewAdapters || [];
        valueAccessors = valueAccessors || [];
        if (valueAccessors.length > 1) {
            throw new Error('More than one custom control value accessor matches!');
        }
        this.viewAdapter = valueAccessors.length > 0
            ? new ControlValueAccessorAdapter(valueAccessors[0])
            : selectViewAdapter(viewAdapters);
    }
    Object.defineProperty(NgrxFormControlDirective.prototype, "ngrxFormControlState", {
        set: function (newState) {
            if (!newState) {
                throw new Error('The control state must not be undefined!');
            }
            var oldState = this.state;
            this.state = newState;
            if (this.isInitialized) {
                this.updateViewIfControlIdChanged(newState, oldState);
                this.updateViewIfValueChanged(newState, oldState);
                this.updateViewIfIsDisabledChanged(newState, oldState);
                this.updateViewIfIsFocusedChanged(newState, oldState);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgrxFormControlDirective.prototype, "ngrxEnableFocusTracking", {
        set: function (value) {
            if (value && !this.dom) {
                throw new Error('focus tracking is only supported on the browser platform');
            }
            this.focusTrackingIsEnabled = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgrxFormControlDirective.prototype, "focusRegionStartAttr", {
        // TODO: move this into a separate directive
        // automatically apply the attribute that's used by the CDK to set initial focus
        get: function () {
            return this.state && this.state.isFocused ? '' : null;
        },
        enumerable: true,
        configurable: true
    });
    NgrxFormControlDirective.prototype.updateViewIfControlIdChanged = function (newState, oldState) {
        if (oldState && newState.id === oldState.id) {
            return;
        }
        this.stateValue = newState.value;
        this.viewValue = this.ngrxValueConverter.convertStateToViewValue(this.stateValue);
        this.viewAdapter.setViewValue(this.viewValue);
        if (this.viewAdapter.setIsDisabled) {
            this.viewAdapter.setIsDisabled(newState.isDisabled);
        }
    };
    NgrxFormControlDirective.prototype.updateViewIfValueChanged = function (newState, _) {
        if (newState.value === this.stateValue) {
            return;
        }
        this.stateValue = newState.value;
        this.viewValue = this.ngrxValueConverter.convertStateToViewValue(newState.value);
        this.viewAdapter.setViewValue(this.viewValue);
    };
    NgrxFormControlDirective.prototype.updateViewIfIsDisabledChanged = function (newState, oldState) {
        if (!this.viewAdapter.setIsDisabled) {
            return;
        }
        if (oldState && newState.isDisabled === oldState.isDisabled) {
            return;
        }
        this.viewAdapter.setIsDisabled(newState.isDisabled);
    };
    NgrxFormControlDirective.prototype.updateViewIfIsFocusedChanged = function (newState, oldState) {
        if (!this.focusTrackingIsEnabled) {
            return;
        }
        if (oldState && newState.isFocused === oldState.isFocused) {
            return;
        }
        if (newState.isFocused) {
            this.el.nativeElement.focus();
        }
        else {
            this.el.nativeElement.blur();
        }
    };
    NgrxFormControlDirective.prototype.dispatchAction = function (action) {
        if (this.actionsSubject !== null) {
            this.actionsSubject.next(action);
        }
        else {
            throw new Error('ActionsSubject must be present in order to dispatch actions!');
        }
    };
    NgrxFormControlDirective.prototype.ngOnInit = function () {
        var _this = this;
        if (!this.state) {
            throw new Error('The form state must not be undefined!');
        }
        this.isInitialized = true;
        this.updateViewIfControlIdChanged(this.state, undefined);
        this.updateViewIfValueChanged(this.state, undefined);
        this.updateViewIfIsDisabledChanged(this.state, undefined);
        this.updateViewIfIsFocusedChanged(this.state, undefined);
        var dispatchMarkAsDirtyAction = function () {
            if (_this.state.isPristine) {
                _this.dispatchAction(new MarkAsDirtyAction(_this.state.id));
            }
        };
        var dispatchSetValueAction = function () {
            _this.stateValue = _this.ngrxValueConverter.convertViewToStateValue(_this.viewValue);
            if (_this.stateValue !== _this.state.value) {
                _this.dispatchAction(new SetValueAction(_this.state.id, _this.stateValue));
                dispatchMarkAsDirtyAction();
            }
        };
        this.viewAdapter.setOnChangeCallback(function (viewValue) {
            _this.viewValue = viewValue;
            if (_this.ngrxUpdateOn === NGRX_UPDATE_ON_TYPE.CHANGE) {
                dispatchSetValueAction();
            }
        });
        this.viewAdapter.setOnTouchedCallback(function () {
            if (!_this.state.isTouched && _this.ngrxUpdateOn !== NGRX_UPDATE_ON_TYPE.NEVER) {
                _this.dispatchAction(new MarkAsTouchedAction(_this.state.id));
            }
            if (_this.ngrxUpdateOn === NGRX_UPDATE_ON_TYPE.BLUR) {
                dispatchSetValueAction();
            }
        });
    };
    NgrxFormControlDirective.prototype.ngAfterViewInit = function () {
        // we need to update the view again after it was initialized since some
        // controls depend on child elements for setting the value (e.g. selects)
        this.viewAdapter.setViewValue(this.viewValue);
        if (this.viewAdapter.setIsDisabled) {
            this.viewAdapter.setIsDisabled(this.state.isDisabled);
        }
    };
    NgrxFormControlDirective.prototype.onFocusChange = function () {
        if (!this.focusTrackingIsEnabled) {
            return;
        }
        var isControlFocused = this.el.nativeElement === this.dom.activeElement;
        if (isControlFocused !== this.state.isFocused) {
            this.dispatchAction(isControlFocused ? new FocusAction(this.state.id) : new UnfocusAction(this.state.id));
        }
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], NgrxFormControlDirective.prototype, "ngrxFormControlState", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String)
    ], NgrxFormControlDirective.prototype, "ngrxUpdateOn", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], NgrxFormControlDirective.prototype, "ngrxEnableFocusTracking", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], NgrxFormControlDirective.prototype, "ngrxValueConverter", void 0);
    tslib_1.__decorate([
        HostBinding('attr.cdk-focus-region-start'),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [])
    ], NgrxFormControlDirective.prototype, "focusRegionStartAttr", null);
    tslib_1.__decorate([
        HostListener('focusin'),
        HostListener('focusout'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], NgrxFormControlDirective.prototype, "onFocusChange", null);
    NgrxFormControlDirective = tslib_1.__decorate([
        Directive({
            // tslint:disable-next-line:directive-selector
            selector: ':not([ngrxFormsAction])[ngrxFormControlState]',
        }),
        tslib_1.__param(1, Optional()), tslib_1.__param(1, Inject(DOCUMENT)),
        tslib_1.__param(2, Optional()), tslib_1.__param(2, Inject(ActionsSubject)),
        tslib_1.__param(3, Self()), tslib_1.__param(3, Optional()), tslib_1.__param(3, Inject(NGRX_FORM_VIEW_ADAPTER)),
        tslib_1.__param(4, Self()), tslib_1.__param(4, Optional()), tslib_1.__param(4, Inject(NG_VALUE_ACCESSOR)),
        tslib_1.__metadata("design:paramtypes", [ElementRef, Object, Object, Array, Array])
    ], NgrxFormControlDirective);
    return NgrxFormControlDirective;
}());
export { NgrxFormControlDirective };
