// survey-lib/esm2015/lib/pages/questionnaire/website/website.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/questionnaire/website/website.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Question } from '../../pages.model';
import { SurveyResponses } from '../../survey-response.model';
import { PagesService } from '../../../shared/pages-services/pages.service';
export class WebsiteComponent {
    /**
     * @param {?} pageService
     */
    constructor(pageService) {
        this.pageService = pageService;
        this.URL_REGEXP = new RegExp('^[A-Za-z][A-Za-z\\d.+-]*:\\/*(?:\\w+(?::\\w+)?@)?[^\\s/]+(?::\\d+)?(?:\\/[\\w#!:.?+=&%@\\-/]*)?$');
        this.submitEvent = new EventEmitter();
        this.isValidUrl = true;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.pageService.stage == 'entry-next') {
            document.getElementById('website-question').className = 'entry-next';
            document.getElementById('website-question-delay').className = 'entry-next-delay';
        }
        else if (this.pageService.stage == 'entry-previous') {
            document.getElementById('website-question').className = 'entry-previous-delay';
            document.getElementById('website-question-delay').className = 'entry-previous';
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    onSubmit(value) {
        console.log(value);
        if (value != undefined) {
            this.isValidUrl = this.test(value);
            if (this.isValidUrl) {
                this.surveyResponse.questionId = this.website.questionId;
                this.surveyResponse.question = this.website.question;
                this.surveyResponse.questionType = this.website.questionType;
                this.surveyResponse.identifier = this.website.identifier;
                this.surveyResponse.answer = value;
                this.pageService.persistSurvey(this.surveyResponse);
                document.getElementById('btn-primary').classList.add('blink-btn-animation');
                setTimeout((/**
                 * @return {?}
                 */
                () => {
                    this.submitEvent.emit();
                }), 350);
            }
        }
        else {
            this.submitEvent.emit();
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
    }
    /**
     * @private
     * @param {?} url
     * @return {?}
     */
    test(url) {
        return this.URL_REGEXP.test(url);
    }
}
WebsiteComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-website',
                template: "<div id=\"website-question\">\n<div class=\"survey-statement\">\n    <span *ngIf=\"website.attributes.required\" class=\"isRequired\">*</span>{{website.question}}\n</div>\n<p class=\"question-description\">\n    {{website.attributes.description}}\n  </p>\n</div>\n\n<div id=\"website-question-delay\">\n<input class=\"survey-input-text-box\" #websiteInput\n       type=\"url\"\n       name=\"name\"\n       [(ngModel)]=\"surveyResponse.answer\"\n       placeholder=\"Enter website url\">\n  <div *ngIf=\"!isValidUrl\" class=\"validation-error-msg\">Please enter valid URL</div>\n<button class=\"btn-primary\" id=\"btn-primary\" type=\"submit\" name=\"name\" (click)=\"onSubmit(websiteInput.value)\">{{isLastQuestion?'Submit': 'Done'}}</button>\n\n<span\n  (click)=\"onSubmit(undefined)\"\n  *ngIf=\"!website.attributes.required\" class=\"skip-this-question skip-this-question-marginLeft\">Skip</span>\n</div>\n",
                styles: [""]
            }] }
];
/** @nocollapse */
WebsiteComponent.ctorParameters = () => [
    { type: PagesService }
];
WebsiteComponent.propDecorators = {
    website: [{ type: Input }],
    surveyResponse: [{ type: Input }],
    submitEvent: [{ type: Output }],
    isLastQuestion: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    WebsiteComponent.prototype.URL_REGEXP;
    /** @type {?} */
    WebsiteComponent.prototype.website;
    /** @type {?} */
    WebsiteComponent.prototype.surveyResponse;
    /** @type {?} */
    WebsiteComponent.prototype.submitEvent;
    /** @type {?} */
    WebsiteComponent.prototype.isValidUrl;
    /** @type {?} */
    WebsiteComponent.prototype.isLastQuestion;
    /**
     * @type {?}
     * @private
     */
    WebsiteComponent.prototype.pageService;
}
