// survey-lib/esm5/lib/survey-lib.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/survey-lib.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var SurveyLibService = /** @class */ (function () {
    function SurveyLibService() {
        this.enableSubmit = false;
        this.defaultThankYouPage = false;
        this.isEmbededInDesigner = false;
        this.questionAnsweredEventEmitter = new EventEmitter();
        this.isQuestionCardCampaign = false;
        this.currentElementEventEmitter = new EventEmitter();
        this.currentElement = new CurrentElement();
        this.currentElement.currentPageIndex = 0;
        this.currentElement.currentQuestionIndex = 0;
    }
    /**
     * @param {?} selectedThemeModel
     * @return {?}
     */
    SurveyLibService.prototype.initializeSurveyTheme = /**
     * @param {?} selectedThemeModel
     * @return {?}
     */
    function (selectedThemeModel) {
        selectedThemeModel.font = {
            name: 'Muli',
            url: 'https://fonts.googleapis.com/css?family=Muli&display=swap'
        };
        console.log(selectedThemeModel);
        /** @type {?} */
        var link = document.createElement('link');
        link.type = "text/css";
        link.rel = "stylesheet";
        link.href = selectedThemeModel.font.url;
        document.head.appendChild(link);
        console.log(document.head);
        document.documentElement.style.setProperty('--font-family', selectedThemeModel.font.name.toString() + ',sans-serif');
        document.documentElement.style.setProperty('--bg-color', selectedThemeModel.colors.backgroundColor);
        document.documentElement.style.setProperty('--primary-color', selectedThemeModel.colors.primaryColor);
        document.documentElement.style.setProperty('--primary-color-dark', selectedThemeModel.colors.primaryColorDark);
        document.documentElement.style.setProperty('--footer-bg-color', selectedThemeModel.colors.footerBackgroundColor);
        document.documentElement.style.setProperty('--footer-primary-color', selectedThemeModel.colors.footerPrimaryColor);
        document.documentElement.style.setProperty('--footer-secondary-color', selectedThemeModel.colors.footerSecondaryColor);
        document.documentElement.style.setProperty('--text-color-primary', selectedThemeModel.colors.textColorPrimary);
        document.documentElement.style.setProperty('--text-color-secondary', selectedThemeModel.colors.textColorSecondary);
        document.documentElement.style.setProperty('--primary-button-text-color', selectedThemeModel.colors.primaryButtonTextColor);
        // document.documentElement.style.setProperty('--organisation-logo', selectedThemeModel.organisationLogo);
    };
    SurveyLibService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    SurveyLibService.ctorParameters = function () { return []; };
    /** @nocollapse */ SurveyLibService.ngInjectableDef = i0.defineInjectable({ factory: function SurveyLibService_Factory() { return new SurveyLibService(); }, token: SurveyLibService, providedIn: "root" });
    return SurveyLibService;
}());
export { SurveyLibService };
if (false) {
    /** @type {?} */
    SurveyLibService.prototype.currentElement;
    /** @type {?} */
    SurveyLibService.prototype.enableSubmit;
    /** @type {?} */
    SurveyLibService.prototype.defaultThankYouPage;
    /** @type {?} */
    SurveyLibService.prototype.isEmbededInDesigner;
    /** @type {?} */
    SurveyLibService.prototype.questionAnsweredEventEmitter;
    /** @type {?} */
    SurveyLibService.prototype.isQuestionCardCampaign;
    /** @type {?} */
    SurveyLibService.prototype.currentElementEventEmitter;
}
var CurrentElement = /** @class */ (function () {
    function CurrentElement() {
    }
    return CurrentElement;
}());
export { CurrentElement };
if (false) {
    /** @type {?} */
    CurrentElement.prototype.currentPageIndex;
    /** @type {?} */
    CurrentElement.prototype.currentQuestionIndex;
}
