// survey-lib/esm5/lib/survey-lib.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/survey-lib.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { SurveyContent } from './pages/pages.model';
import { SurveyLibService } from './survey-lib.service';
import { PagesService } from './shared/pages-services/pages.service';
import { SurveyResponses } from './pages/survey-response.model';
var SurveyLibComponent = /** @class */ (function () {
    function SurveyLibComponent(surveyLibraryService, pageService) {
        this.surveyLibraryService = surveyLibraryService;
        this.pageService = pageService;
        this.surveyResponseEmitter = new EventEmitter();
        this.isQuestionCardCampaign = false;
        this.responseFromQuestionCardCampaign = undefined;
    }
    /**
     * @return {?}
     */
    SurveyLibComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.isQuestionCardCampaign) {
            console.log('Inside question card campaign!');
            if (this.surveyContent.pages[0].pageName === 'Welcome page') {
                this.surveyContent.pages.splice(0, 1);
            }
            this.storeResponseFromQuestionCardCampaign();
            this.surveyLibraryService.isQuestionCardCampaign = true;
        }
        this.surveyLibraryService.initializeSurveyTheme(this.surveyContent.selectedTheme);
    };
    /**
     * @param {?} surveyResponse
     * @return {?}
     */
    SurveyLibComponent.prototype.emitResponse = /**
     * @param {?} surveyResponse
     * @return {?}
     */
    function (surveyResponse) {
        this.surveyResponseEmitter.emit(surveyResponse);
    };
    /**
     * @return {?}
     */
    SurveyLibComponent.prototype.storeResponseFromQuestionCardCampaign = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var question = this.surveyContent.pages[0].questions[0];
        /** @type {?} */
        var surveyResponse = new SurveyResponses();
        surveyResponse.question = question.question;
        surveyResponse.questionType = question.questionType;
        surveyResponse.identifier = question.identifier;
        surveyResponse.questionId = question.questionId;
        switch (question.questionType) {
            case 'Nps':
            case 'Rating':
                surveyResponse.rating = this.responseFromQuestionCardCampaign;
                this.pageService.persistSurvey(surveyResponse);
                break;
            case 'Yes / No':
                surveyResponse.answer = this.responseFromQuestionCardCampaign;
                this.pageService.persistSurvey(surveyResponse);
                break;
            case 'Statement':
                surveyResponse.answer = this.responseFromQuestionCardCampaign;
                this.pageService.persistSurvey(surveyResponse);
                break;
            case 'Date':
                surveyResponse.answer = this.responseFromQuestionCardCampaign;
                this.pageService.persistSurvey(surveyResponse);
                break;
            default:
                break;
        }
    };
    SurveyLibComponent.decorators = [
        { type: Component, args: [{
                    selector: 'omoto-survey-lib',
                    template: "\n    <app-pages [surveyContent]=\"surveyContent\"\n               (surveyResponseEmitter)=\"emitResponse($event)\"\n               [isTestMode]=\"testMode\"\n               [surveyTheme]=\"theme\"></app-pages>\n  "
                }] }
    ];
    /** @nocollapse */
    SurveyLibComponent.ctorParameters = function () { return [
        { type: SurveyLibService },
        { type: PagesService }
    ]; };
    SurveyLibComponent.propDecorators = {
        surveyResponseEmitter: [{ type: Output }],
        surveyContent: [{ type: Input }],
        testMode: [{ type: Input }],
        theme: [{ type: Input }],
        currentPageIndex: [{ type: Input }],
        currentQuestionIndex: [{ type: Input }],
        isQuestionCardCampaign: [{ type: Input }],
        responseFromQuestionCardCampaign: [{ type: Input }]
    };
    return SurveyLibComponent;
}());
export { SurveyLibComponent };
if (false) {
    /** @type {?} */
    SurveyLibComponent.prototype.surveyResponseEmitter;
    /** @type {?} */
    SurveyLibComponent.prototype.surveyContent;
    /** @type {?} */
    SurveyLibComponent.prototype.testMode;
    /** @type {?} */
    SurveyLibComponent.prototype.theme;
    /** @type {?} */
    SurveyLibComponent.prototype.currentPageIndex;
    /** @type {?} */
    SurveyLibComponent.prototype.currentQuestionIndex;
    /** @type {?} */
    SurveyLibComponent.prototype.isQuestionCardCampaign;
    /** @type {?} */
    SurveyLibComponent.prototype.responseFromQuestionCardCampaign;
    /**
     * @type {?}
     * @private
     */
    SurveyLibComponent.prototype.surveyLibraryService;
    /** @type {?} */
    SurveyLibComponent.prototype.pageService;
}
