// survey-lib/esm5/lib/pages/default-thank-you-page/default-thank-you-page.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/default-thank-you-page/default-thank-you-page.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
var DefaultThankYouPageComponent = /** @class */ (function () {
    function DefaultThankYouPageComponent() {
    }
    /**
     * @return {?}
     */
    DefaultThankYouPageComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        //this.thankYouPage = PersistenceService.get(PagesComponent.CURRENT_PAGE);
    };
    /**
     * @return {?}
     */
    DefaultThankYouPageComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
    };
    DefaultThankYouPageComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-default-thank-you-page',
                    template: "<div\n  class=\"survey-container-wrap\">\n  <div class=\"survey-statement\">Thanks for the feedback!</div>\n</div>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    DefaultThankYouPageComponent.ctorParameters = function () { return []; };
    return DefaultThankYouPageComponent;
}());
export { DefaultThankYouPageComponent };
