// survey-lib/esm2015/lib/pages/questionnaire/phone/phone.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/questionnaire/phone/phone.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Question } from '../../pages.model';
import { SurveyResponses } from '../../survey-response.model';
import { PagesService } from '../../../shared/pages-services/pages.service';
export class PhoneComponent {
    /**
     * @param {?} pageService
     */
    constructor(pageService) {
        this.pageService = pageService;
        this.submitEvent = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.pageService.stage == 'entry-next') {
            document.getElementById('phone-question').className = 'entry-next';
            document.getElementById('phone-question-delay').className = 'entry-next-delay';
        }
        else if (this.pageService.stage == 'entry-previous') {
            document.getElementById('phone-question').className = 'entry-previous-delay';
            document.getElementById('phone-question-delay').className = 'entry-previous';
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    onSubmit(value) {
        if (value != undefined) {
            this.surveyResponse.questionId = this.phone.questionId;
            this.surveyResponse.question = this.phone.question;
            this.surveyResponse.questionType = this.phone.questionType;
            this.surveyResponse.identifier = this.phone.identifier;
            this.surveyResponse.answer = value;
            this.pageService.persistSurvey(this.surveyResponse);
            document.getElementById('btn-primary').classList.add('blink-btn-animation');
            setTimeout((/**
             * @return {?}
             */
            () => {
                this.submitEvent.emit();
            }), 350);
        }
        else {
            this.submitEvent.emit();
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
    }
}
PhoneComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-phone',
                template: "<div id=\"phone-question\">\n<div class=\"survey-statement\">\n  <span *ngIf=\"phone.attributes.required\" style=\"color:red;margin-right: 0.5%\">*</span>{{phone.question}}\n</div>\n\n<p class=\"question-description\">\n  {{phone.attributes.description}}\n</p>\n</div>\n\n\n\n<div id=\"phone-question-delay\">\n<input class=\"survey-input-text-box\"\n  #phoneInput\n  type=\"number\"\n  name=\"phone\"\n  [(ngModel)]=\"surveyResponse.answer\"\n  onfocus=\"this.placeholder=''\"\n  placeholder=\"Enter valid phone number\">\n<button class=\"btn-primary\" id=\"btn-primary\" (click)=\"onSubmit(phoneInput.value)\">{{isLastQuestion?'Submit': 'Done'}}</button>\n<span\n  (click)=\"onSubmit(undefined)\"\n  *ngIf=\"!phone.attributes.required\" class=\"skip-this-question skip-this-question-marginLeft\">Skip</span>\n</div>\n",
                styles: [""]
            }] }
];
/** @nocollapse */
PhoneComponent.ctorParameters = () => [
    { type: PagesService }
];
PhoneComponent.propDecorators = {
    phone: [{ type: Input }],
    surveyResponse: [{ type: Input }],
    submitEvent: [{ type: Output }],
    isLastQuestion: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    PhoneComponent.prototype.phone;
    /** @type {?} */
    PhoneComponent.prototype.surveyResponse;
    /** @type {?} */
    PhoneComponent.prototype.submitEvent;
    /** @type {?} */
    PhoneComponent.prototype.isLastQuestion;
    /**
     * @type {?}
     * @private
     */
    PhoneComponent.prototype.pageService;
}
