// ngrx-forms/esm5/src/group/reducer/util.js
import * as tslib_1 from "tslib";
import { formStateReducer } from '../../reducer';
import { computeGroupState } from '../../state';
export function dispatchActionPerChild(controls, actionCreator) {
    var hasChanged = false;
    var newControls = Object.keys(controls)
        .reduce(function (c, key) {
        var _a;
        Object.assign(c, (_a = {}, _a[key] = formStateReducer(controls[key], actionCreator(controls[key].id)), _a));
        hasChanged = hasChanged || c[key] !== controls[key];
        return c;
    }, {});
    return hasChanged ? newControls : controls;
}
function callChildReducers(controls, action) {
    var hasChanged = false;
    var newControls = Object.keys(controls)
        .map(function (key) { return [key, formStateReducer(controls[key], action)]; })
        .reduce(function (res, _a) {
        var _b;
        var _c = tslib_1.__read(_a, 2), key = _c[0], state = _c[1];
        hasChanged = hasChanged || state !== controls[key];
        return Object.assign(res, (_b = {}, _b[key] = state, _b));
    }, {});
    return hasChanged ? newControls : controls;
}
export function childReducer(state, action) {
    var controls = callChildReducers(state.controls, action);
    if (state.controls === controls) {
        return state;
    }
    return computeGroupState(state.id, controls, state.value, state.errors, state.pendingValidations, state.userDefinedProperties, {
        wasOrShouldBeDirty: state.isDirty,
        wasOrShouldBeEnabled: state.isEnabled,
        wasOrShouldBeTouched: state.isTouched,
        wasOrShouldBeSubmitted: state.isSubmitted,
    });
}
