// survey-lib/esm2015/lib/pages/questionnaire/yes-no/yes-no.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/questionnaire/yes-no/yes-no.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Question } from '../../pages.model';
import { SurveyResponses } from '../../survey-response.model';
import { PagesService } from '../../../shared/pages-services/pages.service';
export class YesNoComponent {
    /**
     * @param {?} pageService
     */
    constructor(pageService) {
        this.pageService = pageService;
        this.isActive = false;
        this.submitEvent = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.pageService.stage == 'entry-next') {
            document.getElementById('yes-no-question').className = 'entry-next';
            document.getElementById('yes-no-question-delay').className = 'entry-next-delay';
        }
        else if (this.pageService.stage == 'entry-previous') {
            document.getElementById('yes-no-question').className = 'entry-previous-delay';
            document.getElementById('yes-no-question-delay').className = 'entry-previous';
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    onSelect(value) {
        if (value != undefined) {
            this.isActive = true;
            this.surveyResponse.question = this.yesNoQuestion.question;
            this.surveyResponse.questionType = this.yesNoQuestion.questionType;
            this.surveyResponse.questionId = this.yesNoQuestion.questionId;
            this.surveyResponse.identifier = this.yesNoQuestion.identifier;
            this.surveyResponse.answer = value;
            this.pageService.persistSurvey(this.surveyResponse);
            value == 'Yes' ? document.getElementById('yes-btn').classList.add('blink-btn-animation') : '';
            value == 'No' ? document.getElementById('no-btn').classList.add('blink-btn-animation') : '';
            setTimeout((/**
             * @return {?}
             */
            () => {
                this.submitEvent.emit();
            }), 350);
        }
        else {
            this.submitEvent.emit();
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
    }
}
YesNoComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-yes-no',
                template: "<div id=\"yes-no-question\">\n<div class=\"survey-statement\">\n  <span *ngIf=\"yesNoQuestion.attributes.required\" style=\"color:red;margin-right: 0.5%\">*</span>{{ yesNoQuestion.question }} </div>\n  <p class=\"question-description\">\n  {{yesNoQuestion.attributes.description}}\n</p>\n\n</div>\n\n\n\n<div id=\"yes-no-question-delay\">\n<button\n    name=\"yes-no\" value=\"yes\" id=\"yes-btn\"\n    [ngClass]=\"surveyResponse.answer == 'Yes'?'btn-secondary yes-no-btn btn-focused-state':'btn-secondary yes-no-btn'\"\n    (click)=\"onSelect('Yes')\"><span [ngClass]=\"surveyResponse.answer == 'Yes'? 'checkmark checkmark-checked':'checkmark'\"></span><span style=\"margin-left: 10px\">YES</span>\n</button>\n\n\n\n<button\n    name=\"yes-no\" value=\"no\" id=\"no-btn\"\n    [ngClass]=\"surveyResponse.answer == 'No'?'btn-secondary yes-no-btn btn-focused-state':'btn-secondary yes-no-btn'\"\n    (click)=\"onSelect('No')\"><span [ngClass]=\"surveyResponse.answer == 'No'? 'crossmark crossmark-checked':'crossmark'\"></span><span>NO</span>\n</button>\n\n\n<div style=\"margin-top:2%\" *ngIf=\"!yesNoQuestion.attributes.required\" class=\"skip-this-question\">\n  <span (click)=\"onSelect(undefined)\">Skip</span>\n</div>\n\n</div>\n",
                styles: [""]
            }] }
];
/** @nocollapse */
YesNoComponent.ctorParameters = () => [
    { type: PagesService }
];
YesNoComponent.propDecorators = {
    yesNoQuestion: [{ type: Input }],
    surveyResponse: [{ type: Input }],
    submitEvent: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    YesNoComponent.prototype.isActive;
    /** @type {?} */
    YesNoComponent.prototype.yesNoQuestion;
    /** @type {?} */
    YesNoComponent.prototype.surveyResponse;
    /**
     * @type {?}
     * @private
     */
    YesNoComponent.prototype.submitEvent;
    /**
     * @type {?}
     * @private
     */
    YesNoComponent.prototype.pageService;
}
