// survey-lib/esm5/lib/pages/thank-you-page/thank-you-page.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/thank-you-page/thank-you-page.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { Page } from '../pages.model';
var ThankYouPageComponent = /** @class */ (function () {
    // isActive: boolean = true;
    function ThankYouPageComponent() {
    }
    /**
     * @return {?}
     */
    ThankYouPageComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        //this.thankYouPage = PersistenceService.get(PagesComponent.CURRENT_PAGE);
        document.getElementById("thank-you-page").className = 'entry-next';
        document.getElementById("thank-you-page-delay").className = 'entry-next-delay';
    };
    ThankYouPageComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-thank-you-page',
                    template: "<div id=\"thank-you-page\">\n        <div class=\"welcome-thankyou-heading\">{{ thankYouPage.properties.headingValue }}</div>\n\n   <div class=\"survey-statement welcome-thankyou-text\">  {{ thankYouPage.properties.subHeadingValue }} </div>\n</div>\n\n<div id=\"thank-you-page-delay\">\n<button class=\"btn-primary\" id=\"btn-primary\" >{{ thankYouPage.properties.buttonText }}</button>\n</div>\n\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ThankYouPageComponent.ctorParameters = function () { return []; };
    ThankYouPageComponent.propDecorators = {
        thankYouPage: [{ type: Input }]
    };
    return ThankYouPageComponent;
}());
export { ThankYouPageComponent };
if (false) {
    /** @type {?} */
    ThankYouPageComponent.prototype.thankYouPage;
    /** @type {?} */
    ThankYouPageComponent.prototype.isActive;
}
