// survey-lib/esm5/lib/pages/questionnaire/date/date.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/questionnaire/date/date.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Question } from '../../pages.model';
import { SurveyResponses } from '../../survey-response.model';
import { PagesService } from '../../../shared/pages-services/pages.service';
var DateComponent = /** @class */ (function () {
    function DateComponent(pageService) {
        this.pageService = pageService;
        this.submitEvent = new EventEmitter();
    }
    /**
     * @return {?}
     */
    DateComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.platform = navigator.platform;
        console.log(navigator.platform);
        if (this.pageService.stage == 'entry-next') {
            document.getElementById('date-question').className = 'entry-next';
            document.getElementById('date-question-delay').className = 'entry-next-delay';
        }
        else if (this.pageService.stage == 'entry-previous') {
            document.getElementById('date-question').className = 'entry-previous-delay';
            document.getElementById('date-question-delay').className = 'entry-previous';
        }
    };
    /**
     * @param {?} date
     * @return {?}
     */
    DateComponent.prototype.onDateSelect = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        var _this = this;
        if (date != undefined) {
            this.surveyResponse.questionId = this.dateQuestion.questionId;
            this.surveyResponse.question = this.dateQuestion.question;
            this.surveyResponse.questionType = this.dateQuestion.questionType;
            this.surveyResponse.identifier = this.dateQuestion.identifier;
            this.surveyResponse.answer = date;
            this.pageService.persistSurvey(this.surveyResponse);
            document.getElementById('input').classList.add('blink-btn-animation');
            setTimeout((/**
             * @return {?}
             */
            function () {
                _this.submitEvent.emit();
            }), 350);
        }
        else {
            this.submitEvent.emit();
        }
    };
    /**
     * @return {?}
     */
    DateComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
    };
    DateComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-date',
                    template: "\n<div id=\"date-question\">\n<div class=\"survey-statement\">  <span *ngIf=\"dateQuestion.attributes.required\" class=\"isRequired\">*</span>{{dateQuestion.question}}</div>\n<p class=\"question-description\">\n        {{dateQuestion.attributes.description}}\n      </p>\n</div>\n\n<div id=\"date-question-delay\">\n  <!-- for IOS device function will be fired on (blur) -->\n<input *ngIf=\"platform == ('iPhone' || 'iPad')\" type=\"date\" id=\"input\" style=\"display:inline-block\" [(ngModel)]=\"surveyResponse.answer\"\n       (blur)=\"onDateSelect($event.target.value)\">\n  <!-- for IOS device function will be fired on (input) -->\n  <input *ngIf=\"platform !== 'iPhone'\" type=\"date\" id=\"input\" style=\"display:inline-block\" [(ngModel)]=\"surveyResponse.answer\"\n         (input)=\"onDateSelect($event.target.value)\">\n\n  <span\n  (click)=\"onDateSelect(undefined)\"\n  *ngIf=\"!dateQuestion.attributes.required\" class=\"skip-this-question skip-this-question-marginLeft\">Skip</span>\n</div>\n",
                    styles: ["[type=date]{background:var(--white-25) url(https://cdn1.iconfinder.com/data/icons/cc_mono_icon_set/blacks/16x16/calendar_2.png) 97% 50% no-repeat;height:2em;outline:0}input::-webkit-clear-button,input::-webkit-inner-spin-button{display:none}input::-webkit-calendar-picker-indicator{color:transparent;opacity:1}input::-webkit-calendar-picker-indicator:hover{background:var(--white-25);cursor:pointer}input{margin:10px 0 0;border:1px solid var(--white-25);background-color:var(--white-25);color:var(--text-color-secondary);padding:3px 5px;width:262px}"]
                }] }
    ];
    /** @nocollapse */
    DateComponent.ctorParameters = function () { return [
        { type: PagesService }
    ]; };
    DateComponent.propDecorators = {
        dateQuestion: [{ type: Input }],
        submitEvent: [{ type: Output }],
        surveyResponse: [{ type: Input }]
    };
    return DateComponent;
}());
export { DateComponent };
if (false) {
    /** @type {?} */
    DateComponent.prototype.platform;
    /** @type {?} */
    DateComponent.prototype.dateQuestion;
    /** @type {?} */
    DateComponent.prototype.submitEvent;
    /** @type {?} */
    DateComponent.prototype.surveyResponse;
    /**
     * @type {?}
     * @private
     */
    DateComponent.prototype.pageService;
}
