// survey-lib/esm5/lib/pages/questionnaire/short-text/short-text.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/questionnaire/short-text/short-text.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Question } from '../../pages.model';
import { SurveyResponses } from '../../survey-response.model';
import { PagesService } from '../../../shared/pages-services/pages.service';
var ShortTextComponent = /** @class */ (function () {
    function ShortTextComponent(pageService) {
        this.pageService = pageService;
        this.submitEvent = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ShortTextComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.pageService.stage == 'entry-next') {
            document.getElementById('short-question').className = 'entry-next';
            document.getElementById('short-question-delay').className = 'entry-next-delay';
        }
        else if (this.pageService.stage == 'entry-previous') {
            document.getElementById('short-question').className = 'entry-previous-delay';
            document.getElementById('short-question-delay').className = 'entry-previous';
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    ShortTextComponent.prototype.onInputValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        var _this = this;
        if (value != undefined) {
            this.surveyResponse.questionId = this.shortText.questionId;
            this.surveyResponse.question = this.shortText.question;
            this.surveyResponse.questionType = this.shortText.questionType;
            this.surveyResponse.identifier = this.shortText.identifier;
            this.surveyResponse.answer = value;
            this.pageService.persistSurvey(this.surveyResponse);
            document.getElementById('btn-primary').classList.add('blink-btn-animation');
            setTimeout((/**
             * @return {?}
             */
            function () {
                _this.submitEvent.emit();
            }), 350);
        }
        else {
            this.submitEvent.emit();
        }
    };
    /**
     * @return {?}
     */
    ShortTextComponent.prototype.skipQuestion = /**
     * @return {?}
     */
    function () {
        this.surveyResponse.identifier = this.shortText.identifier;
        this.surveyResponse.answer = undefined;
        this.submitEvent.emit();
    };
    /**
     * @return {?}
     */
    ShortTextComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
    };
    ShortTextComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-short-text',
                    template: "<div id=\"short-question\">\n<div class=\"survey-statement\">\n  <span *ngIf=\"shortText.attributes.required\" class=\"isRequired\">*</span>{{ shortText.question }}</div>\n\n  <p class=\"question-description\">\n  {{shortText.attributes.description}}\n</p>\n\n</div>\n\n<div id=\"short-question-delay\">\n<input class=\"survey-input-text-box\" #shortTextInput\n        type=\"text\" onfocus=\"this.placeholder=''\"\n       [(ngModel)]=\"surveyResponse.answer\"\n       [maxLength]=\"shortText.attributes.maxCharacterCount\"\n       placeholder=\"Enter text here\">\n<button class=\"btn-primary\" id=\"btn-primary\" (click)=\"onInputValue(shortTextInput.value)\">{{isLastQuestion?'Submit': 'Done'}}</button>\n<span\n  (click)=\"onInputValue(undefined)\"\n  *ngIf=\"!shortText.attributes.required\" class=\"skip-this-question skip-this-question-marginLeft\">Skip</span>\n</div>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ShortTextComponent.ctorParameters = function () { return [
        { type: PagesService }
    ]; };
    ShortTextComponent.propDecorators = {
        shortText: [{ type: Input }],
        surveyResponse: [{ type: Input }],
        submitEvent: [{ type: Output }],
        isLastQuestion: [{ type: Input }]
    };
    return ShortTextComponent;
}());
export { ShortTextComponent };
if (false) {
    /** @type {?} */
    ShortTextComponent.prototype.shortText;
    /** @type {?} */
    ShortTextComponent.prototype.surveyResponse;
    /** @type {?} */
    ShortTextComponent.prototype.submitEvent;
    /** @type {?} */
    ShortTextComponent.prototype.isLastQuestion;
    /**
     * @type {?}
     * @private
     */
    ShortTextComponent.prototype.pageService;
}
