// survey-lib/esm5/lib/pages/questionnaire/number/number.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/questionnaire/number/number.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Question } from '../../pages.model';
import { SurveyResponses } from '../../survey-response.model';
import { PagesService } from '../../../shared/pages-services/pages.service';
var NumberComponent = /** @class */ (function () {
    function NumberComponent(pageService) {
        this.pageService = pageService;
        this.submitEvent = new EventEmitter();
    }
    /**
     * @return {?}
     */
    NumberComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.pageService.stage == 'entry-next') {
            document.getElementById('number-question').className = 'entry-next';
            document.getElementById('number-question-delay').className = 'entry-next-delay';
        }
        else if (this.pageService.stage == 'entry-previous') {
            document.getElementById('number-question').className = 'entry-previous-delay';
            document.getElementById('number-question-delay').className = 'entry-previous';
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    NumberComponent.prototype.onSubmit = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        var _this = this;
        if (value != undefined) {
            this.surveyResponse.questionId = this.phone.questionId;
            this.surveyResponse.question = this.phone.question;
            this.surveyResponse.identifier = this.phone.identifier;
            this.surveyResponse.questionType = this.phone.questionType;
            this.surveyResponse.answer = value;
            this.pageService.persistSurvey(this.surveyResponse);
            document.getElementById('btn-primary').classList.add('blink-btn-animation');
            setTimeout((/**
             * @return {?}
             */
            function () {
                _this.submitEvent.emit();
            }), 350);
        }
        else {
            this.submitEvent.emit();
        }
    };
    /**
     * @return {?}
     */
    NumberComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
    };
    NumberComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-number',
                    template: "<div id=\"number-question\">\n<div class=\"survey-statement\">\n  <span *ngIf=\"phone.attributes.required\" class=\"isRequired\">*</span>{{phone.question}}\n</div>\n<p class=\"question-description\">\n  {{phone.attributes.description}}\n</p>\n</div>\n\n<div id=\"number-question-delay\">\n<input class=\"survey-input-text-box\"\n  #numberInput\n  type=\"number\"\n  name=\"number\"\n  [(ngModel)]=\"surveyResponse.answer\"\n  onfocus=\"this.placeholder=''\" placeholder=\"Number\">\n<button class=\"btn-primary\" id=\"btn-primary\" (click)=\"onSubmit(numberInput.value)\">{{isLastQuestion?'Submit': 'Done'}}</button>\n<span\n  (click)=\"onSubmit(undefined)\"\n  *ngIf=\"!phone.attributes.required\"\n  class=\"skip-this-question skip-this-question-marginLeft\">Skip</span>\n</div>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    NumberComponent.ctorParameters = function () { return [
        { type: PagesService }
    ]; };
    NumberComponent.propDecorators = {
        phone: [{ type: Input }],
        surveyResponse: [{ type: Input }],
        submitEvent: [{ type: Output }],
        isLastQuestion: [{ type: Input }]
    };
    return NumberComponent;
}());
export { NumberComponent };
if (false) {
    /** @type {?} */
    NumberComponent.prototype.phone;
    /** @type {?} */
    NumberComponent.prototype.surveyResponse;
    /** @type {?} */
    NumberComponent.prototype.submitEvent;
    /** @type {?} */
    NumberComponent.prototype.isLastQuestion;
    /**
     * @type {?}
     * @private
     */
    NumberComponent.prototype.pageService;
}
