// survey-lib/esm5/lib/pages/questionnaire/agree-disagree/agree-disagree.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/questionnaire/agree-disagree/agree-disagree.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Question } from '../../pages.model';
import { PagesService } from '../../../shared/pages-services/pages.service';
import { SurveyResponses } from '../../survey-response.model';
var AgreeDisagreeComponent = /** @class */ (function () {
    function AgreeDisagreeComponent(pageService) {
        this.pageService = pageService;
        this.submitEvent = new EventEmitter();
    }
    /**
     * @return {?}
     */
    AgreeDisagreeComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.pageService.stage == 'entry-next') {
            document.getElementById('agree-disagree-question').className = 'entry-next';
            document.getElementById('agree-disagree-question-delay').className = 'entry-next-delay';
        }
        else if (this.pageService.stage == 'entry-previous') {
            document.getElementById('agree-disagree-question').className = 'entry-previous-delay';
            document.getElementById('agree-disagree-question-delay').className = 'entry-previous';
        }
    };
    /**
     * @param {?} answer
     * @return {?}
     */
    AgreeDisagreeComponent.prototype.onSubmit = /**
     * @param {?} answer
     * @return {?}
     */
    function (answer) {
        var _this = this;
        if (answer != undefined) {
            this.surveyResponse.questionId = this.agreeDisagree.questionId;
            this.surveyResponse.question = this.agreeDisagree.question;
            this.surveyResponse.questionType = this.agreeDisagree.questionType;
            this.surveyResponse.identifier = this.agreeDisagree.identifier;
            this.surveyResponse.answer = answer;
            this.pageService.persistSurvey(this.surveyResponse);
            answer == 'Agree' ? document.getElementById('agree').classList.add('blink-btn-animation') : '';
            answer == 'Disagree' ? document.getElementById('disagree').classList.add('blink-btn-animation') : '';
            setTimeout((/**
             * @return {?}
             */
            function () {
                _this.submitEvent.emit();
            }), 350);
        }
        else {
            this.submitEvent.emit();
        }
    };
    /**
     * @return {?}
     */
    AgreeDisagreeComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
    };
    AgreeDisagreeComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-agree-disagree',
                    template: "<div id=\"agree-disagree-question\">\n<div class=\"survey-statement\">\n  <span *ngIf=\"agreeDisagree.attributes.required\" class=\"isRequired\">*</span>{{agreeDisagree.question}}\n</div>\n<p class=\"question-description\">\n  {{agreeDisagree.attributes.description}}\n</p>\n</div>\n\n\n<div id=\"agree-disagree-question-delay\">\n<button id=\"agree\"\n  [ngClass]=\"surveyResponse.answer == 'Agree'?'btn-secondary agree-disagree-btn btn-focused-state':'btn-secondary agree-disagree-btn'\"\n  (click)=\"onSubmit('Agree')\"><span [ngClass]=\"surveyResponse.answer == 'Agree'?'checkmark checkmark-checked':'checkmark'\"></span><span style=\"margin-left: 10px;\">AGREE</span>\n</button>\n\n<button id=\"disagree\"\n  [ngClass]=\"surveyResponse.answer == 'Disagree'?'btn-secondary agree-disagree-btn btn-focused-state':'btn-secondary agree-disagree-btn'\"\n  (click)=\"onSubmit('Disagree')\"><span [ngClass]=\"surveyResponse.answer == 'Disagree'?'crossmark crossmark-checked':'crossmark'\"></span><span>DISAGREE</span>\n</button>\n\n<div style=\"margin-top: 2%;\" class=\"skip-this-question\" *ngIf=\"!agreeDisagree.attributes.required\">\n  <span (click)=\"onSubmit(undefined)\" >Skip</span>\n</div>\n</div>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    AgreeDisagreeComponent.ctorParameters = function () { return [
        { type: PagesService }
    ]; };
    AgreeDisagreeComponent.propDecorators = {
        agreeDisagree: [{ type: Input }],
        surveyResponse: [{ type: Input }],
        submitEvent: [{ type: Output }]
    };
    return AgreeDisagreeComponent;
}());
export { AgreeDisagreeComponent };
if (false) {
    /** @type {?} */
    AgreeDisagreeComponent.prototype.agreeDisagree;
    /** @type {?} */
    AgreeDisagreeComponent.prototype.surveyResponse;
    /** @type {?} */
    AgreeDisagreeComponent.prototype.submitEvent;
    /**
     * @type {?}
     * @private
     */
    AgreeDisagreeComponent.prototype.pageService;
}
