// survey-lib/esm5/lib/pages/questionnaire/long-text/long-text.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/questionnaire/long-text/long-text.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Question } from '../../pages.model';
import { SurveyResponses } from '../../survey-response.model';
import { PagesService } from '../../../shared/pages-services/pages.service';
import { SurveyLibService } from '../../../survey-lib.service';
var LongTextComponent = /** @class */ (function () {
    function LongTextComponent(pageService, surveyLibServices) {
        this.pageService = pageService;
        this.surveyLibServices = surveyLibServices;
        this.submitEvent = new EventEmitter();
    }
    /**
     * @return {?}
     */
    LongTextComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.pageService.stage == 'entry-next') {
            document.getElementById('long-question').className = 'entry-next';
            document.getElementById('long-question-delay').className = 'entry-next-delay';
        }
        else if (this.pageService.stage == 'entry-previous') {
            document.getElementById('long-question').className = 'entry-previous-delay';
            document.getElementById('long-question-delay').className = 'entry-previous';
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    LongTextComponent.prototype.onInputValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        var _this = this;
        if (value != undefined) {
            this.surveyResponse.questionId = this.longText.questionId;
            this.surveyResponse.question = this.longText.question;
            this.surveyResponse.questionType = this.longText.questionType;
            this.surveyResponse.answer = value;
            this.surveyResponse.identifier = this.longText.identifier;
            this.pageService.persistSurvey(this.surveyResponse);
            document.getElementById('btn-primary').classList.add('blink-btn-animation');
            setTimeout((/**
             * @return {?}
             */
            function () {
                _this.submitEvent.emit();
            }), 350);
        }
        else {
            this.submitEvent.emit();
        }
    };
    /**
     * @return {?}
     */
    LongTextComponent.prototype.skipQuestion = /**
     * @return {?}
     */
    function () {
        this.submitEvent.emit();
    };
    /**
     * @return {?}
     */
    LongTextComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
    };
    LongTextComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-long-text',
                    template: "<div id=\"long-question\">\n\n<div class=\"survey-statement\">\n  <span *ngIf=\"longText.attributes.required\" class=\"isRequired\">*</span>{{ longText.question }}</div>\n  <p class=\"question-description\">\n  {{longText.attributes.description}}\n</p>\n\n</div>\n\n<div id=\"long-question-delay\">\n<textarea class=\"survey-input-textarea\" #longTextInput onfocus=\"this.placeholder=''\" name=\"message\"\n          placeholder=\"Add comment here\"\n          [maxLength]=\"longText.attributes.maxCharacterCount\"\n          [(ngModel)]=\"surveyResponse.answer\">\n  </textarea>\n<button class=\"btn-primary\"  id=\"btn-primary\" (click)=\"onInputValue(longTextInput.value)\">{{isLastQuestion?'Submit': 'Done'}}</button>\n<span *ngIf=\"!longText.attributes.required\"\n     (click)=\"onInputValue(undefined)\"\n     class=\"skip-this-question skip-this-question-marginLeft\">Skip</span>\n<!--<div class=\"row\">\n<button id=\"skip-button\"\n        (click)=\"skipQuestion()\">Skip</button>\n<button id=\"done-button\"\n        (click)=\"onInputValue()\">Done</button>\n</div>-->\n</div>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    LongTextComponent.ctorParameters = function () { return [
        { type: PagesService },
        { type: SurveyLibService }
    ]; };
    LongTextComponent.propDecorators = {
        longText: [{ type: Input }],
        surveyResponse: [{ type: Input }],
        submitEvent: [{ type: Output }],
        isLastQuestion: [{ type: Input }]
    };
    return LongTextComponent;
}());
export { LongTextComponent };
if (false) {
    /** @type {?} */
    LongTextComponent.prototype.longText;
    /** @type {?} */
    LongTextComponent.prototype.surveyResponse;
    /** @type {?} */
    LongTextComponent.prototype.submitEvent;
    /** @type {?} */
    LongTextComponent.prototype.isLastQuestion;
    /**
     * @type {?}
     * @private
     */
    LongTextComponent.prototype.pageService;
    /** @type {?} */
    LongTextComponent.prototype.surveyLibServices;
}
