// survey-lib/esm2015/lib/pages/questionnaire/mandatory-question-validation.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/questionnaire/mandatory-question-validation.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class MandatoryQuestionValidation {
    constructor() {
    }
    /**
     * @param {?} questionType
     * @param {?} response
     * @return {?}
     */
    isQuestionAnswered(questionType, response) {
        console.log('Inside isQuestionAnswered!');
        console.log('Question type: ' + questionType);
        if (response == undefined)
            return false;
        switch (questionType) {
            case 'Address':
                return this.isAddressFormAnswered(response);
            case 'Choice':
                return this.isChoiceQuestionAnswered(response);
            case 'Contact form':
                return this.isContactFormAnswered(response);
            case 'Date':
                return this.isDateQuestionAnswered(response);
            case 'Drop Down':
                return this.isDropDownQuestionAnswered(response);
            case 'Email':
                return this.isEmailQuestionAnswered(response);
            case 'Short text':
            case 'Long text':
                return this.isTextQuestionAnswered(response);
            case 'Nps':
            case 'Rating':
                return this.isRatingQuestionAnswered(response);
            case 'Statement':
                return this.isStatementQuestionAnswered(response);
            case 'Yes No':
            case 'Legal':
                return this.isYesNoQuestionAnswered(response);
            default: return false;
        }
    }
    /**
     * @private
     * @param {?} response
     * @return {?}
     */
    isAddressFormAnswered(response) {
        return false;
    }
    /**
     * @private
     * @param {?} response
     * @return {?}
     */
    isContactFormAnswered(response) {
        return false;
    }
    /**
     * @private
     * @param {?} response
     * @return {?}
     */
    isChoiceQuestionAnswered(response) {
        return response.choices != undefined && response.choices.length > 0;
    }
    /**
     * @private
     * @param {?} response
     * @return {?}
     */
    isDateQuestionAnswered(response) {
        return response.answer != undefined && response.answer.length > 0;
    }
    /**
     * @private
     * @param {?} response
     * @return {?}
     */
    isDropDownQuestionAnswered(response) {
        return response.answerId != undefined;
    }
    /**
     * @private
     * @param {?} response
     * @return {?}
     */
    isEmailQuestionAnswered(response) {
        return response.answer != undefined && response.answer.length > 0;
    }
    /**
     * @private
     * @param {?} response
     * @return {?}
     */
    isTextQuestionAnswered(response) {
        return response.answer != undefined && response.answer.length > 0;
    }
    /**
     * @private
     * @param {?} response
     * @return {?}
     */
    isRatingQuestionAnswered(response) {
        return response.rating != undefined && response.rating > -1;
    }
    /**
     * @private
     * @param {?} response
     * @return {?}
     */
    isStatementQuestionAnswered(response) {
        return false;
    }
    /**
     * @private
     * @param {?} response
     * @return {?}
     */
    isYesNoQuestionAnswered(response) {
        return response.answer != undefined && response.answer.length > 0;
    }
}
