// survey-lib/esm2015/lib/pages/questionnaire/questionnaire.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/questionnaire/questionnaire.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Page } from '../pages.model';
import { ExpressionBuilder } from './expression-builder';
import { MandatoryQuestionValidation } from './mandatory-question-validation';
import { PagesService } from '../../shared/pages-services/pages.service';
import { SurveyLibService } from '../../survey-lib.service';
export class QuestionnaireComponent {
    /**
     * @param {?} pagesService
     * @param {?} surveyLibService
     */
    constructor(pagesService, surveyLibService) {
        this.pagesService = pagesService;
        this.surveyLibService = surveyLibService;
        this.previousPageEvent = new EventEmitter();
        this.submitResponse = new EventEmitter();
        this.mandatoryQuestion = new MandatoryQuestionValidation();
        this.expressionBuilder = new ExpressionBuilder();
        this.isMandatoryQuestionAnswered = true;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        console.log(this.questionnairePage.questions.length);
        console.log(this.surveyLibService.currentElement.currentQuestionIndex);
        if (this.surveyLibService.isQuestionCardCampaign) {
            this.nextQuestion(true);
        }
        this.pagesService.stage = 'entry-next';
        this.pagesService.sequenceOfQuestionTraversed.push(this.surveyLibService.currentElement.currentQuestionIndex);
        this.previousQuestionSubscription = this.pagesService.previousQuestion.subscribe((/**
         * @return {?}
         */
        () => this.previousQuestion()));
        this.nextQuestionSubscription = this.pagesService.nextQuestion.subscribe((/**
         * @return {?}
         */
        () => this.nextQuestion()));
    }
    /**
     * @param {?=} skipAnimation
     * @return {?}
     */
    nextQuestion(skipAnimation = false) {
        console.log("nextQuestion");
        if (this.questionnairePage.questions[this.surveyLibService.currentElement.currentQuestionIndex].attributes.required) {
            if (this.mandatoryQuestion.isQuestionAnswered(this.questionnairePage
                .questions[this.surveyLibService.currentElement.currentQuestionIndex].questionType, this.pagesService.getResponse(this.questionnairePage.questions[this.surveyLibService
                .currentElement.currentQuestionIndex].identifier))) {
                this.jump(skipAnimation);
                this.isMandatoryQuestionAnswered = true;
            }
            else {
                this.isMandatoryQuestionAnswered = false;
                this.showErrorMessage();
            }
        }
        else
            this.jump(skipAnimation);
        console.log(this.surveyLibService.currentElement);
        this.surveyLibService.currentElementEventEmitter.emit(this.surveyLibService.currentElement);
        this.updateProgressBar();
    }
    /**
     * @return {?}
     */
    previousQuestion() {
        console.log(this.pagesService.sequenceOfQuestionTraversed);
        if (!this.surveyLibService.isEmbededInDesigner) {
            if (this.pagesService.sequenceOfQuestionTraversed.length > 1) {
                this.animatePreviousExit(this.questionnairePage.questions[this.surveyLibService.currentElement.currentQuestionIndex].questionType);
                setTimeout((/**
                 * @return {?}
                 */
                () => {
                    /** @type {?} */
                    let previousQuestionIndex = this.pagesService.sequenceOfQuestionTraversed[this.pagesService.sequenceOfQuestionTraversed.length - 2];
                    this.pagesService.sequenceOfQuestionTraversed = this.pagesService.sequenceOfQuestionTraversed.filter((/**
                     * @param {?} sequence
                     * @return {?}
                     */
                    sequence => sequence < this.surveyLibService.currentElement.currentQuestionIndex));
                    this.surveyLibService.currentElement.currentQuestionIndex = previousQuestionIndex;
                    this.updateProgressBar();
                    this.pagesService.stage = 'entry-previous';
                }), 400);
            }
            else {
                //this.resetResponse();
                this.previousPageEvent.emit();
            }
        }
        else {
            if (this.surveyLibService.currentElement.currentQuestionIndex > 0) {
                this.updateProgressBar();
                this.surveyLibService.currentElement.currentQuestionIndex--;
            }
            else {
                if (this.surveyLibService.currentElement.currentPageIndex > 0) {
                    this.previousPageEvent.emit();
                }
            }
        }
        this.surveyLibService.currentElementEventEmitter.emit(this.surveyLibService.currentElement);
    }
    /**
     * @private
     * @param {?} skipAnimation
     * @return {?}
     */
    jump(skipAnimation) {
        this.animateExitNext(this.questionnairePage.questions[this.surveyLibService.currentElement.currentQuestionIndex].questionType);
        setTimeout((/**
         * @return {?}
         */
        () => {
            if (this.questionnairePage.questions[this.surveyLibService.currentElement.currentQuestionIndex].logicJumps != undefined && this.questionnairePage.questions[this.surveyLibService.currentElement.currentQuestionIndex].logicJumps.blocks.length > 0)
                this.logicJump();
            else
                this.defaultJump();
            this.pagesService.stage = 'entry-next';
        }), skipAnimation ? 0 : 400);
    }
    /**
     * @private
     * @return {?}
     */
    logicJump() {
        console.log('Inside logicJump: ' + this.surveyLibService.currentElement.currentQuestionIndex);
        /** @type {?} */
        let logicJump = false;
        for (let block of this.questionnairePage.questions[this.surveyLibService.currentElement.currentQuestionIndex].logicJumps.blocks) {
            console.warn(this.pagesService.surveyResponses);
            if (eval(this.expressionBuilder.build(block.conditions, this.pagesService.surveyResponses, this.questionnairePage.questions[this.surveyLibService.currentElement.currentQuestionIndex].questionType))) {
                /** @type {?} */
                let jumpIndex = this.questionnairePage.questions.findIndex((/**
                 * @param {?} question
                 * @return {?}
                 */
                question => question.questionId == block.thatIndex));
                if (QuestionnaireComponent.isBackwardJump(this.surveyLibService.currentElement.currentQuestionIndex, jumpIndex)) {
                    this.backwardJump(jumpIndex);
                }
                else {
                    this.forwardJump(jumpIndex);
                }
                logicJump = true;
                break;
            }
        }
        if (!logicJump) {
            this.defaultJump();
            /*this.surveyLibService.currentElement.currentQuestionIndex = this.questionnairePage.questions.findIndex(question => question.questionId == this.questionnairePage.questions[this.surveyLibService.currentElement.currentQuestionIndex].logicJumps.defaultJump);
            this.questionnairePage.questions[this.surveyLibService.currentElement.currentQuestionIndex] = this.questionnairePage.questions[this.surveyLibService.currentElement.currentQuestionIndex];
            this.pagesService.sequenceOfQuestionTraversed.push(this.surveyLibService.currentElement.currentQuestionIndex);*/
        }
    }
    /**
     * @private
     * @return {?}
     */
    defaultJump() {
        if (this.surveyLibService.currentElement.currentQuestionIndex < this.questionnairePage.questions.length - 1) {
            this.surveyLibService.currentElement.currentQuestionIndex++;
            this.questionnairePage.questions[this.surveyLibService.currentElement.currentQuestionIndex] = this.questionnairePage.questions[this.surveyLibService.currentElement.currentQuestionIndex];
            this.pagesService.sequenceOfQuestionTraversed.push(this.surveyLibService.currentElement.currentQuestionIndex);
            console.log('Inside default jump: ' + this.surveyLibService.currentElement.currentQuestionIndex);
            this.surveyLibService.enableSubmit = false;
        }
        else {
            this.surveyLibService.enableSubmit = true;
            this.submitResponse.emit();
        }
    }
    /**
     * @private
     * @param {?} currentQuestionIndex
     * @param {?} jumpIndex
     * @return {?}
     */
    static isBackwardJump(currentQuestionIndex, jumpIndex) {
        return jumpIndex < currentQuestionIndex;
    }
    /**
     * @private
     * @param {?} jumpIndex
     * @return {?}
     */
    backwardJump(jumpIndex) {
        console.log('Inside backwardJump: ' + this.surveyLibService.currentElement.currentQuestionIndex);
        this.surveyLibService.currentElement.currentQuestionIndex = jumpIndex;
        this.questionnairePage.questions[this.surveyLibService.currentElement.currentQuestionIndex] = this.questionnairePage.questions[this.surveyLibService.currentElement.currentQuestionIndex];
        this.pagesService.sequenceOfQuestionTraversed = this.pagesService.sequenceOfQuestionTraversed.filter((/**
         * @param {?} sequence
         * @return {?}
         */
        sequence => sequence <= jumpIndex));
        this.removeResponses(this.pagesService.sequenceOfQuestionTraversed);
    }
    /**
     * @private
     * @param {?} jumpIndex
     * @return {?}
     */
    forwardJump(jumpIndex) {
        console.log('Inside forwardJump: ' + this.surveyLibService.currentElement.currentQuestionIndex);
        this.surveyLibService.currentElement.currentQuestionIndex = jumpIndex;
        this.questionnairePage.questions[this.surveyLibService.currentElement.currentQuestionIndex] = this.questionnairePage.questions[this.surveyLibService.currentElement.currentQuestionIndex];
        this.pagesService.sequenceOfQuestionTraversed.push(this.surveyLibService.currentElement.currentQuestionIndex);
    }
    /**
     * @private
     * @param {?} questionIndexes
     * @return {?}
     */
    removeResponses(questionIndexes) {
        /** @type {?} */
        let excludeQuestionIds = [];
        for (let index of questionIndexes) {
            excludeQuestionIds.push(this.questionnairePage.questions[index].questionId);
        }
        this.pagesService.removeResponsesOfExcludedQuestionIds(excludeQuestionIds);
    }
    /**
     * @private
     * @return {?}
     */
    updateProgressBar() {
        document.getElementById('step-number').innerText = this.pagesService.surveyResponses.length.toString();
        //document.getElementById('progress').setAttribute('value', this.pagesService.response.length.toString());
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.previousQuestionSubscription.unsubscribe();
        this.nextQuestionSubscription.unsubscribe();
    }
    /**
     * @private
     * @return {?}
     */
    showErrorMessage() {
        /** @type {?} */
        let x = document.getElementById("error_snackbar");
        document.getElementById('message').innerText = this.questionnairePage
            .questions[this.surveyLibService.currentElement.currentQuestionIndex].attributes
            .requiredErrorMessage;
        x.className = "show";
        setTimeout((/**
         * @return {?}
         */
        () => { x.className = x.className.replace("show", ""); }), 3000);
    }
    /**
     * @private
     * @param {?} questionType
     * @return {?}
     */
    animateExitNext(questionType) {
        console.log(questionType);
        /** @type {?} */
        let elementById = null;
        switch (questionType) {
            case 'Nps':
                elementById = document.getElementById('nps-question');
                if (elementById !== null) {
                    document.getElementById('nps-question').className = 'exit-next';
                    document.getElementById('nps-question-delay').className = 'exit-next-delay';
                }
                break;
            case 'Choice':
                elementById = document.getElementById('choice-question');
                if (elementById !== null) {
                    document.getElementById('choice-question').className = 'exit-next';
                    document.getElementById('choice-question-delay').className = 'exit-next-delay';
                }
                break;
            case 'Rating':
                elementById = document.getElementById('rating-question');
                if (elementById !== null) {
                    document.getElementById('rating-question').className = 'exit-next';
                    document.getElementById('rating-question-delay').className = 'exit-next-delay';
                }
                break;
            case 'Drop down':
                elementById = document.getElementById('drop-down-question');
                if (elementById !== null) {
                    document.getElementById('drop-down-question').className = 'exit-next';
                    document.getElementById('drop-down-question-delay').className = 'exit-next-delay';
                }
                break;
            case 'Yes / No':
                elementById = document.getElementById('yes-no-question');
                if (elementById !== null) {
                    document.getElementById('yes-no-question').className = 'exit-next';
                    document.getElementById('yes-no-question-delay').className = 'exit-next-delay';
                }
                break;
            case 'Short text':
                elementById = document.getElementById('short-question');
                if (elementById !== null) {
                    document.getElementById('short-question').className = 'exit-next';
                    document.getElementById('short-question-delay').className = 'exit-next-delay';
                }
                break;
            case 'Long text':
                elementById = document.getElementById('long-question');
                if (elementById !== null) {
                    document.getElementById('long-question').className = 'exit-next';
                    document.getElementById('long-question-delay').className = 'exit-next-delay';
                }
                break;
            case 'Statement':
                elementById = document.getElementById('statement-question');
                if (elementById !== null) {
                    document.getElementById('statement-question').className = 'exit-next';
                    document.getElementById('statement-question-delay').className = 'exit-next-delay';
                }
                break;
            case 'Email':
                elementById = document.getElementById('email-question');
                if (elementById !== null) {
                    document.getElementById('email-question').className = 'exit-next';
                    document.getElementById('email-question-delay').className = 'exit-next-delay';
                }
                break;
            case 'Phone number':
                elementById = document.getElementById('phone-question');
                if (elementById !== null) {
                    document.getElementById('phone-question').className = 'exit-next';
                    document.getElementById('phone-question-delay').className = 'exit-next-delay';
                }
                break;
            case 'Number':
                elementById = document.getElementById('number-question');
                if (elementById !== null) {
                    document.getElementById('number-question').className = 'exit-next';
                    document.getElementById('number-question-delay').className = 'exit-next-delay';
                }
                break;
            case 'Name':
                elementById = document.getElementById('name-question');
                if (elementById !== null) {
                    document.getElementById('name-question').className = 'exit-next';
                    document.getElementById('name-question-delay').className = 'exit-next-delay';
                }
                break;
            case 'Website':
                elementById = document.getElementById('website-question');
                if (elementById !== null) {
                    document.getElementById('website-question').className = 'exit-next';
                    document.getElementById('website-question-delay').className = 'exit-next-delay';
                }
                break;
            case 'Legal':
                elementById = document.getElementById('agree-disagree-question');
                if (elementById !== null) {
                    document.getElementById('agree-disagree-question').className = 'exit-next';
                    document.getElementById('agree-disagree-question-delay').className = 'exit-next-delay';
                }
                break;
            case 'Date':
                elementById = document.getElementById('date-question');
                if (elementById !== null) {
                    document.getElementById('date-question').className = 'exit-next';
                    document.getElementById('date-question-delay').className = 'exit-next-delay';
                }
                break;
        }
    }
    /**
     * @private
     * @param {?} questionType
     * @return {?}
     */
    animatePreviousExit(questionType) {
        switch (questionType) {
            case 'Nps':
                document.getElementById('nps-question').className = 'exit-previous-delay';
                document.getElementById('nps-question-delay').className = 'exit-previous';
                break;
            case 'Choice':
                document.getElementById('choice-question').className = 'exit-previous-delay';
                document.getElementById('choice-question-delay').className = 'exit-previous';
                break;
            case 'Rating':
                document.getElementById('rating-question').className = 'exit-previous-delay';
                document.getElementById('rating-question-delay').className = 'exit-previous';
                break;
            case 'Drop down':
                document.getElementById('drop-down-question').className = 'exit-previous-delay';
                document.getElementById('drop-down-question-delay').className = 'exit-previous';
                break;
            case 'Yes / No':
                document.getElementById('yes-no-question').className = 'exit-previous-delay';
                document.getElementById('yes-no-question-delay').className = 'exit-previous';
                break;
            case 'Short text':
                document.getElementById('short-question').className = 'exit-previous-delay';
                document.getElementById('short-question-delay').className = 'exit-previous';
                break;
            case 'Long text':
                document.getElementById('long-question').className = 'exit-previous-delay';
                document.getElementById('long-question-delay').className = 'exit-previous';
                break;
            case 'Statement':
                document.getElementById('statement-question').className = 'exit-previous-delay';
                document.getElementById('statement-question-delay').className = 'exit-previous';
                break;
            case 'Email':
                document.getElementById('email-question').className = 'exit-previous-delay';
                document.getElementById('email-question-delay').className = 'exit-previous';
                break;
            case 'Phone number':
                document.getElementById('phone-question').className = 'exit-previous-delay';
                document.getElementById('phone-question-delay').className = 'exit-previous';
                break;
            case 'Number':
                document.getElementById('number-question').className = 'exit-previous-delay';
                document.getElementById('number-question-delay').className = 'exit-previous';
                break;
            case 'Name':
                document.getElementById('name-question').className = 'exit-previous-delay';
                document.getElementById('name-question-delay').className = 'exit-previous';
                break;
            case 'Website':
                document.getElementById('website-question').className = 'exit-previous-delay';
                document.getElementById('website-question-delay').className = 'exit-previous';
                break;
            case 'Legal':
                document.getElementById('agree-disagree-question').className = 'exit-previous-delay';
                document.getElementById('agree-disagree-question-delay').className = 'exit-previous';
                break;
            case 'Date':
                document.getElementById('date-question').className = 'exit-previous-delay';
                document.getElementById('date-question-delay').className = 'exit-previous';
                break;
        }
    }
}
QuestionnaireComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-questionnaire',
                template: "<app-choice\n  [isLastQuestion]=\"(questionnairePage.questions.length -1) === this.surveyLibService.currentElement.currentQuestionIndex \"\n\n  [choiceQuestion]=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex]\"\n            *ngIf=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].questionType == 'Choice'\"\n            [isMultipleChoice]=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].attributes.multipleChoice\"\n            [surveyResponse]=\"pagesService.getResponse(questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].identifier)\"\n            (submitEvent)=\"nextQuestion()\">\n</app-choice>\n\n<app-nps\n  *ngIf=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].questionType == 'Nps'\"\n  [npsQuestion]=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex]\"\n  [surveyResponse]=\"pagesService.getResponse(questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].identifier)\"\n  (submitEvent)=\"nextQuestion()\">\n</app-nps>\n\n<app-yes-no\n  [yesNoQuestion]=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex]\"\n  [surveyResponse]=\"pagesService.getResponse(questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].identifier)\"\n  *ngIf=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].questionType == 'Yes / No'\"\n  (submitEvent)=\"nextQuestion()\">\n</app-yes-no>\n\n<app-rating\n  *ngIf=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].questionType == 'Rating'\"\n  [surveyResponse]=\"pagesService.getResponse(questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].identifier)\"\n  [ratingQuestion]=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex]\"\n  (submitEvent)=\"nextQuestion()\">\n</app-rating>\n\n\n<app-drop-down\n  [surveyResponse]=\"pagesService.getResponse(questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].identifier)\"\n  *ngIf=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].questionType == 'Drop down'\"\n  [dropDown]=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex]\"\n  (submitEvent)=\"nextQuestion()\">\n</app-drop-down>\n\n\n<app-short-text\n  [isLastQuestion]=\"(questionnairePage.questions.length -1) === this.surveyLibService.currentElement.currentQuestionIndex \"\n\n  [surveyResponse]=\"pagesService.getResponse(questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].identifier)\"\n  *ngIf=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].questionType == 'Short text'\"\n  [shortText]=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex]\"\n  (submitEvent)=\"nextQuestion()\">\n</app-short-text>\n\n\n<app-long-text\n  [isLastQuestion]=\"(questionnairePage.questions.length -1) === this.surveyLibService.currentElement.currentQuestionIndex \"\n\n  [surveyResponse]=\"pagesService.getResponse(questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].identifier)\"\n  [longText]=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex]\"\n  *ngIf=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].questionType == 'Long text'\"\n  (submitEvent)=\"nextQuestion()\">\n</app-long-text>\n\n<app-statement\n  [surveyResponse]=\"pagesService.getResponse(questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].identifier)\"\n  *ngIf=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].questionType == 'Statement'\"\n  [statement]=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex]\"\n  (submitEvent)=\"nextQuestion()\">\n</app-statement>\n\n\n<app-email\n  [isLastQuestion]=\"(questionnairePage.questions.length -1) === this.surveyLibService.currentElement.currentQuestionIndex \"\n  [surveyResponse]=\"pagesService.getResponse(questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].identifier)\"\n  *ngIf=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].questionType == 'Email'\"\n  [email]=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex]\"\n  (submitEvent)=\"nextQuestion()\">\n</app-email>\n\n<app-phone\n  [isLastQuestion]=\"(questionnairePage.questions.length -1) === this.surveyLibService.currentElement.currentQuestionIndex \"\n\n  [surveyResponse]=\"pagesService.getResponse(questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].identifier)\"\n  *ngIf=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].questionType == 'Phone number'\"\n  [phone]=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex]\"\n  (submitEvent)=\"nextQuestion()\">\n</app-phone>\n\n<app-number\n  [isLastQuestion]=\"(questionnairePage.questions.length -1) === this.surveyLibService.currentElement.currentQuestionIndex \"\n\n  [surveyResponse]=\"pagesService.getResponse(questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].identifier)\"\n  *ngIf=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].questionType == 'Number'\"\n  [phone]=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex]\"\n  (submitEvent)=\"nextQuestion()\">\n</app-number>\n\n\n<app-name [name]=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex]\"\n          [surveyResponse]=\"pagesService.getResponse(questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].identifier)\"\n          *ngIf=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].questionType == 'Name'\"\n          (submitEvent)=\"nextQuestion()\">\n</app-name>\n\n<app-website\n  [isLastQuestion]=\"(questionnairePage.questions.length -1) === this.surveyLibService.currentElement.currentQuestionIndex \"\n  [website]=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex]\"\n             [surveyResponse]=\"pagesService.getResponse(questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].identifier)\"\n             *ngIf=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].questionType == 'Website'\"\n             (submitEvent)=\"nextQuestion()\"></app-website>\n\n<app-agree-disagree\n  *ngIf=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].questionType == 'Legal'\"\n  [agreeDisagree]=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex]\"\n  [surveyResponse]=\"pagesService.getResponse(questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].identifier)\"\n  (submitEvent)=\"nextQuestion()\">\n</app-agree-disagree>\n\n<app-date [dateQuestion]=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex]\"\n          *ngIf=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].questionType == 'Date'\"\n          [surveyResponse]=\"pagesService.getResponse(questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].identifier)\"\n          (submitEvent)=\"nextQuestion()\">\n</app-date>\n<!-- <app-contact-form *ngIf=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].questionType == 'Contact Form'\"\n                   [contactForm] = \"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex]\"\n                   (submitEvent)=\"nextQuestion()\">\n </app-contact-form>\n <app-address *ngIf=\"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex].questionType == 'Address Form'\"\n              [addressForm] = \"questionnairePage.questions[surveyLibService.currentElement.currentQuestionIndex]\"\n              (submitEvent)=\"nextQuestion()\">\n </app-address>-->\n\n\n\n\n<!--<button class=\"back-button\" mat-button (click)=\"previousQuestion()\">Back</button>\n<button class=\"submit-button\"\n        mat-button\n        *ngIf=\"isLastQuestion\"\n        (click)=\"submitResponse()\">Submit</button>-->\n\n<div id=\"error_snackbar\"><span id=\"message\">Error</span></div>\n",
                styles: ["/*!* START OF ANIMATION*!*//*!* END OF ANIMATION*!*/", ":root{--white-25:rgba(255, 255, 255, 0.25);--font-family:'Ubuntu',sans-serif;--bg-color:#f4a06e;--primary-color:#004e89;--primary-color-dark:#001B56;--footer-bg-color:#f9cfb6;--footer-primary-color:#b98f76;--footer-secondary-color:#fce7da;--text-color-primary:#ffffff;--text-color-secondary:#4a4a4a;--primary-button-text-color:#fffff;--positiveTranslate:translateX(350px);--negativeTranslate:translateX(-350px);--negativeTranslate3D:translate3d(-350px,0,0)}*,:after,:before{box-sizing:inherit}body,html{margin:0;padding:0;height:100%;width:100%;box-sizing:border-box;transition:none!important}.entry-next{-webkit-animation:.6s cubic-bezier(.215,.61,.355,1) both animMoveRightToCurrent;animation:.6s cubic-bezier(.215,.61,.355,1) both animMoveRightToCurrent}.entry-next-delay{-webkit-animation:.6s cubic-bezier(.215,.61,.355,1) .1s both animMoveRightToCurrent;animation:.6s cubic-bezier(.215,.61,.355,1) .1s both animMoveRightToCurrent}@-webkit-keyframes animMoveRightToCurrent{from{opacity:0;-webkit-transform:var(--positiveTranslate)}to{opacity:1;-webkit-transform:translateY(0)}}@keyframes animMoveRightToCurrent{from{opacity:0;-webkit-transform:var(--positiveTranslate)}to{opacity:1;-webkit-transform:translateY(0)}}.exit-next{-webkit-animation:.6s cubic-bezier(.215,.61,.355,1) forwards animMoveCurrentToLeft;animation:.6s cubic-bezier(.215,.61,.355,1) forwards animMoveCurrentToLeft}.exit-next-delay{-webkit-animation:.6s cubic-bezier(.215,.61,.355,1) .1s forwards animMoveCurrentToLeft;animation:.6s cubic-bezier(.215,.61,.355,1) .1s forwards animMoveCurrentToLeft}@-webkit-keyframes animMoveCurrentToLeft{to{opacity:0;-webkit-transform:var(--negativeTranslate3D)}}@keyframes animMoveCurrentToLeft{to{opacity:0;-webkit-transform:var(--negativeTranslate3D);transform:var(--negativeTranslate3D)}}.entry-previous{-webkit-animation:.6s cubic-bezier(.215,.61,.355,1) both animMoveLeftToCurrent;animation:.6s cubic-bezier(.215,.61,.355,1) .1s both animMoveLeftToCurrent}.entry-previous-delay{-webkit-animation:.6s cubic-bezier(.215,.61,.355,1) .1s both animMoveLeftToCurrent;animation:.6s cubic-bezier(.215,.61,.355,1) .1s both animMoveLeftToCurrent}@-webkit-keyframes animMoveLeftToCurrent{from{opacity:0;-webkit-transform:var(--negativeTranslate)}to{opacity:1;-webkit-transform:translateY(0)}}@keyframes animMoveLeftToCurrent{from{opacity:0;-webkit-transform:var(--negativeTranslate);transform:var(--negativeTranslate)}to{opacity:1;-webkit-transform:translateY(0);transform:translateY(0)}}.exit-previous{-webkit-animation:.6s cubic-bezier(.215,.61,.355,1) .1s forwards animMoveCurrentToRight;animation:.6s cubic-bezier(.215,.61,.355,1) .1s forwards animMoveCurrentToRight}.exit-previous-delay{-webkit-animation:.6s cubic-bezier(.215,.61,.355,1) forwards animMoveCurrentToRight;animation:.6s cubic-bezier(.215,.61,.355,1) forwards animMoveCurrentToRight}@-webkit-keyframes animMoveCurrentToRight{from{opacity:1;-webkit-transform:translateY(0)}to{opacity:0;-webkit-transform:var(--positiveTranslate)}}@keyframes animMoveCurrentToRight{from{opacity:1;-webkit-transform:translateY(0);transform:translateY(0)}to{opacity:0;-webkit-transform:var(--positiveTranslate);transform:var(--positiveTranslate)}}.input{border-radius:3px;padding:5px 0 5px 10px;color:var(--text-color-secondary);background-color:var(--white-25);font-size:.875rem;position:relative;box-shadow:1px 1px 3px 0 rgba(0,0,0,.5);height:31px}.select{cursor:pointer;position:relative;width:250px}.select .input:after,.select .input:before{content:'';position:absolute;left:calc(100% - 20px);-webkit-transform:translateY(-50%);transform:translateY(-50%);width:0;height:0}.select .input:before{border-top:5px solid var(--text-color-secondary);border-left:5px solid transparent;border-right:5px solid transparent;top:50%}.select .input:after{border-top:5px solid var(--white-25);border-left:5px solid transparent;border-right:5px solid transparent;top:calc(50% - 1px)}.select-options{position:absolute;width:100%;border-radius:4px;z-index:100;top:104%;left:0;display:none;overflow:hidden;box-shadow:1px 1px 3px 0 rgba(0,0,0,.5)}.select-options.visible{display:block;background-color:var(--white-25);z-index:1}.select-options ul{list-style:none;padding:0;margin:0}.select-options ul li{padding:5px 0 5px 10px;border-bottom:1px solid #ddd;font-size:.875rem;color:var(--text-color-secondary);background-color:var(--white-25);position:relative;overflow:hidden}.select-options ul li:hover{background-color:var(--primary-color);color:var(--text-color-primary)}.select-options ul li:last-child{border-bottom:0}.select-options ul li.selected{color:var(--text-color-primary);background-color:var(--primary-color);padding:5px 5px 5px 35px}.select-options ul li.selected:before{content:'';width:5px;height:10px;color:#000;border:2px solid var(--text-color-primary);border-right:0;border-bottom:0;right:calc(100% - 20px);top:50%;-webkit-transform:translateY(-50%) rotate(-135deg);transform:translateY(-50%) rotate(-135deg);position:absolute}.survey-container{font-weight:400;font-style:normal;font-stretch:normal;letter-spacing:normal;font-family:var(--font-family);line-height:1.5;background-color:var(--bg-color);height:calc(100vh - 80px);width:100%;overflow-y:auto;overflow-x:hidden;margin:0;padding:0}.survey-container-wrap{margin-left:auto;margin-right:auto;margin-bottom:10px;padding-top:8.18em;max-width:46.125em;color:var(--primary-color)}.skip-this-question{color:var(--text-color-secondary);cursor:pointer;font-size:.85em;clear:both}.skip-this-question-marginLeft{margin-left:2%}.skip-this-question:hover{color:var(--primary-color)}.question-description{margin-top:5px;font-size:14px}.footer{height:80px;position:absolute;bottom:0;width:100%;font-family:var(--font-family);background-color:var(--footer-bg-color)}.footer-container{position:absolute;bottom:0;left:50%;max-width:46.125em;width:46.125em;margin-left:-23.0625em}.footer-right{padding:15px 0;float:right}.footer-right img{width:81px}.footer-left{float:left;margin:10px 0 0}.progress-container{width:16.25em;display:inline-block}.progress-container p{margin:0;line-height:normal;font-size:.875rem;color:var(--footer-primary-color)}.progress-bar{width:100%;display:inline-block}.arrowUp-arrowDown-container{display:inline-block;margin:0 0 0 1em}.arrowUp-arrowDown-btn{width:25px;height:25px;background-color:var(--footer-bg-color);border-radius:3px;margin:0 10px 0 0;border:1px solid var(--footer-primary-color);cursor:pointer;padding:0}.arrowUp-arrowDown-btn:active{background:var(--footer-primary-color);outline:var(--footer-primary-color);color:var(--footer-secondary-color)}.arrowUp-arrowDown-btn:active span{border:solid var(--footer-secondary-color);border-width:0 2px 2px 0}.arrowUp-arrowDown-btn:focus{outline:var(--footer-primary-color)}.arrowUp-arrowDown-container span{border:solid var(--footer-primary-color);border-width:0 2px 2px 0;display:inline-block;padding:3px}.arrowUp{transform:rotate(-45deg);-webkit-transform:rotate(-45deg);margin-right:3px}.arrowDown{transform:rotate(135deg);-webkit-transform:rotate(135deg);margin-left:3px}progress{width:100%;border:1px solid var(--footer-primary-color);border-radius:14px}progress::-moz-progress-bar{border-radius:14px;background:var(--footer-primary-color);transition:.5s ease-out}progress::-webkit-progress-bar{border-radius:14px;background:var(--footer-secondary-color)}progress::-webkit-progress-value{border-radius:14px;background:var(--footer-primary-color);transition:.5s ease-out}.org-logo{display:block;max-width:250px;max-height:70px;min-height:50px}.welcome-thankyou-heading{font-size:1.75em;margin:2rem 0 0;font-weight:700}.welcome-thankyou-text{margin:0!important}.survey-statement{font-size:1.31em;font-weight:500;margin:2rem 0 0}.blink-btn-animation{-webkit-animation:.25s 2 blinkAnimation;animation:.25s 2 blinkAnimation}@-webkit-keyframes blinkAnimation{50%{opacity:.3}}@keyframes blinkAnimation{50%{opacity:.3}}.btn-primary{margin:1.56em 0 0;border-radius:4px;text-align:center;background-color:var(--primary-color);border:1px solid var(--primary-color);font-size:1rem;min-width:7.31em;min-height:2.5em;color:var(--primary-button-text-color);font-family:var(--font-family);cursor:pointer}.btn-primary:focus{background-color:var(--primary-color-dark);border:1px solid var(--primary-color-dark);outline:0}.btn-primary:hover{box-shadow:0 11px 22px -6px rgba(0,0,0,.5)}.btn-primary:active{background-color:var(--primary-color-dark)}.btn-secondary{border-radius:4px;margin:10px 10px 0 0;cursor:pointer;color:var(--text-color-secondary);border:1px solid var(--white-25);background-color:var(--white-25);font-family:var(--font-family);font-size:1rem;box-shadow:1px 1px 3px 0 rgba(0,0,0,.5);transition:.3s}.btn-secondary:hover{box-shadow:0 0 3px 0 var(--primary-color);border:solid 1px var(--primary-color)}.btn-secondary:focus{background-color:var(--primary-color);color:var(--text-color-primary);border:var(--primary-color);outline:0;box-shadow:none}.btn-secondary:active{background-color:var(--primary-color);box-shadow:1px 1px 5px 1px rgba(0,0,0,.5);outline:0}.btn-focused-state{background-color:var(--primary-color);color:var(--text-color-primary);border:var(--primary-color);outline:0;box-shadow:none}.checkmark{display:inline-block;-webkit-transform:rotate(45deg);transform:rotate(45deg);height:16px;width:9px;border-bottom:3px solid var(--text-color-secondary);border-right:3px solid var(--text-color-secondary)}.checkmark-checked{border-bottom:3px solid var(--text-color-primary);border-right:3px solid var(--text-color-primary)}.crossmark{position:relative;display:inline-block;margin-bottom:7px;margin-right:25px}.crossmark:after{position:absolute;content:'';display:block;width:18px;height:0;border:solid var(--text-color-secondary);border-width:0 0 3px;-webkit-transform:rotate(45deg);transform:rotate(45deg);left:0;top:0}.crossmark:before{position:absolute;content:'';display:block;width:18px;height:0;border:solid var(--text-color-secondary);border-width:0 0 3px;-webkit-transform:rotate(-45deg);transform:rotate(-45deg);left:0;top:0}.crossmark-checked:after,.crossmark-checked:before{border:solid var(--text-color-primary);border-width:0 0 3px}.nps-btn-group{display:flex;justify-content:space-between;width:75%;margin:7px 0 0}.nps-btn{width:40px;height:40px;font-size:1.5em;line-height:1.19;margin:0}.nps-btn-focus{background-color:var(--primary-color);color:var(--text-color-primary);border:var(--primary-color);outline:0;box-shadow:none}.nps-legend{margin:5px 0 0;font-size:.75em;color:var(--text-color-secondary);position:relative;width:75%}#left-nps-label{position:absolute;width:8%;left:0}#center-nps-label{position:absolute;width:8%;left:47.5%}#right-nps-label{position:absolute;width:8%;right:0;text-align:end}.multiple-choice{width:320px;min-height:32px;margin:7px 0 0;padding:5px 0 0 7px;font-size:.875em;display:block;border-radius:4px;border:1px solid var(--white-25);box-shadow:1px 1px 3px 0 rgba(0,0,0,.5);color:var(--text-color-secondary);transition:.3s;background-color:var(--white-25);cursor:pointer}.multiple-choice-checkbox{display:none}.multiple-choice-checkbox:checked+span{background-color:var(--primary-color);color:var(--text-color-primary);border:none;box-shadow:none}.survey-input-textarea{width:95%;min-height:5em;max-height:21em;font-size:.875rem;color:var(--text-color-secondary);margin:7px 0 0;padding:6px 0 0 5px;display:block;background-color:var(--white-25);border:var(--white-25);resize:vertical;font-family:var(--font-family)}.survey-input-textarea:active,.survey-input-textarea:focus{outline:1px solid var(--primary-color)}.survey-input-text-box{width:100%;height:32px;margin:7px 0 0;padding:0 0 0 5px;font-size:.875rem;color:var(--text-color-secondary);background-color:var(--white-25);border:var(--white-25);font-family:var(--font-family)}.survey-input-text-box:active,.survey-input-text-box:focus{outline:1px solid var(--primary-color)}.yes-no-btn{width:71px;height:32px;padding:0}.thumbs-btn{width:40px;height:32px;padding:3px 0 0}.rating{direction:rtl;display:block;margin:7px 0 0;float:left}.rating input[type=radio]{display:none}.rating label{width:32px;height:32px;display:inline-block;cursor:pointer;margin:0 7px 0 0;position:relative}.ratingFillIcon,.ratingOutlineIcon{fill:var(--primary-color);position:absolute}.ratingFillIcon{display:none}.rating input[type=radio]:checked~label>div>.ratingFillIcon,.rating label:hover .ratingFillIcon,.rating label:hover~label .ratingFillIcon{fill:var(--primary-color);display:block}.rating input[type=radio]:checked~label>div>.ratingFillIcon{-webkit-animation:.25s 2 blinkAnimation;animation:.25s 2 blinkAnimation}.drop-down{margin:7px 0 0;width:100%;height:32px;font-size:.75em;line-height:1.5;color:var(--text-color-secondary);padding:7px 0 7px 12px;border:1px solid var(--white-25);border-radius:4px;background-color:var(--white-25);cursor:pointer}.drop-down:hover .dropdown-content{display:block}.drop-down:focus{outline:1px solid var(--primary-color)}.agree-disagree-btn{min-width:115px;height:32px}.matrix{display:flex;justify-content:center;align-items:center}.matrix-left{display:inline-block}.matrix-left p{font-size:1.2em;line-height:1.43;color:var(--p-font-color);margin:12px 0 0}.matrix-right{margin:20px 0 0 20px;padding:0 0 10px}.matrix-star-rating{float:none}.matrix-star-rating label{width:25px;height:25px;margin:0}.matrix-star-rating input[type=radio]:checked~label>img,.matrix-star-rating label:hover,.matrix-star-rating label:hover~label{background:url(https://cdn.omoto.io/star-fill.png) no-repeat}.matrix-drop-down{width:157px;height:32px}.isRequired{color:red;margin-right:3px}.validation-error-msg{color:red;font-size:.75em}@media (min-width:481px) and (max-width:767px){.survey-container{height:100%}.survey-container-wrap{padding:20px 20px 10px;background-color:var(--bg-color);height:calc(100% - 59px);overflow-x:hidden;overflow-y:auto}.survey-statement{font-size:1.125em}.multiple-choice{width:100%;max-width:320px}.footer{height:59px;position:fixed}.footer-container{padding:0 20px;position:initial;max-width:initial;width:initial;margin:0}.footer-right img{width:60px;height:30px}.progress-container{width:10em}.footer-left{margin:5px 0 0}.btn-primary{margin:20px 0 0}.nps-btn-group{margin-top:7px;width:85%}.nps-legend{width:85%}.nps-btn{width:35px;height:35px;font-size:1.06em;line-height:2.2;margin:0;padding:0}.matrix{margin:0 auto}.matrix-left p{width:138px;height:20px;font-size:.875em}.matrix-right{padding:4px 0 0}}@media (min-width:320px) and (max-width:375px){.survey-container{height:100%}.org-logo{width:100%;height:auto}.survey-container-wrap{margin:0;padding:20px 20px 10px;height:calc(100vh - 59px);background-color:var(--bg-color);overflow-y:auto;overflow-x:hidden}.welcome-thankyou-heading{margin:1rem 0 0}.survey-statement{font-size:14px;margin:1rem 0 0}.multiple-choice{width:100%;max-width:320px;height:28px;padding:4px 0 0 7px;font-size:12px}.footer{height:59px;position:fixed}.footer-container{padding:0 20px;position:initial;max-width:initial;width:initial;margin:0}.footer-right img{width:60px;height:30px}.footer-left{margin:5px 0 0}.progress-container{width:8.3em}.btn-primary{margin:16px 0 0;font-size:14px}.nps-btn-group{margin-top:7px;width:100%}.nps-btn{width:25px;height:25px;font-size:1.06em;line-height:1.06;margin:0 3px 0 0;padding:0}#right-nps-label,.nps-legend{width:100%}.ratingIconAbove6{width:25px!important;margin:0 3px 0 0!important}.ratingIconAbove6 .ratingFillIcon,.ratingIconAbove6 .ratingOutlineIcon{width:25px}.matrix{margin:0 auto}.matrix-left p{width:138px;height:20px;font-size:.875em}.matrix-right{padding:4px 0 0}}@media (min-width:375px) and (max-width:480px){.survey-container{height:100%}.org-logo{width:100%;height:auto}.survey-container-wrap{margin:0;padding:20px 20px 10px;height:calc(100vh - 59px);background-color:var(--bg-color);overflow-y:auto;overflow-x:hidden}.survey-statement{font-size:1.125em}.multiple-choice{width:100%;max-width:320px}.footer{height:59px;position:fixed}.footer-container{padding:0 20px;position:initial;max-width:initial;width:initial;margin:0}.footer-right img{width:60px;height:30px}.footer-left{margin:5px 0 0}.progress-container{width:8.3em}.btn-primary{margin:20px 0 0}.nps-btn-group{margin-top:7px;width:100%}.nps-btn{width:25px;height:25px;font-size:1.06em;line-height:1.06;margin:0;padding:0}#right-nps-label,.nps-legend{width:100%}.ratingIconAbove6{width:25px!important;margin:0 3px 0 0!important}.ratingIconAbove6 .ratingFillIcon,.ratingIconAbove6 .ratingOutlineIcon{width:25px}.matrix{margin:0 auto}.matrix-left p{width:138px;height:20px;font-size:.875em}.matrix-right{padding:4px 0 0}}#error_snackbar{visibility:hidden;width:250px;margin-left:-125px;background-color:rgba(0,0,0,.55);color:#fff;text-align:center;border-radius:2px;padding:16px;position:absolute;z-index:1;left:50%;bottom:10%;font-size:1em}#error_snackbar.show{visibility:visible;-webkit-animation:.5s fadein,.5s 2.5s fadeout;animation:.5s fadein,.5s 2.5s fadeout}@-webkit-keyframes fadein{from{bottom:0;opacity:0}to{bottom:10%;opacity:1}}@keyframes fadein{from{bottom:0;opacity:0}to{bottom:10%;opacity:1}}@-webkit-keyframes fadeout{from{bottom:10%;opacity:1}to{bottom:0;opacity:0}}@keyframes fadeout{from{bottom:10%;opacity:1}to{bottom:0;opacity:0}}@media screen and (-webkit-min-device-pixel-ratio:0){progress{height:10px}input,select,textarea{font-size:16px}}"]
            }] }
];
/** @nocollapse */
QuestionnaireComponent.ctorParameters = () => [
    { type: PagesService },
    { type: SurveyLibService }
];
QuestionnaireComponent.propDecorators = {
    previousPageEvent: [{ type: Output }],
    currentQuestionIndex: [{ type: Input }],
    submitResponse: [{ type: Output }],
    questionnairePage: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    QuestionnaireComponent.prototype.previousPageEvent;
    /** @type {?} */
    QuestionnaireComponent.prototype.currentQuestionIndex;
    /** @type {?} */
    QuestionnaireComponent.prototype.submitResponse;
    /** @type {?} */
    QuestionnaireComponent.prototype.questionnairePage;
    /** @type {?} */
    QuestionnaireComponent.prototype.mandatoryQuestion;
    /** @type {?} */
    QuestionnaireComponent.prototype.expressionBuilder;
    /** @type {?} */
    QuestionnaireComponent.prototype.isMandatoryQuestionAnswered;
    /**
     * @type {?}
     * @private
     */
    QuestionnaireComponent.prototype.nextQuestionSubscription;
    /**
     * @type {?}
     * @private
     */
    QuestionnaireComponent.prototype.previousQuestionSubscription;
    /** @type {?} */
    QuestionnaireComponent.prototype.pagesService;
    /** @type {?} */
    QuestionnaireComponent.prototype.surveyLibService;
}
