// survey-lib/esm2015/lib/pages/welcome-page/welcome-page.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/welcome-page/welcome-page.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Page } from '../pages.model';
import { PagesService } from '../../shared/pages-services/pages.service';
export class WelcomePageComponent {
    /**
     * @param {?} pageService
     */
    constructor(pageService) {
        this.pageService = pageService;
        this.nextPageEvent = pageService.nextPageEvent;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.pageService.stage != 'entry-next') {
            document.getElementById("welcome-page").className = 'entry-next';
            document.getElementById("welcome-page-delay").className = 'entry-next-delay';
        }
        else {
            document.getElementById("welcome-page").className = 'entry-previous-delay';
            document.getElementById("welcome-page-delay").className = 'entry-previous';
        }
    }
    /**
     * @return {?}
     */
    nextPage() {
        //this.pageService.store.dispatch(new Reset());
        document.getElementById("btn-primary").classList.add('blink-btn-animation');
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.nextPageEvent.emit();
        }), 350);
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
    }
}
WelcomePageComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-welcome-page',
                template: "<div id=\"welcome-page\">\n  <div class=\"welcome-thankyou-heading\">{{welcomePage.properties.headingValue}}</div>\n  <div class=\"survey-statement welcome-thankyou-text\"> {{ welcomePage.properties.subHeadingValue }} </div>\n</div>\n\n<div id=\"welcome-page-delay\" >\n<button class=\"btn-primary\" id=\"btn-primary\"  (click)=\"nextPage()\">{{ welcomePage.properties.buttonText}}</button>\n</div>\n\n\n\n",
                styles: [""]
            }] }
];
/** @nocollapse */
WelcomePageComponent.ctorParameters = () => [
    { type: PagesService }
];
WelcomePageComponent.propDecorators = {
    welcomePage: [{ type: Input }],
    nextPageEvent: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    WelcomePageComponent.prototype.welcomePage;
    /** @type {?} */
    WelcomePageComponent.prototype.nextPageEvent;
    /** @type {?} */
    WelcomePageComponent.prototype.showPrev;
    /**
     * @type {?}
     * @private
     */
    WelcomePageComponent.prototype.pageService;
}
