// survey-lib/esm2015/lib/pages/questionnaire/name/name.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/questionnaire/name/name.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Question } from '../../pages.model';
import { SurveyResponses } from '../../survey-response.model';
import { PagesService } from '../../../shared/pages-services/pages.service';
export class NameComponent {
    /**
     * @param {?} pageService
     */
    constructor(pageService) {
        this.pageService = pageService;
        this.submitEvent = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.pageService.stage == 'entry-next') {
            document.getElementById('name-question').className = 'entry-next';
            document.getElementById('name-question-delay').className = 'entry-next-delay';
        }
        else if (this.pageService.stage == 'entry-previous') {
            document.getElementById('name-question').className = 'entry-previous-delay';
            document.getElementById('name-question-delay').className = 'entry-previous';
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    onSubmit(value) {
        if (value != undefined) {
            this.surveyResponse.questionId = this.name.questionId;
            this.surveyResponse.question = this.name.question;
            this.surveyResponse.questionType = this.name.questionType;
            this.surveyResponse.identifier = this.name.identifier;
            this.surveyResponse.answer = value;
            this.pageService.persistSurvey(this.surveyResponse);
            document.getElementById('btn-primary').classList.add('blink-btn-animation');
            setTimeout((/**
             * @return {?}
             */
            () => {
                this.submitEvent.emit();
            }), 350);
        }
        else {
            this.submitEvent.emit();
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
    }
}
NameComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-name',
                template: "<div id=\"name-question\">\n<div class=\"survey-statement\">\n  <span *ngIf=\"name.attributes.required\" class=\"isRequired\">*</span>{{name.question}}\n</div>\n<p class=\"question-description\">\n  {{name.attributes.description}}\n</p>\n</div>\n\n\n\n<div id=\"name-question-delay\">\n<input class=\"survey-input-text-box\"\n  #nameInput\n  type=\"text\" name=\"name\"\n  [(ngModel)]=\"surveyResponse.answer\" onfocus=\"this.placeholder=''\"\n  placeholder=\"Enter first name followed by last name\">\n<button class=\"btn-primary\" id=\"btn-primary\" type=\"submit\" name=\"name\" (click)=\"onSubmit(nameInput.value)\">Done</button>\n<span\n  *ngIf=\"!name.attributes.required\"\n  (click)=\"onSubmit(undefined)\"\n  class=\"skip-this-question\">Skip</span>\n</div>\n",
                styles: [""]
            }] }
];
/** @nocollapse */
NameComponent.ctorParameters = () => [
    { type: PagesService }
];
NameComponent.propDecorators = {
    name: [{ type: Input }],
    surveyResponse: [{ type: Input }],
    submitEvent: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    NameComponent.prototype.name;
    /** @type {?} */
    NameComponent.prototype.surveyResponse;
    /** @type {?} */
    NameComponent.prototype.submitEvent;
    /**
     * @type {?}
     * @private
     */
    NameComponent.prototype.pageService;
}
