// survey-lib/esm5/lib/shared/pages-services/pages.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/shared/pages-services/pages.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Injectable } from '@angular/core';
import { SurveyResponses } from '../../pages/survey-response.model';
var PagesService = /** @class */ (function () {
    function PagesService() {
        this.previousQuestion = new EventEmitter();
        this.nextQuestion = new EventEmitter();
        this.nextPageEvent = new EventEmitter();
        this.surveyResponses = [];
        this.sequenceOfQuestionTraversed = [];
    }
    /**
     * @param {?} response
     * @return {?}
     */
    PagesService.prototype.persistSurvey = /**
     * @param {?} response
     * @return {?}
     */
    function (response) {
        /** @type {?} */
        var index = this.surveyResponses.findIndex((/**
         * @param {?} surveyResponse
         * @return {?}
         */
        function (surveyResponse) { return surveyResponse.identifier === response.identifier; }));
        if (index !== -1) {
            this.surveyResponses[index] = response;
        }
        else {
            this.surveyResponses.push(response);
        }
    };
    /**
     * @param {?} payload
     * @return {?}
     */
    PagesService.prototype.removeResponsesOfExcludedQuestionIds = /**
     * @param {?} payload
     * @return {?}
     */
    function (payload) {
        return this.surveyResponses.filter((/**
         * @param {?} response
         * @return {?}
         */
        function (response) { return payload.find((/**
         * @param {?} questionId
         * @return {?}
         */
        function (questionId) { return questionId == response.questionId; })); }));
    };
    /**
     * @param {?} identifier
     * @return {?}
     */
    PagesService.prototype.getResponse = /**
     * @param {?} identifier
     * @return {?}
     */
    function (identifier) {
        console.log(this.surveyResponses);
        /** @type {?} */
        var index = this.surveyResponses.findIndex((/**
         * @param {?} res
         * @return {?}
         */
        function (res) { return res.identifier == identifier; }));
        if (index > -1) {
            return this.surveyResponses[index];
        }
        else {
            return new SurveyResponses();
        }
    };
    PagesService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    PagesService.ctorParameters = function () { return []; };
    return PagesService;
}());
export { PagesService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PagesService.prototype.surveyUrl;
    /**
     * @type {?}
     * @private
     */
    PagesService.prototype.submitResponseUrl;
    /** @type {?} */
    PagesService.prototype.previousQuestion;
    /** @type {?} */
    PagesService.prototype.nextQuestion;
    /** @type {?} */
    PagesService.prototype.nextPageEvent;
    /** @type {?} */
    PagesService.prototype.surveyResponses;
    /** @type {?} */
    PagesService.prototype.sequenceOfQuestionTraversed;
    /** @type {?} */
    PagesService.prototype.stage;
}
