// survey-lib/esm5/lib/pages/questionnaire/statement/statement.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/questionnaire/statement/statement.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Question } from '../../pages.model';
import { SurveyResponses } from '../../survey-response.model';
import { PagesService } from '../../../shared/pages-services/pages.service';
var StatementComponent = /** @class */ (function () {
    function StatementComponent(pageService) {
        this.pageService = pageService;
        this.submitEvent = new EventEmitter();
    }
    /**
     * @return {?}
     */
    StatementComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.pageService.stage == 'entry-next') {
            document.getElementById('statement-question').className = 'entry-next';
            document.getElementById('statement-question-delay').className = 'entry-next-delay';
        }
        else if (this.pageService.stage == 'entry-previous') {
            document.getElementById('statement-question').className = 'entry-previous-delay';
            document.getElementById('statement-question-delay').className = 'entry-previous';
        }
    };
    /**
     * @return {?}
     */
    StatementComponent.prototype.onSubmit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.surveyResponse.answer = 'Ok';
        this.surveyResponse.identifier = this.statement.identifier;
        this.surveyResponse.questionId = this.statement.questionId;
        this.surveyResponse.questionType = this.statement.questionType;
        this.pageService.persistSurvey(this.surveyResponse);
        document.getElementById('btn-primary').classList.add('blink-btn-animation');
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.submitEvent.emit();
        }), 350);
    };
    /**
     * @return {?}
     */
    StatementComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.submitEvent.unsubscribe();
    };
    /**
     * @param {?} value
     * @return {?}
     */
    StatementComponent.prototype.onSkip = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value != undefined) {
            this.surveyResponse.answer = value;
            this.surveyResponse.identifier = this.statement.identifier;
            this.surveyResponse.questionId = this.statement.questionId;
            this.surveyResponse.questionType = this.statement.questionType;
            this.pageService.persistSurvey(this.surveyResponse);
        }
        this.submitEvent.emit();
    };
    StatementComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-statement',
                    template: "<div id=\"statement-question\">\n<div class=\"survey-statement\"> <span *ngIf=\"statement.attributes.required\" class=\"isRequired\">*</span>{{ statement.question }} </div>\n<p class=\"question-description\">\n  {{statement.attributes.description}}\n</p>\n\n</div>\n\n\n\n<div id=\"statement-question-delay\">\n<button class=\"btn-primary\"\n        id=\"btn-primary\"\n        type=\"submit\"\n        *ngIf=\"surveyResponse.answer != 'Ok'\"\n        name=\"name\"\n        (click)=\"onSubmit()\">Done</button>\n<button class=\"btn-primary\"\n        type=\"submit\"\n        id=\"btn-primary\"\n        name=\"name\"\n        appAutofocus\n        *ngIf=\"surveyResponse.answer == 'Ok'\"\n        (click)=\"onSubmit()\">Done</button>\n<span\n  (click)=\"onSkip(undefined)\"\n  *ngIf=\"!statement.attributes.required\" class=\"skip-this-question skip-this-question-marginLeft\">Skip</span>\n</div>\n",
                    styles: [".statement-button{position:absolute;left:0;margin-left:1%}"]
                }] }
    ];
    /** @nocollapse */
    StatementComponent.ctorParameters = function () { return [
        { type: PagesService }
    ]; };
    StatementComponent.propDecorators = {
        statement: [{ type: Input }],
        surveyResponse: [{ type: Input }],
        submitEvent: [{ type: Output }]
    };
    return StatementComponent;
}());
export { StatementComponent };
if (false) {
    /** @type {?} */
    StatementComponent.prototype.statement;
    /** @type {?} */
    StatementComponent.prototype.surveyResponse;
    /** @type {?} */
    StatementComponent.prototype.submitEvent;
    /**
     * @type {?}
     * @private
     */
    StatementComponent.prototype.pageService;
}
