// survey-lib/esm5/lib/pages/questionnaire/mandatory-question-validation.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/questionnaire/mandatory-question-validation.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var MandatoryQuestionValidation = /** @class */ (function () {
    function MandatoryQuestionValidation() {
    }
    /**
     * @param {?} questionType
     * @param {?} response
     * @return {?}
     */
    MandatoryQuestionValidation.prototype.isQuestionAnswered = /**
     * @param {?} questionType
     * @param {?} response
     * @return {?}
     */
    function (questionType, response) {
        console.log('Inside isQuestionAnswered!');
        console.log('Question type: ' + questionType);
        if (response == undefined)
            return false;
        switch (questionType) {
            case 'Address':
                return this.isAddressFormAnswered(response);
            case 'Choice':
                return this.isChoiceQuestionAnswered(response);
            case 'Contact form':
                return this.isContactFormAnswered(response);
            case 'Date':
                return this.isDateQuestionAnswered(response);
            case 'Drop Down':
                return this.isDropDownQuestionAnswered(response);
            case 'Email':
                return this.isEmailQuestionAnswered(response);
            case 'Short text':
            case 'Long text':
                return this.isTextQuestionAnswered(response);
            case 'Nps':
            case 'Rating':
                return this.isRatingQuestionAnswered(response);
            case 'Statement':
                return this.isStatementQuestionAnswered(response);
            case 'Yes No':
            case 'Legal':
                return this.isYesNoQuestionAnswered(response);
            default: return false;
        }
    };
    /**
     * @private
     * @param {?} response
     * @return {?}
     */
    MandatoryQuestionValidation.prototype.isAddressFormAnswered = /**
     * @private
     * @param {?} response
     * @return {?}
     */
    function (response) {
        return false;
    };
    /**
     * @private
     * @param {?} response
     * @return {?}
     */
    MandatoryQuestionValidation.prototype.isContactFormAnswered = /**
     * @private
     * @param {?} response
     * @return {?}
     */
    function (response) {
        return false;
    };
    /**
     * @private
     * @param {?} response
     * @return {?}
     */
    MandatoryQuestionValidation.prototype.isChoiceQuestionAnswered = /**
     * @private
     * @param {?} response
     * @return {?}
     */
    function (response) {
        return response.choices != undefined && response.choices.length > 0;
    };
    /**
     * @private
     * @param {?} response
     * @return {?}
     */
    MandatoryQuestionValidation.prototype.isDateQuestionAnswered = /**
     * @private
     * @param {?} response
     * @return {?}
     */
    function (response) {
        return response.answer != undefined && response.answer.length > 0;
    };
    /**
     * @private
     * @param {?} response
     * @return {?}
     */
    MandatoryQuestionValidation.prototype.isDropDownQuestionAnswered = /**
     * @private
     * @param {?} response
     * @return {?}
     */
    function (response) {
        return response.answerId != undefined;
    };
    /**
     * @private
     * @param {?} response
     * @return {?}
     */
    MandatoryQuestionValidation.prototype.isEmailQuestionAnswered = /**
     * @private
     * @param {?} response
     * @return {?}
     */
    function (response) {
        return response.answer != undefined && response.answer.length > 0;
    };
    /**
     * @private
     * @param {?} response
     * @return {?}
     */
    MandatoryQuestionValidation.prototype.isTextQuestionAnswered = /**
     * @private
     * @param {?} response
     * @return {?}
     */
    function (response) {
        return response.answer != undefined && response.answer.length > 0;
    };
    /**
     * @private
     * @param {?} response
     * @return {?}
     */
    MandatoryQuestionValidation.prototype.isRatingQuestionAnswered = /**
     * @private
     * @param {?} response
     * @return {?}
     */
    function (response) {
        return response.rating != undefined && response.rating > -1;
    };
    /**
     * @private
     * @param {?} response
     * @return {?}
     */
    MandatoryQuestionValidation.prototype.isStatementQuestionAnswered = /**
     * @private
     * @param {?} response
     * @return {?}
     */
    function (response) {
        return false;
    };
    /**
     * @private
     * @param {?} response
     * @return {?}
     */
    MandatoryQuestionValidation.prototype.isYesNoQuestionAnswered = /**
     * @private
     * @param {?} response
     * @return {?}
     */
    function (response) {
        return response.answer != undefined && response.answer.length > 0;
    };
    return MandatoryQuestionValidation;
}());
export { MandatoryQuestionValidation };
