// survey-lib/esm2015/lib/pages/pages.model.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/pages.model.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class PagesModel {
    /**
     * @return {?}
     */
    get testMode() {
        return this._testMode;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set testMode(value) {
        this._testMode = value;
    }
    /**
     * @return {?}
     */
    get id() {
        return this._id;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set id(value) {
        this._id = value;
    }
    /**
     * @return {?}
     */
    get userId() {
        return this._userId;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set userId(value) {
        this._userId = value;
    }
    /**
     * @return {?}
     */
    get themeId() {
        return this._themeId;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set themeId(value) {
        this._themeId = value;
    }
    /**
     * @return {?}
     */
    get logoUrl() {
        return this._logoUrl;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set logoUrl(value) {
        this._logoUrl = value;
    }
    /**
     * @return {?}
     */
    get surveyContent() {
        return this._surveyContent;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set surveyContent(value) {
        this._surveyContent = value;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    PagesModel.prototype._id;
    /**
     * @type {?}
     * @private
     */
    PagesModel.prototype._userId;
    /**
     * @type {?}
     * @private
     */
    PagesModel.prototype._themeId;
    /**
     * @type {?}
     * @private
     */
    PagesModel.prototype._testMode;
    /**
     * @type {?}
     * @private
     */
    PagesModel.prototype._logoUrl;
    /**
     * @type {?}
     * @private
     */
    PagesModel.prototype._surveyContent;
}
export class SurveyContent {
    /**
     * @return {?}
     */
    get name() {
        return this._name;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @return {?}
     */
    get description() {
        return this._description;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @return {?}
     */
    get pages() {
        return this._pages;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set pages(value) {
        this._pages = value;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    SurveyContent.prototype._name;
    /**
     * @type {?}
     * @private
     */
    SurveyContent.prototype._description;
    /**
     * @type {?}
     * @private
     */
    SurveyContent.prototype._pages;
    /** @type {?} */
    SurveyContent.prototype.selectedTheme;
    /** @type {?} */
    SurveyContent.prototype.organizationLogo;
}
export class SelectedThemeModel {
}
if (false) {
    /** @type {?} */
    SelectedThemeModel.prototype.name;
    /** @type {?} */
    SelectedThemeModel.prototype.font;
    /** @type {?} */
    SelectedThemeModel.prototype.colors;
}
export class Font {
}
if (false) {
    /** @type {?} */
    Font.prototype.name;
    /** @type {?} */
    Font.prototype.url;
}
export class Colors {
}
if (false) {
    /** @type {?} */
    Colors.prototype.backgroundColor;
    /** @type {?} */
    Colors.prototype.primaryColor;
    /** @type {?} */
    Colors.prototype.primaryColorDark;
    /** @type {?} */
    Colors.prototype.footerBackgroundColor;
    /** @type {?} */
    Colors.prototype.footerSecondaryColor;
    /** @type {?} */
    Colors.prototype.footerPrimaryColor;
    /** @type {?} */
    Colors.prototype.textColorPrimary;
    /** @type {?} */
    Colors.prototype.textColorSecondary;
    /** @type {?} */
    Colors.prototype.primaryButtonTextColor;
}
export class Page {
    /**
     * @return {?}
     */
    get pageName() {
        return this._pageName;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set pageName(value) {
        this._pageName = value;
    }
    /**
     * @return {?}
     */
    get properties() {
        return this._properties;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set properties(value) {
        this._properties = value;
    }
    /**
     * @return {?}
     */
    get questions() {
        return this._questions;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set questions(value) {
        this._questions = value;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    Page.prototype._pageName;
    /**
     * @type {?}
     * @private
     */
    Page.prototype._properties;
    /**
     * @type {?}
     * @private
     */
    Page.prototype._questions;
}
export class Question {
}
if (false) {
    /** @type {?} */
    Question.prototype.identifier;
    /** @type {?} */
    Question.prototype.questionId;
    /** @type {?} */
    Question.prototype.questionType;
    /** @type {?} */
    Question.prototype.question;
    /** @type {?} */
    Question.prototype.choices;
    /** @type {?} */
    Question.prototype.others;
    /** @type {?} */
    Question.prototype.attributes;
    /** @type {?} */
    Question.prototype.pictures;
    /** @type {?} */
    Question.prototype.rankingItems;
    /** @type {?} */
    Question.prototype.rows;
    /** @type {?} */
    Question.prototype.columns;
    /** @type {?} */
    Question.prototype.contactForm;
    /** @type {?} */
    Question.prototype.addressForm;
    /** @type {?} */
    Question.prototype.logicJumps;
}
export class Others {
}
if (false) {
    /** @type {?} */
    Others.prototype.value;
    /** @type {?} */
    Others.prototype.isChecked;
    /** @type {?} */
    Others.prototype.showCheckbox;
}
export class LogicJumps {
}
if (false) {
    /** @type {?} */
    LogicJumps.prototype.blocks;
    /** @type {?} */
    LogicJumps.prototype.defaultJump;
}
export class Block {
}
if (false) {
    /** @type {?} */
    Block.prototype.that;
    /** @type {?} */
    Block.prototype.thatIndex;
    /** @type {?} */
    Block.prototype.conditions;
}
export class Condition {
}
if (false) {
    /** @type {?} */
    Condition.prototype.condition;
    /** @type {?} */
    Condition.prototype._if;
    /** @type {?} */
    Condition.prototype._this;
    /** @type {?} */
    Condition.prototype.relationWithNextCondition;
}
export class RankingItem {
}
if (false) {
    /** @type {?} */
    RankingItem.prototype.itemId;
    /** @type {?} */
    RankingItem.prototype.text;
}
export class ContactForm {
}
if (false) {
    /** @type {?} */
    ContactForm.prototype.label;
    /** @type {?} */
    ContactForm.prototype.type;
    /** @type {?} */
    ContactForm.prototype.value;
    /** @type {?} */
    ContactForm.prototype.validation;
    /** @type {?} */
    ContactForm.prototype.keyName;
}
export class Validation {
}
if (false) {
    /** @type {?} */
    Validation.prototype.required;
}
export class Column {
}
if (false) {
    /** @type {?} */
    Column.prototype.columnId;
    /** @type {?} */
    Column.prototype.label;
    /** @type {?} */
    Column.prototype.value;
}
export class Picture {
    /**
     * @return {?}
     */
    get id() {
        return this._id;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set id(value) {
        this._id = value;
    }
    /**
     * @return {?}
     */
    get picture() {
        return this._picture;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set picture(value) {
        this._picture = value;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    Picture.prototype._id;
    /**
     * @type {?}
     * @private
     */
    Picture.prototype._picture;
}
export class AddressForm {
    /**
     * @return {?}
     */
    get text() {
        return this._text;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set text(value) {
        this._text = value;
    }
    /**
     * @return {?}
     */
    get required() {
        return this._required;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set required(value) {
        this._required = value;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    AddressForm.prototype._text;
    /**
     * @type {?}
     * @private
     */
    AddressForm.prototype._required;
}
export class Row {
    /**
     * @return {?}
     */
    get id() {
        return this._id;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set id(value) {
        this._id = value;
    }
    /**
     * @return {?}
     */
    get text() {
        return this._text;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set text(value) {
        this._text = value;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    Row.prototype._id;
    /**
     * @type {?}
     * @private
     */
    Row.prototype._text;
}
export class Attributes {
}
if (false) {
    /** @type {?} */
    Attributes.prototype.image;
    /** @type {?} */
    Attributes.prototype.description;
    /** @type {?} */
    Attributes.prototype.vertical;
    /** @type {?} */
    Attributes.prototype.video;
    /** @type {?} */
    Attributes.prototype.multipleChoice;
    /** @type {?} */
    Attributes.prototype.randomize;
    /** @type {?} */
    Attributes.prototype.required;
    /** @type {?} */
    Attributes.prototype.requiredErrorMessage;
    /** @type {?} */
    Attributes.prototype.others;
    /** @type {?} */
    Attributes.prototype.maxCharacter;
    /** @type {?} */
    Attributes.prototype.quotation;
    /** @type {?} */
    Attributes.prototype.minimumNumber;
    /** @type {?} */
    Attributes.prototype.maximumNumber;
    /** @type {?} */
    Attributes.prototype.numberOfSelection;
    /** @type {?} */
    Attributes.prototype.defaultCountryCode;
    /** @type {?} */
    Attributes.prototype.alphabeticalSort;
    /** @type {?} */
    Attributes.prototype.includeAddressForm;
    /** @type {?} */
    Attributes.prototype.startScaleAt;
    /** @type {?} */
    Attributes.prototype.leftLabel;
    /** @type {?} */
    Attributes.prototype.centerLabel;
    /** @type {?} */
    Attributes.prototype.rightLabel;
    /** @type {?} */
    Attributes.prototype.isDescriptionEnabled;
    /** @type {?} */
    Attributes.prototype.steps;
    /** @type {?} */
    Attributes.prototype.ratingIcon;
    /** @type {?} */
    Attributes.prototype.selection;
    /** @type {?} */
    Attributes.prototype.maxCharacterCount;
    /** @type {?} */
    Attributes.prototype.minCharacterCount;
}
export class Selection {
}
if (false) {
    /** @type {?} */
    Selection.prototype.type;
    /** @type {?} */
    Selection.prototype.exact;
    /** @type {?} */
    Selection.prototype.rangeFrom;
    /** @type {?} */
    Selection.prototype.rangeTo;
}
export class Step {
}
if (false) {
    /** @type {?} */
    Step.prototype.label;
    /** @type {?} */
    Step.prototype.value;
}
export class Choice {
}
if (false) {
    /** @type {?} */
    Choice.prototype.identifier;
    /** @type {?} */
    Choice.prototype.choiceId;
    /** @type {?} */
    Choice.prototype.text;
    /** @type {?} */
    Choice.prototype.isChecked;
}
export class Properties {
    /**
     * @return {?}
     */
    get buttonText() {
        return this._buttonText;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set buttonText(value) {
        this._buttonText = value;
    }
    /**
     * @return {?}
     */
    get subHeadingValue() {
        return this._subHeadingValue;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set subHeadingValue(value) {
        this._subHeadingValue = value;
    }
    /**
     * @return {?}
     */
    get headingValue() {
        return this._headingValue;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set headingValue(value) {
        this._headingValue = value;
    }
    /**
     * @return {?}
     */
    get backgroundImage() {
        return this._backgroundImage;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set backgroundImage(value) {
        this._backgroundImage = value;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    Properties.prototype._headingValue;
    /**
     * @type {?}
     * @private
     */
    Properties.prototype._subHeadingValue;
    /**
     * @type {?}
     * @private
     */
    Properties.prototype._backgroundImage;
    /**
     * @type {?}
     * @private
     */
    Properties.prototype._buttonText;
}
