// survey-lib/esm2015/lib/pages/questionnaire/drop-down/drop-down.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/questionnaire/drop-down/drop-down.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Question } from '../../pages.model';
import { SurveyResponses } from '../../survey-response.model';
import { PagesService } from '../../../shared/pages-services/pages.service';
export class DropDownComponent {
    /**
     * @param {?} pageService
     */
    constructor(pageService) {
        this.pageService = pageService;
        this.submitEvent = new EventEmitter();
    }
    /**
     * @param {?} text
     * @return {?}
     */
    onSelect(text) {
        if (text != undefined) {
            this.surveyResponse.question = this.dropDown.question;
            this.surveyResponse.questionType = this.dropDown.questionType;
            this.surveyResponse.questionId = this.dropDown.questionId;
            this.surveyResponse.answerId = this.getChoiceIdByText(text);
            this.surveyResponse.identifier = this.dropDown.identifier;
            this.surveyResponse.answer = text;
            this.pageService.persistSurvey(this.surveyResponse);
            document.getElementById('input').classList.add('blink-btn-animation');
            setTimeout((/**
             * @return {?}
             */
            () => {
                this.submitEvent.emit();
            }), 350);
        }
        else {
            this.submitEvent.emit();
        }
    }
    /**
     * @private
     * @param {?} choiceId
     * @return {?}
     */
    getChoiceById(choiceId) {
        /** @type {?} */
        let index = this.dropDown.choices.findIndex((/**
         * @param {?} choice
         * @return {?}
         */
        choice => choice.choiceId == choiceId));
        return this.dropDown.choices[index].text;
    }
    /**
     * @private
     * @param {?} text
     * @return {?}
     */
    getChoiceIdByText(text) {
        /** @type {?} */
        let index = this.dropDown.choices.findIndex((/**
         * @param {?} choice
         * @return {?}
         */
        choice => choice.text == text));
        return this.dropDown.choices[index].choiceId;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.pageService.stage == 'entry-next') {
            document.getElementById('drop-down-question').className = 'entry-next';
            document.getElementById('drop-down-question-delay').className = 'entry-next-delay';
        }
        else if (this.pageService.stage == 'entry-previous') {
            document.getElementById('drop-down-question').className = 'entry-previous-delay';
            document.getElementById('drop-down-question-delay').className = 'entry-previous';
        }
    }
    /**
     * @return {?}
     */
    toggleCustomDropDown() {
        document.getElementById('select-options').classList.toggle('visible');
        document.getElementById('skipBtn').classList.toggle('invisible');
    }
    /**
     * @param {?} text
     * @param {?} i
     * @return {?}
     */
    onDropDownSelection(text, i) {
        /** @type {?} */
        let selectedOptionsEl = document.getElementsByTagName('li');
        selectedOptionsEl[i].classList.add('selected');
        for (let j = 0; j < selectedOptionsEl.length; j++) {
            if (i == j) {
                selectedOptionsEl[j].classList.add('selected');
            }
            else {
                selectedOptionsEl[j].classList.remove('selected');
            }
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
    }
}
DropDownComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-drop-down',
                template: "<div id=\"drop-down-question\">\n<div class=\"survey-statement\">\n    <span *ngIf=\"dropDown.attributes.required\" class=\"isRequired\">*</span>{{dropDown.question}}\n</div>\n<p class=\"question-description\">\n    {{dropDown.attributes.description}}\n  </p>\n\n</div>\n\n\n<div id=\"drop-down-question-delay\">\n  <div>\n  <div style=\"display: inline-block\" class=\"select\" (click)=\"toggleCustomDropDown()\">\n    <div class=\"input\" id=\"input\">{{surveyResponse.answer == undefined ? 'Please select' : surveyResponse.answer }}</div>\n    <div class=\"select-options\" id=\"select-options\">\n      <ul>\n        <li (click)=\"onSelect(choice.text); onDropDownSelection(choice.text, index)\" *ngFor=\"let choice of dropDown.choices; let index = index\"\n            [ngClass]=\"surveyResponse.answer == choice.text ? 'selected':''\"\n        >\n          <span>{{choice.text}}</span>\n        </li>\n      </ul>\n    </div>\n  </div>\n\n  <span\n    (click)=\"onSelect(undefined)\"\n    *ngIf=\"!dropDown.attributes.required\" id=\"skipBtn\" class=\"skip-this-question skip-this-question-marginLeft\">Skip</span>\n</div>\n\n</div>\n",
                styles: [".invisible{display:none}"]
            }] }
];
/** @nocollapse */
DropDownComponent.ctorParameters = () => [
    { type: PagesService }
];
DropDownComponent.propDecorators = {
    dropDown: [{ type: Input }],
    surveyResponse: [{ type: Input }],
    submitEvent: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    DropDownComponent.prototype.dropDown;
    /** @type {?} */
    DropDownComponent.prototype.surveyResponse;
    /** @type {?} */
    DropDownComponent.prototype.submitEvent;
    /**
     * @type {?}
     * @private
     */
    DropDownComponent.prototype.pageService;
}
