// survey-lib/esm5/lib/pages/questionnaire/email/email.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/questionnaire/email/email.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Question } from '../../pages.model';
import { SurveyResponses } from '../../survey-response.model';
import { PagesService } from '../../../shared/pages-services/pages.service';
var EmailComponent = /** @class */ (function () {
    function EmailComponent(pageService) {
        this.pageService = pageService;
        this.EMAIL_REGEXP = new RegExp('^(([^<>()\\[\\]\\\\.,;:\\s@"]+(\\.[^<>()\\[\\]\\\\.,;:\\s@"]+)*)|(".+"))@((\\[[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\])|(([a-zA-Z\\-0-9]+\\.)+[a-zA-Z]{2,}))$');
        this.submitEvent = new EventEmitter();
        this.isValidEmail = true;
    }
    /**
     * @return {?}
     */
    EmailComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.pageService.stage == 'entry-next') {
            document.getElementById('email-question').className = 'entry-next';
            document.getElementById('email-question-delay').className = 'entry-next-delay';
        }
        else if (this.pageService.stage == 'entry-previous') {
            document.getElementById('email-question').className = 'entry-previous-delay';
            document.getElementById('email-question-delay').className = 'entry-previous';
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    EmailComponent.prototype.onSubmit = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        var _this = this;
        console.log(value);
        if (value != undefined) {
            this.isValidEmail = this.isValidEmailId(value);
            console.log('Is valid email: ' + this.isValidEmail);
            if (this.isValidEmail) {
                this.surveyResponse.questionId = this.email.questionId;
                this.surveyResponse.question = this.email.question;
                this.surveyResponse.questionType = this.email.questionType;
                this.surveyResponse.identifier = this.email.identifier;
                this.surveyResponse.answer = value;
                this.pageService.persistSurvey(this.surveyResponse);
                document.getElementById('btn-primary').classList.add('blink-btn-animation');
                setTimeout((/**
                 * @return {?}
                 */
                function () {
                    _this.submitEvent.emit();
                }), 350);
            }
        }
        else {
            this.submitEvent.emit();
        }
    };
    /**
     * @return {?}
     */
    EmailComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @private
     * @param {?} email
     * @return {?}
     */
    EmailComponent.prototype.isValidEmailId = /**
     * @private
     * @param {?} email
     * @return {?}
     */
    function (email) {
        return this.EMAIL_REGEXP.test(email);
    };
    EmailComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-email',
                    template: "<div id=\"email-question\">\n<div class=\"survey-statement\">\n    <span *ngIf=\"email.attributes.required\" class=\"isRequired\">*</span>{{ email.question }}\n</div>\n<p class=\"question-description\">\n    {{email.attributes.description}}\n  </p>\n\n</div>\n\n\n<div id=\"email-question-delay\">\n<input class=\"survey-input-text-box\"\n  #emailInput\n  type=\"email\"\n  name=\"email\"\n  onfocus=\"this.placeholder=''\"\n  placeholder=\"Enter valid email ID\"\n  [(ngModel)]=\"surveyResponse.answer\">\n  <div *ngIf=\"!isValidEmail\" class=\"validation-error-msg\">Please enter valid email</div>\n<button class=\"btn-primary\" id=\"btn-primary\"  (click)=\"onSubmit(emailInput.value)\">{{isLastQuestion?'Submit': 'Done'}}</button>\n<span (click)=\"onSubmit(undefined)\"\n  *ngIf=\"!email.attributes.required\" class=\"skip-this-question skip-this-question-marginLeft\">Skip</span>\n</div>\n",
                    styles: [".email-text-box{position:absolute;left:0;margin-left:1%}"]
                }] }
    ];
    /** @nocollapse */
    EmailComponent.ctorParameters = function () { return [
        { type: PagesService }
    ]; };
    EmailComponent.propDecorators = {
        email: [{ type: Input }],
        submitEvent: [{ type: Output }],
        surveyResponse: [{ type: Input }],
        isLastQuestion: [{ type: Input }]
    };
    return EmailComponent;
}());
export { EmailComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    EmailComponent.prototype.EMAIL_REGEXP;
    /** @type {?} */
    EmailComponent.prototype.email;
    /** @type {?} */
    EmailComponent.prototype.submitEvent;
    /** @type {?} */
    EmailComponent.prototype.surveyResponse;
    /** @type {?} */
    EmailComponent.prototype.isLastQuestion;
    /** @type {?} */
    EmailComponent.prototype.isValidEmail;
    /**
     * @type {?}
     * @private
     */
    EmailComponent.prototype.pageService;
}
