// survey-lib/esm5/lib/pages/store/survey-actions.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/store/survey-actions.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var CURRENT_STATE = 'current_state';
/** @type {?} */
export var SAVE_CHOICE = 'save_choice';
/** @type {?} */
export var SAVE_EMAIL = 'save_email';
/** @type {?} */
export var SAVE_AGREE_DISAGREE = 'save_agree_disagree';
/** @type {?} */
export var SAVE_PHONE = 'save_phone';
/** @type {?} */
export var SAVE_ADDRESS = 'save_address';
/** @type {?} */
export var SAVE_DATE = 'save_date';
/** @type {?} */
export var SAVE_NPS = 'save_nps';
/** @type {?} */
export var SAVE_YES_NO = 'save_yes_no';
/** @type {?} */
export var SAVE_RATING = 'save_rating';
/** @type {?} */
export var SAVE_DROP_DOWN = 'save_rating';
/** @type {?} */
export var SAVE_LONG_TEXT = 'save_long_text';
/** @type {?} */
export var SAVE_SHORT_TEXT = 'save_short_text';
/** @type {?} */
export var RESET_RESPONSE = 'reset_response';
/** @type {?} */
export var REMOVE_RESPONSES = 'remove_responses';
/** @type {?} */
export var REMOVE_RESPONSES_OF_EXCLUDED_QUESTION_IDS = 'remove_responses_of_excluded_question_ids';
var Email = /** @class */ (function () {
    function Email(payload) {
        this.payload = payload;
        this.type = SAVE_EMAIL;
    }
    return Email;
}());
export { Email };
if (false) {
    /** @type {?} */
    Email.prototype.type;
    /** @type {?} */
    Email.prototype.payload;
}
var Phone = /** @class */ (function () {
    function Phone(payload) {
        this.payload = payload;
        this.type = SAVE_PHONE;
    }
    return Phone;
}());
export { Phone };
if (false) {
    /** @type {?} */
    Phone.prototype.type;
    /** @type {?} */
    Phone.prototype.payload;
}
var AgreeDisagree = /** @class */ (function () {
    function AgreeDisagree(payload) {
        this.payload = payload;
        this.type = SAVE_AGREE_DISAGREE;
    }
    return AgreeDisagree;
}());
export { AgreeDisagree };
if (false) {
    /** @type {?} */
    AgreeDisagree.prototype.type;
    /** @type {?} */
    AgreeDisagree.prototype.payload;
}
var Choice = /** @class */ (function () {
    function Choice(payload) {
        this.payload = payload;
        this.type = SAVE_CHOICE;
    }
    return Choice;
}());
export { Choice };
if (false) {
    /** @type {?} */
    Choice.prototype.type;
    /** @type {?} */
    Choice.prototype.payload;
}
var Address = /** @class */ (function () {
    function Address(payload) {
        this.payload = payload;
        this.type = SAVE_ADDRESS;
    }
    return Address;
}());
export { Address };
if (false) {
    /** @type {?} */
    Address.prototype.type;
    /** @type {?} */
    Address.prototype.payload;
}
var Date = /** @class */ (function () {
    function Date(payload) {
        this.payload = payload;
        this.type = SAVE_DATE;
    }
    return Date;
}());
export { Date };
if (false) {
    /** @type {?} */
    Date.prototype.type;
    /** @type {?} */
    Date.prototype.payload;
}
var Nps = /** @class */ (function () {
    function Nps(payload) {
        this.payload = payload;
        this.type = SAVE_NPS;
    }
    return Nps;
}());
export { Nps };
if (false) {
    /** @type {?} */
    Nps.prototype.type;
    /** @type {?} */
    Nps.prototype.payload;
}
var YesNo = /** @class */ (function () {
    function YesNo(payload) {
        this.payload = payload;
        this.type = SAVE_YES_NO;
    }
    return YesNo;
}());
export { YesNo };
if (false) {
    /** @type {?} */
    YesNo.prototype.type;
    /** @type {?} */
    YesNo.prototype.payload;
}
var Rating = /** @class */ (function () {
    function Rating(payload) {
        this.payload = payload;
        this.type = SAVE_RATING;
    }
    return Rating;
}());
export { Rating };
if (false) {
    /** @type {?} */
    Rating.prototype.type;
    /** @type {?} */
    Rating.prototype.payload;
}
var DropDown = /** @class */ (function () {
    function DropDown(payload) {
        this.payload = payload;
        this.type = SAVE_DROP_DOWN;
    }
    return DropDown;
}());
export { DropDown };
if (false) {
    /** @type {?} */
    DropDown.prototype.type;
    /** @type {?} */
    DropDown.prototype.payload;
}
var LongText = /** @class */ (function () {
    function LongText(payload) {
        this.payload = payload;
        this.type = SAVE_LONG_TEXT;
    }
    return LongText;
}());
export { LongText };
if (false) {
    /** @type {?} */
    LongText.prototype.type;
    /** @type {?} */
    LongText.prototype.payload;
}
var ShortText = /** @class */ (function () {
    function ShortText(payload) {
        this.payload = payload;
        this.type = SAVE_SHORT_TEXT;
    }
    return ShortText;
}());
export { ShortText };
if (false) {
    /** @type {?} */
    ShortText.prototype.type;
    /** @type {?} */
    ShortText.prototype.payload;
}
var RemoveResponses = /** @class */ (function () {
    function RemoveResponses(payload) {
        this.payload = payload;
        this.type = REMOVE_RESPONSES;
    }
    return RemoveResponses;
}());
export { RemoveResponses };
if (false) {
    /** @type {?} */
    RemoveResponses.prototype.type;
    /** @type {?} */
    RemoveResponses.prototype.payload;
}
var RemoveResponsesOfExcludedQuestionIds = /** @class */ (function () {
    function RemoveResponsesOfExcludedQuestionIds(payload) {
        this.payload = payload;
        this.type = REMOVE_RESPONSES_OF_EXCLUDED_QUESTION_IDS;
    }
    return RemoveResponsesOfExcludedQuestionIds;
}());
export { RemoveResponsesOfExcludedQuestionIds };
if (false) {
    /** @type {?} */
    RemoveResponsesOfExcludedQuestionIds.prototype.type;
    /** @type {?} */
    RemoveResponsesOfExcludedQuestionIds.prototype.payload;
}
var Reset = /** @class */ (function () {
    function Reset() {
        this.type = RESET_RESPONSE;
    }
    return Reset;
}());
export { Reset };
if (false) {
    /** @type {?} */
    Reset.prototype.type;
}
var CurrentState = /** @class */ (function () {
    function CurrentState() {
        this.type = CURRENT_STATE;
    }
    return CurrentState;
}());
export { CurrentState };
if (false) {
    /** @type {?} */
    CurrentState.prototype.type;
}
