// survey-lib/esm5/lib/pages/pages.model.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/pages.model.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var PagesModel = /** @class */ (function () {
    function PagesModel() {
    }
    Object.defineProperty(PagesModel.prototype, "testMode", {
        get: /**
         * @return {?}
         */
        function () {
            return this._testMode;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._testMode = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PagesModel.prototype, "id", {
        get: /**
         * @return {?}
         */
        function () {
            return this._id;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._id = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PagesModel.prototype, "userId", {
        get: /**
         * @return {?}
         */
        function () {
            return this._userId;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._userId = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PagesModel.prototype, "themeId", {
        get: /**
         * @return {?}
         */
        function () {
            return this._themeId;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._themeId = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PagesModel.prototype, "logoUrl", {
        get: /**
         * @return {?}
         */
        function () {
            return this._logoUrl;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._logoUrl = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PagesModel.prototype, "surveyContent", {
        get: /**
         * @return {?}
         */
        function () {
            return this._surveyContent;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._surveyContent = value;
        },
        enumerable: true,
        configurable: true
    });
    return PagesModel;
}());
export { PagesModel };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PagesModel.prototype._id;
    /**
     * @type {?}
     * @private
     */
    PagesModel.prototype._userId;
    /**
     * @type {?}
     * @private
     */
    PagesModel.prototype._themeId;
    /**
     * @type {?}
     * @private
     */
    PagesModel.prototype._testMode;
    /**
     * @type {?}
     * @private
     */
    PagesModel.prototype._logoUrl;
    /**
     * @type {?}
     * @private
     */
    PagesModel.prototype._surveyContent;
}
var SurveyContent = /** @class */ (function () {
    function SurveyContent() {
    }
    Object.defineProperty(SurveyContent.prototype, "name", {
        get: /**
         * @return {?}
         */
        function () {
            return this._name;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._name = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SurveyContent.prototype, "description", {
        get: /**
         * @return {?}
         */
        function () {
            return this._description;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._description = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SurveyContent.prototype, "pages", {
        get: /**
         * @return {?}
         */
        function () {
            return this._pages;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._pages = value;
        },
        enumerable: true,
        configurable: true
    });
    return SurveyContent;
}());
export { SurveyContent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    SurveyContent.prototype._name;
    /**
     * @type {?}
     * @private
     */
    SurveyContent.prototype._description;
    /**
     * @type {?}
     * @private
     */
    SurveyContent.prototype._pages;
    /** @type {?} */
    SurveyContent.prototype.selectedTheme;
    /** @type {?} */
    SurveyContent.prototype.organizationLogo;
}
var SelectedThemeModel = /** @class */ (function () {
    function SelectedThemeModel() {
    }
    return SelectedThemeModel;
}());
export { SelectedThemeModel };
if (false) {
    /** @type {?} */
    SelectedThemeModel.prototype.name;
    /** @type {?} */
    SelectedThemeModel.prototype.font;
    /** @type {?} */
    SelectedThemeModel.prototype.colors;
}
var Font = /** @class */ (function () {
    function Font() {
    }
    return Font;
}());
export { Font };
if (false) {
    /** @type {?} */
    Font.prototype.name;
    /** @type {?} */
    Font.prototype.url;
}
var Colors = /** @class */ (function () {
    function Colors() {
    }
    return Colors;
}());
export { Colors };
if (false) {
    /** @type {?} */
    Colors.prototype.backgroundColor;
    /** @type {?} */
    Colors.prototype.primaryColor;
    /** @type {?} */
    Colors.prototype.primaryColorDark;
    /** @type {?} */
    Colors.prototype.footerBackgroundColor;
    /** @type {?} */
    Colors.prototype.footerSecondaryColor;
    /** @type {?} */
    Colors.prototype.footerPrimaryColor;
    /** @type {?} */
    Colors.prototype.textColorPrimary;
    /** @type {?} */
    Colors.prototype.textColorSecondary;
    /** @type {?} */
    Colors.prototype.primaryButtonTextColor;
}
var Page = /** @class */ (function () {
    function Page() {
    }
    Object.defineProperty(Page.prototype, "pageName", {
        get: /**
         * @return {?}
         */
        function () {
            return this._pageName;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._pageName = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Page.prototype, "properties", {
        get: /**
         * @return {?}
         */
        function () {
            return this._properties;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._properties = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Page.prototype, "questions", {
        get: /**
         * @return {?}
         */
        function () {
            return this._questions;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._questions = value;
        },
        enumerable: true,
        configurable: true
    });
    return Page;
}());
export { Page };
if (false) {
    /**
     * @type {?}
     * @private
     */
    Page.prototype._pageName;
    /**
     * @type {?}
     * @private
     */
    Page.prototype._properties;
    /**
     * @type {?}
     * @private
     */
    Page.prototype._questions;
}
var Question = /** @class */ (function () {
    function Question() {
    }
    return Question;
}());
export { Question };
if (false) {
    /** @type {?} */
    Question.prototype.identifier;
    /** @type {?} */
    Question.prototype.questionId;
    /** @type {?} */
    Question.prototype.questionType;
    /** @type {?} */
    Question.prototype.question;
    /** @type {?} */
    Question.prototype.choices;
    /** @type {?} */
    Question.prototype.others;
    /** @type {?} */
    Question.prototype.attributes;
    /** @type {?} */
    Question.prototype.pictures;
    /** @type {?} */
    Question.prototype.rankingItems;
    /** @type {?} */
    Question.prototype.rows;
    /** @type {?} */
    Question.prototype.columns;
    /** @type {?} */
    Question.prototype.contactForm;
    /** @type {?} */
    Question.prototype.addressForm;
    /** @type {?} */
    Question.prototype.logicJumps;
}
var Others = /** @class */ (function () {
    function Others() {
    }
    return Others;
}());
export { Others };
if (false) {
    /** @type {?} */
    Others.prototype.value;
    /** @type {?} */
    Others.prototype.isChecked;
    /** @type {?} */
    Others.prototype.showCheckbox;
}
var LogicJumps = /** @class */ (function () {
    function LogicJumps() {
    }
    return LogicJumps;
}());
export { LogicJumps };
if (false) {
    /** @type {?} */
    LogicJumps.prototype.blocks;
    /** @type {?} */
    LogicJumps.prototype.defaultJump;
}
var Block = /** @class */ (function () {
    function Block() {
    }
    return Block;
}());
export { Block };
if (false) {
    /** @type {?} */
    Block.prototype.that;
    /** @type {?} */
    Block.prototype.thatIndex;
    /** @type {?} */
    Block.prototype.conditions;
}
var Condition = /** @class */ (function () {
    function Condition() {
    }
    return Condition;
}());
export { Condition };
if (false) {
    /** @type {?} */
    Condition.prototype.condition;
    /** @type {?} */
    Condition.prototype._if;
    /** @type {?} */
    Condition.prototype._this;
    /** @type {?} */
    Condition.prototype.relationWithNextCondition;
}
var RankingItem = /** @class */ (function () {
    function RankingItem() {
    }
    return RankingItem;
}());
export { RankingItem };
if (false) {
    /** @type {?} */
    RankingItem.prototype.itemId;
    /** @type {?} */
    RankingItem.prototype.text;
}
var ContactForm = /** @class */ (function () {
    function ContactForm() {
    }
    return ContactForm;
}());
export { ContactForm };
if (false) {
    /** @type {?} */
    ContactForm.prototype.label;
    /** @type {?} */
    ContactForm.prototype.type;
    /** @type {?} */
    ContactForm.prototype.value;
    /** @type {?} */
    ContactForm.prototype.validation;
    /** @type {?} */
    ContactForm.prototype.keyName;
}
var Validation = /** @class */ (function () {
    function Validation() {
    }
    return Validation;
}());
export { Validation };
if (false) {
    /** @type {?} */
    Validation.prototype.required;
}
var Column = /** @class */ (function () {
    function Column() {
    }
    return Column;
}());
export { Column };
if (false) {
    /** @type {?} */
    Column.prototype.columnId;
    /** @type {?} */
    Column.prototype.label;
    /** @type {?} */
    Column.prototype.value;
}
var Picture = /** @class */ (function () {
    function Picture() {
    }
    Object.defineProperty(Picture.prototype, "id", {
        get: /**
         * @return {?}
         */
        function () {
            return this._id;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._id = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Picture.prototype, "picture", {
        get: /**
         * @return {?}
         */
        function () {
            return this._picture;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._picture = value;
        },
        enumerable: true,
        configurable: true
    });
    return Picture;
}());
export { Picture };
if (false) {
    /**
     * @type {?}
     * @private
     */
    Picture.prototype._id;
    /**
     * @type {?}
     * @private
     */
    Picture.prototype._picture;
}
var AddressForm = /** @class */ (function () {
    function AddressForm() {
    }
    Object.defineProperty(AddressForm.prototype, "text", {
        get: /**
         * @return {?}
         */
        function () {
            return this._text;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._text = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AddressForm.prototype, "required", {
        get: /**
         * @return {?}
         */
        function () {
            return this._required;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._required = value;
        },
        enumerable: true,
        configurable: true
    });
    return AddressForm;
}());
export { AddressForm };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AddressForm.prototype._text;
    /**
     * @type {?}
     * @private
     */
    AddressForm.prototype._required;
}
var Row = /** @class */ (function () {
    function Row() {
    }
    Object.defineProperty(Row.prototype, "id", {
        get: /**
         * @return {?}
         */
        function () {
            return this._id;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._id = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Row.prototype, "text", {
        get: /**
         * @return {?}
         */
        function () {
            return this._text;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._text = value;
        },
        enumerable: true,
        configurable: true
    });
    return Row;
}());
export { Row };
if (false) {
    /**
     * @type {?}
     * @private
     */
    Row.prototype._id;
    /**
     * @type {?}
     * @private
     */
    Row.prototype._text;
}
var Attributes = /** @class */ (function () {
    function Attributes() {
    }
    return Attributes;
}());
export { Attributes };
if (false) {
    /** @type {?} */
    Attributes.prototype.image;
    /** @type {?} */
    Attributes.prototype.description;
    /** @type {?} */
    Attributes.prototype.vertical;
    /** @type {?} */
    Attributes.prototype.video;
    /** @type {?} */
    Attributes.prototype.multipleChoice;
    /** @type {?} */
    Attributes.prototype.randomize;
    /** @type {?} */
    Attributes.prototype.required;
    /** @type {?} */
    Attributes.prototype.requiredErrorMessage;
    /** @type {?} */
    Attributes.prototype.others;
    /** @type {?} */
    Attributes.prototype.maxCharacter;
    /** @type {?} */
    Attributes.prototype.quotation;
    /** @type {?} */
    Attributes.prototype.minimumNumber;
    /** @type {?} */
    Attributes.prototype.maximumNumber;
    /** @type {?} */
    Attributes.prototype.numberOfSelection;
    /** @type {?} */
    Attributes.prototype.defaultCountryCode;
    /** @type {?} */
    Attributes.prototype.alphabeticalSort;
    /** @type {?} */
    Attributes.prototype.includeAddressForm;
    /** @type {?} */
    Attributes.prototype.startScaleAt;
    /** @type {?} */
    Attributes.prototype.leftLabel;
    /** @type {?} */
    Attributes.prototype.centerLabel;
    /** @type {?} */
    Attributes.prototype.rightLabel;
    /** @type {?} */
    Attributes.prototype.isDescriptionEnabled;
    /** @type {?} */
    Attributes.prototype.steps;
    /** @type {?} */
    Attributes.prototype.ratingIcon;
    /** @type {?} */
    Attributes.prototype.selection;
    /** @type {?} */
    Attributes.prototype.maxCharacterCount;
    /** @type {?} */
    Attributes.prototype.minCharacterCount;
}
var Selection = /** @class */ (function () {
    function Selection() {
    }
    return Selection;
}());
export { Selection };
if (false) {
    /** @type {?} */
    Selection.prototype.type;
    /** @type {?} */
    Selection.prototype.exact;
    /** @type {?} */
    Selection.prototype.rangeFrom;
    /** @type {?} */
    Selection.prototype.rangeTo;
}
var Step = /** @class */ (function () {
    function Step() {
    }
    return Step;
}());
export { Step };
if (false) {
    /** @type {?} */
    Step.prototype.label;
    /** @type {?} */
    Step.prototype.value;
}
var Choice = /** @class */ (function () {
    function Choice() {
    }
    return Choice;
}());
export { Choice };
if (false) {
    /** @type {?} */
    Choice.prototype.identifier;
    /** @type {?} */
    Choice.prototype.choiceId;
    /** @type {?} */
    Choice.prototype.text;
    /** @type {?} */
    Choice.prototype.isChecked;
}
var Properties = /** @class */ (function () {
    function Properties() {
    }
    Object.defineProperty(Properties.prototype, "buttonText", {
        get: /**
         * @return {?}
         */
        function () {
            return this._buttonText;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._buttonText = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Properties.prototype, "subHeadingValue", {
        get: /**
         * @return {?}
         */
        function () {
            return this._subHeadingValue;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._subHeadingValue = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Properties.prototype, "headingValue", {
        get: /**
         * @return {?}
         */
        function () {
            return this._headingValue;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._headingValue = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Properties.prototype, "backgroundImage", {
        get: /**
         * @return {?}
         */
        function () {
            return this._backgroundImage;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._backgroundImage = value;
        },
        enumerable: true,
        configurable: true
    });
    return Properties;
}());
export { Properties };
if (false) {
    /**
     * @type {?}
     * @private
     */
    Properties.prototype._headingValue;
    /**
     * @type {?}
     * @private
     */
    Properties.prototype._subHeadingValue;
    /**
     * @type {?}
     * @private
     */
    Properties.prototype._backgroundImage;
    /**
     * @type {?}
     * @private
     */
    Properties.prototype._buttonText;
}
