// ngrx-forms/esm2015/src/group/directive.js
import * as tslib_1 from "tslib";
import { Directive, HostListener, Inject, Input, Optional } from '@angular/core';
import { ActionsSubject } from '@ngrx/store';
import { MarkAsSubmittedAction } from '../actions';
let NgrxFormDirective = class NgrxFormDirective {
    constructor(actionsSubject) {
        this.actionsSubject = actionsSubject;
        this.actionsSubject = actionsSubject;
    }
    dispatchAction(action) {
        if (this.actionsSubject !== null) {
            this.actionsSubject.next(action);
        }
        else {
            throw new Error('ActionsSubject must be present in order to dispatch actions!');
        }
    }
    ngOnInit() {
        if (!this.state) {
            throw new Error('The form state must not be undefined!');
        }
    }
    onSubmit(event) {
        event.preventDefault();
        if (this.state.isUnsubmitted) {
            this.dispatchAction(new MarkAsSubmittedAction(this.state.id));
        }
    }
};
tslib_1.__decorate([
    Input('ngrxFormState'),
    tslib_1.__metadata("design:type", Object)
], NgrxFormDirective.prototype, "state", void 0);
tslib_1.__decorate([
    HostListener('submit', ['$event']),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object]),
    tslib_1.__metadata("design:returntype", void 0)
], NgrxFormDirective.prototype, "onSubmit", null);
NgrxFormDirective = tslib_1.__decorate([
    Directive({
        // tslint:disable-next-line:directive-selector
        selector: 'form:not([ngrxFormsAction])[ngrxFormState]',
    }),
    tslib_1.__param(0, Optional()), tslib_1.__param(0, Inject(ActionsSubject)),
    tslib_1.__metadata("design:paramtypes", [Object])
], NgrxFormDirective);
export { NgrxFormDirective };
