// ngrx-forms/esm5/src/view-adapter/option.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Host, Input, Optional, Renderer2, } from '@angular/core';
import { NgrxSelectViewAdapter } from './select';
import { NgrxSelectMultipleViewAdapter } from './select-multiple';
var ɵ0 = function () { return void 0; };
// tslint:disable:directive-class-suffix
var NULL_RENDERER = {
    setProperty: ɵ0,
};
/**
 * This directive is necessary to restore the default behaviour of Angular
 * when an `option` is used without an **ngrx-forms** form state. Since it
 * is not possible to select an element with a selector that considers its
 * parent the `option` directives for `select` and `select[multiple]` will
 * always be applied and therefore overriding the `[value]` binding which
 * disabled Angular's normal behaviour. This directive restores this
 * behaviour if no `select` or `select[multiple]` view adapter is found.
 * This is not a perfect solution since it may interfere with other
 * directives that try to set the `[value]` but that is very unlikely.
 */
var NgrxFallbackSelectOption = /** @class */ (function () {
    function NgrxFallbackSelectOption(element, renderer, viewAdapter, multipleViewAdapter) {
        this.element = element;
        this.renderer = renderer;
        this.renderer = viewAdapter || multipleViewAdapter ? NULL_RENDERER : renderer;
    }
    Object.defineProperty(NgrxFallbackSelectOption.prototype, "value", {
        set: function (value) {
            this.renderer.setProperty(this.element.nativeElement, 'value', value);
        },
        enumerable: true,
        configurable: true
    });
    tslib_1.__decorate([
        Input('value'),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], NgrxFallbackSelectOption.prototype, "value", null);
    NgrxFallbackSelectOption = tslib_1.__decorate([
        Directive({
            // tslint:disable-next-line:directive-selector
            selector: 'option',
        }),
        tslib_1.__param(2, Host()), tslib_1.__param(2, Optional()),
        tslib_1.__param(3, Host()), tslib_1.__param(3, Optional()),
        tslib_1.__metadata("design:paramtypes", [ElementRef,
            Renderer2,
            NgrxSelectViewAdapter,
            NgrxSelectMultipleViewAdapter])
    ], NgrxFallbackSelectOption);
    return NgrxFallbackSelectOption;
}());
export { NgrxFallbackSelectOption };
export { ɵ0 };
