// ngrx-forms/esm5/src/view-adapter/radio.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, forwardRef, HostListener, Input, Renderer2 } from '@angular/core';
import { NGRX_FORM_VIEW_ADAPTER } from './view-adapter';
// tslint:disable:directive-class-suffix
var NgrxRadioViewAdapter = /** @class */ (function () {
    function NgrxRadioViewAdapter(renderer, elementRef) {
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.nativeNameWasSet = false;
        this.onChange = function () { return void 0; };
        this.onTouched = function () { return void 0; };
    }
    NgrxRadioViewAdapter_1 = NgrxRadioViewAdapter;
    Object.defineProperty(NgrxRadioViewAdapter.prototype, "value", {
        set: function (val) {
            if (val !== this.latestValue) {
                this.latestValue = val;
                if (this.isChecked) {
                    this.onChange();
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NgrxRadioViewAdapter.prototype, "ngrxFormControlState", {
        set: function (value) {
            if (!value) {
                throw new Error('The control state must not be undefined!');
            }
            this.state = value;
            var nativeName = this.elementRef.nativeElement.name;
            var shouldSetNativeName = value.id !== nativeName && this.nativeNameWasSet;
            if (shouldSetNativeName) {
                this.renderer.setProperty(this.elementRef.nativeElement, 'name', value.id);
            }
        },
        enumerable: true,
        configurable: true
    });
    NgrxRadioViewAdapter.prototype.ngOnInit = function () {
        this.isChecked = this.elementRef.nativeElement.checked;
    };
    NgrxRadioViewAdapter.prototype.ngAfterViewInit = function () {
        var nativeName = this.elementRef.nativeElement.name;
        var shouldSetNativeName = this.state.id !== nativeName && !nativeName;
        if (shouldSetNativeName) {
            this.renderer.setProperty(this.elementRef.nativeElement, 'name', this.state.id);
            this.nativeNameWasSet = true;
        }
    };
    NgrxRadioViewAdapter.prototype.setViewValue = function (value) {
        this.isChecked = value === this.latestValue;
        this.renderer.setProperty(this.elementRef.nativeElement, 'checked', this.isChecked);
    };
    NgrxRadioViewAdapter.prototype.setOnChangeCallback = function (fn) {
        var _this = this;
        this.onChange = function () { return fn(_this.latestValue); };
    };
    NgrxRadioViewAdapter.prototype.setOnTouchedCallback = function (fn) {
        this.onTouched = fn;
    };
    NgrxRadioViewAdapter.prototype.setIsDisabled = function (isDisabled) {
        this.renderer.setProperty(this.elementRef.nativeElement, 'disabled', isDisabled);
    };
    var NgrxRadioViewAdapter_1;
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], NgrxRadioViewAdapter.prototype, "value", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], NgrxRadioViewAdapter.prototype, "ngrxFormControlState", null);
    tslib_1.__decorate([
        HostListener('change'),
        tslib_1.__metadata("design:type", Function)
    ], NgrxRadioViewAdapter.prototype, "onChange", void 0);
    tslib_1.__decorate([
        HostListener('blur'),
        tslib_1.__metadata("design:type", Function)
    ], NgrxRadioViewAdapter.prototype, "onTouched", void 0);
    NgrxRadioViewAdapter = NgrxRadioViewAdapter_1 = tslib_1.__decorate([
        Directive({
            // tslint:disable-next-line:directive-selector
            selector: 'input[type=radio][ngrxFormControlState]',
            providers: [{
                    provide: NGRX_FORM_VIEW_ADAPTER,
                    useExisting: forwardRef(function () { return NgrxRadioViewAdapter_1; }),
                    multi: true,
                }],
        }),
        tslib_1.__metadata("design:paramtypes", [Renderer2,
            ElementRef])
    ], NgrxRadioViewAdapter);
    return NgrxRadioViewAdapter;
}());
export { NgrxRadioViewAdapter };
