// survey-lib/esm2015/lib/shared/pages-services/pages.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/shared/pages-services/pages.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Injectable } from '@angular/core';
import { SurveyResponses } from '../../pages/survey-response.model';
export class PagesService {
    constructor() {
        this.previousQuestion = new EventEmitter();
        this.nextQuestion = new EventEmitter();
        this.nextPageEvent = new EventEmitter();
        this.surveyResponses = [];
        this.sequenceOfQuestionTraversed = [];
    }
    /**
     * @param {?} response
     * @return {?}
     */
    persistSurvey(response) {
        /** @type {?} */
        let index = this.surveyResponses.findIndex((/**
         * @param {?} surveyResponse
         * @return {?}
         */
        surveyResponse => surveyResponse.identifier === response.identifier));
        if (index !== -1) {
            this.surveyResponses[index] = response;
        }
        else {
            this.surveyResponses.push(response);
        }
    }
    /**
     * @param {?} payload
     * @return {?}
     */
    removeResponsesOfExcludedQuestionIds(payload) {
        return this.surveyResponses.filter((/**
         * @param {?} response
         * @return {?}
         */
        response => payload.find((/**
         * @param {?} questionId
         * @return {?}
         */
        questionId => questionId == response.questionId))));
    }
    /**
     * @param {?} identifier
     * @return {?}
     */
    getResponse(identifier) {
        console.log(this.surveyResponses);
        /** @type {?} */
        let index = this.surveyResponses.findIndex((/**
         * @param {?} res
         * @return {?}
         */
        res => res.identifier == identifier));
        if (index > -1) {
            return this.surveyResponses[index];
        }
        else {
            return new SurveyResponses();
        }
    }
}
PagesService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
PagesService.ctorParameters = () => [];
if (false) {
    /**
     * @type {?}
     * @private
     */
    PagesService.prototype.surveyUrl;
    /**
     * @type {?}
     * @private
     */
    PagesService.prototype.submitResponseUrl;
    /** @type {?} */
    PagesService.prototype.previousQuestion;
    /** @type {?} */
    PagesService.prototype.nextQuestion;
    /** @type {?} */
    PagesService.prototype.nextPageEvent;
    /** @type {?} */
    PagesService.prototype.surveyResponses;
    /** @type {?} */
    PagesService.prototype.sequenceOfQuestionTraversed;
    /** @type {?} */
    PagesService.prototype.stage;
}
