// survey-lib/esm5/lib/pages/questionnaire/address/address.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pages/questionnaire/address/address.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { Question } from '../../pages.model';
var AddressComponent = /** @class */ (function () {
    function AddressComponent() {
    }
    /**
     * @return {?}
     */
    AddressComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    AddressComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-address',
                    template: "<div>\n  <span *ngIf=\"addressForm.questionType == 'Address Form'\">\n    <span *ngIf=\"addressForm.attributes.required\" class=\"isRequired\">*</span>\n    {{addressForm.question}}\n  </span>\n  <table *ngFor=\"let address of addressForm.addressForm; index as addressIndex\">\n    <tr>\n      <td *ngIf=\"address.text == 'Street 1'\">\n        <label for=\"street 1\">\n          <input type=\"text\" id=\"street 1\" placeholder=\"Street 1\">\n        </label>\n      </td>\n\n      <td *ngIf=\"address.text == 'Street 2'\">\n        <label for=\"street 2\">\n          <input type=\"text\" id=\"street 2\" placeholder=\"Street 2\">\n        </label>\n      </td>\n\n      <td *ngIf=\"address.text == 'City'\">\n        <label for=\"city\">\n          <input type=\"text\" id=\"city\" placeholder=\"City\">\n        </label>\n      </td>\n\n      <td *ngIf=\"address.text == 'State'\">\n        <label for=\"state\">\n          <input type=\"text\" id=\"state\" placeholder=\"State\">\n        </label>\n      </td>\n\n      <td *ngIf=\"address.text == 'Country'\">\n        <label for=\"country\">\n          <input type=\"text\" id=\"country\" placeholder=\"Country\">\n        </label>\n      </td>\n\n      <td *ngIf=\"address.text == 'Zip code'\">\n        <label for=\"zip-code\">\n          <input type=\"text\" id=\"zip-code\" placeholder=\"Zip code\">\n        </label>\n      </td>\n\n    </tr>\n  </table>\n</div>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    AddressComponent.ctorParameters = function () { return []; };
    AddressComponent.propDecorators = {
        addressForm: [{ type: Input }]
    };
    return AddressComponent;
}());
export { AddressComponent };
if (false) {
    /** @type {?} */
    AddressComponent.prototype.addressForm;
}
